<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Boris Steipe (boris.steipe@utoronto.ca)">
<meta name="dcterms.date" content="2025-09-29">

<title>Computing Amino Acid Similarity</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="computingAminoAcidSimilarity_files/libs/clipboard/clipboard.min.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/quarto-html/quarto.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/quarto-html/popper.min.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/quarto-html/anchor.min.js"></script>
<link href="computingAminoAcidSimilarity_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="computingAminoAcidSimilarity_files/libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="computingAminoAcidSimilarity_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="computingAminoAcidSimilarity_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="computingAminoAcidSimilarity_files/libs/bootstrap/bootstrap-c0367b04c37547644fece4185067e4a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="computingAminoAcidSimilarity_files/libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="computingAminoAcidSimilarity_files/libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/plotly-binding-4.11.0/plotly.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/typedarray-0.1/typedarray.min.js"></script>
<script src="computingAminoAcidSimilarity_files/libs/jquery-3.5.1/jquery.min.js"></script>
<link href="computingAminoAcidSimilarity_files/libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet">
<script src="computingAminoAcidSimilarity_files/libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>
<link href="computingAminoAcidSimilarity_files/libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="computingAminoAcidSimilarity_files/libs/plotly-main-2.11.1/plotly-latest.min.js"></script>
<style>
  .version-info {
    font-size: 80%;
    color: #AAAAAE;
    font-style: italic;
  }
</style>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#initializations." id="toc-initializations." class="nav-link active" data-scroll-target="#initializations."><span class="header-section-number">1</span> Initializations.</a></li>
  <li><a href="#preparing-the-data" id="toc-preparing-the-data" class="nav-link" data-scroll-target="#preparing-the-data"><span class="header-section-number">2</span> Preparing the data</a>
  <ul class="collapse">
  <li><a href="#scaling-the-scales" id="toc-scaling-the-scales" class="nav-link" data-scroll-target="#scaling-the-scales"><span class="header-section-number">2.1</span> Scaling the scales</a></li>
  </ul></li>
  <li><a href="#a-table-of-amino-acid-scale-types-and-values" id="toc-a-table-of-amino-acid-scale-types-and-values" class="nav-link" data-scroll-target="#a-table-of-amino-acid-scale-types-and-values"><span class="header-section-number">3</span> A table of Amino Acid scale types and values</a>
  <ul class="collapse">
  <li><a href="#selecting-categories-of-amino-acid-scales" id="toc-selecting-categories-of-amino-acid-scales" class="nav-link" data-scroll-target="#selecting-categories-of-amino-acid-scales"><span class="header-section-number">3.1</span> Selecting categories of amino acid scales</a></li>
  </ul></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span class="header-section-number">4</span> Principal Component Analysis (PCA)</a>
  <ul class="collapse">
  <li><a href="#computing-the-pca" id="toc-computing-the-pca" class="nav-link" data-scroll-target="#computing-the-pca"><span class="header-section-number">4.1</span> Computing the PCA</a></li>
  <li><a href="#selecting-important-components" id="toc-selecting-important-components" class="nav-link" data-scroll-target="#selecting-important-components"><span class="header-section-number">4.2</span> Selecting “important” components</a></li>
  <li><a href="#constructing-a-feature-space" id="toc-constructing-a-feature-space" class="nav-link" data-scroll-target="#constructing-a-feature-space"><span class="header-section-number">4.3</span> Constructing a Feature Space</a></li>
  </ul></li>
  <li><a href="#relationships-between-amino-acids" id="toc-relationships-between-amino-acids" class="nav-link" data-scroll-target="#relationships-between-amino-acids"><span class="header-section-number">5</span> Relationships between amino acids</a>
  <ul class="collapse">
  <li><a href="#summarize-and-save" id="toc-summarize-and-save" class="nav-link" data-scroll-target="#summarize-and-save"><span class="header-section-number">5.1</span> Summarize and Save</a></li>
  </ul></li>
  <li><a href="#milestone-amino-acid-feature-space" id="toc-milestone-amino-acid-feature-space" class="nav-link" data-scroll-target="#milestone-amino-acid-feature-space"><span class="header-section-number">6</span> MILESTONE: Amino Acid Feature Space</a></li>
  <li><a href="#interpreting-the-principal-components" id="toc-interpreting-the-principal-components" class="nav-link" data-scroll-target="#interpreting-the-principal-components"><span class="header-section-number">7</span> Interpreting the Principal Components</a>
  <ul class="collapse">
  <li><a href="#corelations-between-pcs-and-observed-scales" id="toc-corelations-between-pcs-and-observed-scales" class="nav-link" data-scroll-target="#corelations-between-pcs-and-observed-scales"><span class="header-section-number">7.1</span> Corelations between PCs and observed scales</a></li>
  <li><a href="#examine-the-dimensions" id="toc-examine-the-dimensions" class="nav-link" data-scroll-target="#examine-the-dimensions"><span class="header-section-number">7.2</span> Examine the dimensions</a></li>
  <li><a href="#global-interpretation-of-axes-and-scales" id="toc-global-interpretation-of-axes-and-scales" class="nav-link" data-scroll-target="#global-interpretation-of-axes-and-scales"><span class="header-section-number">7.3</span> Global Interpretation of Axes and Scales</a>
  <ul class="collapse">
  <li><a href="#interpreting-in-2d" id="toc-interpreting-in-2d" class="nav-link" data-scroll-target="#interpreting-in-2d"><span class="header-section-number">7.3.1</span> Interpreting in 2D</a></li>
  <li><a href="#joint-t-sne-embedding-of-scales-pcs-2d" id="toc-joint-t-sne-embedding-of-scales-pcs-2d" class="nav-link" data-scroll-target="#joint-t-sne-embedding-of-scales-pcs-2d"><span class="header-section-number">7.3.2</span> Joint t-SNE embedding of scales + PCs (2D)</a></li>
  <li><a href="#joint-t-sne-embedding-of-pcs-and-scales-in-3d" id="toc-joint-t-sne-embedding-of-pcs-and-scales-in-3d" class="nav-link" data-scroll-target="#joint-t-sne-embedding-of-pcs-and-scales-in-3d"><span class="header-section-number">7.3.3</span> Joint t-SNE embedding of PCs and scales in 3d</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#amino-acid-similarity---the-function-aasim" id="toc-amino-acid-similarity---the-function-aasim" class="nav-link" data-scroll-target="#amino-acid-similarity---the-function-aasim"><span class="header-section-number">8</span> Amino Acid Similarity - the Function <code>aaSim()</code></a>
  <ul class="collapse">
  <li><a href="#test-aasim-and-validate" id="toc-test-aasim-and-validate" class="nav-link" data-scroll-target="#test-aasim-and-validate"><span class="header-section-number">8.1</span> Test <code>aaSim()</code> and validate</a>
  <ul class="collapse">
  <li><a href="#distinct-amino-acids" id="toc-distinct-amino-acids" class="nav-link" data-scroll-target="#distinct-amino-acids"><span class="header-section-number">8.1.1</span> Distinct Amino Acids</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Computing Amino Acid Similarity</h1>
<p class="subtitle lead">Data handling, PCA, and Feature Space</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Boris Steipe (boris.steipe@utoronto.ca) </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 29, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!--
Version history:
  v1.1 Fix project directory issue (syntax fixed in ./_quarto.yml), add
       version as metadata, add a css block; clean up file and directory paths
       to work relatively to ./ . Add interpretation of space axes (PCs)
       relative to original scales using a 2D t-SNE embedding. Add 3D t-SNE. 
       Completely revise text and add many notes on coding principles.
  v1.0 First quarto version, based on prior teaching code in
       aminoAcidSimlarity.R and aaSim.R .
-->
<div class="version-info">
<p>Version v1.1, last updated September 29, 2025</p>
</div>
<p>In this quarto document, we combine some ideas we discussed in class about amino acid properties, and how to construct a function that can compute values of amino acid similarity that we can use to evaluate the structure of a genetic code.</p>
<section id="initializations." class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Initializations.</h1>
<p>We first check the existence of required files, and whether your course directory is “in a sane state”. That goes on under the hood of the first code chunk - you can inspect in in the <code>qmd</code> source, but it is hidden in the HTML document.</p>
</section>
<section id="preparing-the-data" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Preparing the data</h1>
<p>First, we read the sheets we need from the downloaded, published excel files that were distributed with AAontology paper (<a href="https://www.sciencedirect.com/science/article/pii/S0022283624003267">Breimann <em>et al.</em>, 2024</a>). These are the <em>Supplementary Tables</em> 1 and 3.</p>
<p>I often use names like “mySomething” as variable names, to avoid conflicts with existing names or reserved words in the language:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>  myScales <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">"dat/Breimann_2024_Supplementary_Table_1.xlsx"</span>,</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sheet =</span> <span class="st">"Normalized"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  myCats   <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">"dat/Breimann_2024_Supplementary_Table_3.xlsx"</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sheet =</span> <span class="st">"Scales"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># These are returned as "tibbles", we simplify them to data frames</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  myScales <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(myScales)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  myCats   <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(myCats)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The first column of myScales contains amino acid labels. We assign </span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># these to rownames, and then remove that column:</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(myScales) <span class="ot">&lt;-</span> myScales[ , <span class="dv">1</span>]</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  myScales[ , <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Always examine the result and check for sanity</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  myScales[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ANDN920101 ARGP820101 ARGP820102 ARGP820103 BEGF750101 BEGF750102 BEGF750103
A      0.494      0.230      0.355      0.504      1.000      0.512      0.000
C      0.864      0.404      0.579      0.387      0.000      0.233      0.783
D      1.000      0.174      0.000      0.000      0.404      0.233      1.000
E      0.420      0.177      0.019      0.032      0.713      0.000      0.267
F      0.877      0.762      0.601      0.670      0.574      1.000      0.267</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str</span>(myCats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   586 obs. of  5 variables:
 $ scale_id         : chr  "LINS030110" "LINS030113" "JANJ780101" "JANJ780103" ...
 $ category         : chr  "ASA/Volume" "ASA/Volume" "ASA/Volume" "ASA/Volume" ...
 $ subcategory      : chr  "Accessible surface area (ASA)" "Accessible surface area (ASA)" "Accessible surface area (ASA)" "Accessible surface area (ASA)" ...
 $ scale_name       : chr  "ASA (folded coil/turn)" "ASA (folded coil/turn)" "ASA (folded protein)" "ASA (folded protein)" ...
 $ scale_description: chr  "Total median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)" "% total accessible surfaces of whole residues in random coil/turn structures (Lins et al., 2003)" "Average accessible surface area (Janin et al., 1978)" "Percentage of exposed residues (Janin et al., 1978)" ...</code></pre>
</div>
</div>
<p>Confirm that the IDs in both tables are the same. This kind of validation is important. Here we use the function <code>setequal()</code> since we don’t assume that the order matches, only the IDs themselves.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setequal</span>(<span class="fu">colnames</span>(myScales), myCats<span class="sc">$</span>scale_id)  <span class="co"># must be TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">any</span>(<span class="fu">duplicated</span>(myCats<span class="sc">$</span>scale_id))               <span class="co"># must be FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>Define a vector of amino acid labels explicitly. This is the standard ordering of amino acids in this code. Using this variable guarantees that we don’t accidentally mix up ordering. (For example the order in the AAindex is not the same as in aaOntology…)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>  (aaLabels <span class="ot">&lt;-</span> <span class="fu">rownames</span>(myScales))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A" "C" "D" "E" "F" "G" "H" "I" "K" "L" "M" "N" "P" "Q" "R" "S" "T" "V" "W"
[20] "Y"</code></pre>
</div>
</div>
<section id="scaling-the-scales" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="scaling-the-scales"><span class="header-section-number">2.1</span> Scaling the scales</h2>
<p>The raw data in the observations has widely different scales, which depend on the experimenters’ choice of experimental datails. However, the dimensionality reduction method that we will use later (PCA) is sensitive to the variance of the data: it would give observations that use a numerically larger scale more weight than those that use a samller scale. We will compensate for that by re-scaling each data set to a mean of 0 and a standard deviation of 1, with the R commans <code>scale()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> myScales   <span class="co"># make a temporary copy for validation</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">ncol</span>(myScales))) {</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    myScales[ , i] <span class="ot">&lt;-</span> <span class="fu">scale</span>(myScales[ , i])</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Validate</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="dv">63</span>                           <span class="co"># Pick a random dataset</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(myScales[ , idx])              <span class="co"># Must be (nearly) zero</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -4.440892e-17</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sd</span>(myScales[ , idx])                <span class="co"># Must be (nearly) one</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Clean up</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="a-table-of-amino-acid-scale-types-and-values" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> A table of Amino Acid scale types and values</h1>
<p>We build a skeleton data structure that can contain both the categories and subcategories in <code>myCats</code>, as well as the actual values for the 20 amino acids from <code>myScales</code>. There are ways to merge such tables with a single expression in R, based on the shared scale IDs. However, here we do this explicitly, rather than inheriting the structure of the incoming data. We <strong>build from intent!</strong> This ensures we control the column names and their order, we know exactly what item goes where, and we can be sure that all assumptions we have about the resulting data are met.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>aaOntology <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaleID          =</span> myCats<span class="sc">$</span>scale_id,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">category         =</span> myCats<span class="sc">$</span>category,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subcategory      =</span> myCats<span class="sc">$</span>subcategory,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaleName        =</span> myCats<span class="sc">$</span>scale_name,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaleDescription =</span> myCats<span class="sc">$</span>scale_description)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(aaOntology) <span class="ot">&lt;-</span> aaOntology<span class="sc">$</span>scaleID <span class="co"># this gives us a convenient</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>                                             <span class="co"># way to access each row by</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>                                             <span class="co"># its ID. </span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add amino acid columns, initialized to NA</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (aa <span class="cf">in</span> aaLabels) {</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    aaOntology[ , aa] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we fill in the values scale by scale. We select them by the scaleID, not by row position. This makes the code robust to reordering. Use this as a general principle: match by semantics (meaning, content), not by structure (row index, relative position).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This takes the rows for each ID, and puts the values in the right </span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># columns for the ID in the joint table</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (sID <span class="cf">in</span> aaOntology<span class="sc">$</span>scaleID) {</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    aaOntology[sID, aaLabels] <span class="ot">&lt;-</span> myScales[ aaLabels , sID]</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note: it is now easy to pass the twenty or so lines of code to an AI, as part of a specification, and give it an exact description of the structure of our data.</p>
<p>Final sanity check</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>  <span class="co">#    Pick one random scale ID ...</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  (testID <span class="ot">&lt;-</span> <span class="fu">sample</span>(aaOntology<span class="sc">$</span>scaleID, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "QIAN880123"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print the reconstructed row (AA values only)</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Reconstructed values for scale"</span>, testID, <span class="st">":</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Reconstructed values for scale QIAN880123 :</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>  aaOntology[testID, aaLabels]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   A         C          D          E           F         G
QIAN880123 -1.304055 0.9191849 -0.3664457 -0.6776174 -0.09212321 0.7267498
                  H         I          K           L          M         N
QIAN880123 0.763599 0.2558978 -0.8741469 -0.05527393 -0.4073893 0.6080132
                   P         Q           R       S        T         V
QIAN880123 -1.459641 -1.848606 -0.09212321 1.46783 2.245759 0.6489568
                    W           Y
QIAN880123 -0.4442386 -0.01433028</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print the original column from myScales</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Original values from myScales:</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Original values from myScales:</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>  myScales[[testID]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             [,1]
 [1,] -1.30405520
 [2,]  0.91918490
 [3,] -0.36644565
 [4,] -0.67761738
 [5,] -0.09212321
 [6,]  0.72674976
 [7,]  0.76359904
 [8,]  0.25589780
 [9,] -0.87414689
[10,] -0.05527393
[11,] -0.40738930
[12,]  0.60801318
[13,] -1.45964106
[14,] -1.84860572
[15,] -0.09212321
[16,]  1.46782979
[17,]  2.24575911
[18,]  0.64895682
[19,] -0.44423858
[20,] -0.01433028
attr(,"scaled:center")
[1] 0.4515
attr(,"scaled:scale")
[1] 0.2442381</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Confirm equality (the identical() functionis VERY strict,</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># TRUE means success)</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">identical</span>(</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.numeric</span>(aaOntology[testID, aaLabels]),</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    myScales[[testID]]</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<section id="selecting-categories-of-amino-acid-scales" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="selecting-categories-of-amino-acid-scales"><span class="header-section-number">3.1</span> Selecting categories of amino acid scales</h2>
<p>To select only a subset of subcategorites, we prepare a data frame of subcategory IDs and 0 / 1 selectors. 0 will mean: don’t use that subcategory. We want to do this explicitly, so we have a record of our choices. Therefore, we ask the code to write code for us, which we can then edit interactively and select by hand. This is a reproducible way to scaffold manual decisions: generate template code programmatically, then curate it by hand.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>  mySubCats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cat =</span> aaOntology<span class="sc">$</span>category,</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">subcat =</span> aaOntology<span class="sc">$</span>subcategory,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">use =</span> <span class="fu">numeric</span>(<span class="fu">nrow</span>(aaOntology)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remove duplicates: we only need each unique subcategory listed once. The <code>! duplicated()</code> idiom is very useful to keep only unique values. This keeps only the first occurrence of each - which is fine, for this purpose they are all equivalent.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>  mySubCats <span class="ot">&lt;-</span> mySubCats[<span class="sc">!</span> <span class="fu">duplicated</span>(mySubCats<span class="sc">$</span>subcat), ]</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  (<span class="fu">nrow</span>(mySubCats))  <span class="co"># The number of unique subcategories in the dataset </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 74</code></pre>
</div>
</div>
<p>The following code can be used to write code … the repeated lines of assignments were created this way. Once it was run, I copied the result, pasted it into this script (below) and used it to select categories for inclusion.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mySubCats)) {</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">"mySubCats$use[%d] &lt;- 0    # %s : %s</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                i,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                mySubCats<span class="sc">$</span>cat[i],</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>                mySubCats<span class="sc">$</span>subcat[i]))</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Why do we need to select? Why can’t we just use all categories?</p>
<p>We want to use the result to compute “amino acid similarity” in order to evaluate the structure of the genetic code. Therefore we need to exclude all scales that are in some way influenced by the genetic code structure itself, rather than the inherent properties of the amino acid itself. If we are not careful about this point, we easily fall into a trap of <em>circular reasoning</em>.</p>
<p>In particular, things like codon counts, exchange probabilities, propensities to be observed in specific structural context in existin proteins, or categories that are not well defined or unclassified should be excluded.</p>
<p>The following section is deliberately long and explicit: it’s our record of assumptions. This ensures future readers (including ourselves) know exactly what we excluded, and how we interpreted our rationale above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># ASA/Volume : Accessible surface area (ASA)</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># ASA/Volume : Buried</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># ASA/Volume : Hydrophobic ASA</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># ASA/Volume : Partial specific volume</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># ASA/Volume : Volume</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">6</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : AA composition</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">7</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : AA composition (surface)</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">8</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : Membrane proteins (MPs)</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[ <span class="dv">9</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : Mitochondrial proteins</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">10</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : MPs (anchor)</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">11</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Composition : Unclassified (Composition)</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">12</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Coil</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">13</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Coil (C-term)</span></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">14</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Coil (N-term)</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">15</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Linker (&gt;14 AA)</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">16</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Linker (6-14 AA)</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">17</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : Unclassified (Conformation)</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">18</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix</span></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">19</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (C-cap)</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">20</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (C-term)</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">21</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (C-term, out)</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">22</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (left-handed)</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">23</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (N-cap)</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">24</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (N-term)</span></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">25</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (N-term, out)</span></span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">26</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : α-helix (α-proteins)</span></span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">27</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β/α-bridge</span></span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">28</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-sheet</span></span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">29</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-sheet (C-term)</span></span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">30</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-sheet (N-term)</span></span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">31</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-strand</span></span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">32</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-turn</span></span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">33</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-turn (C-term)</span></span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">34</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-turn (N-term)</span></span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">35</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : β-turn (TM helix)</span></span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">36</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Conformation : π-helix</span></span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">37</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Charge</span></span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">38</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Charge (negative)</span></span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">39</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Charge (positive)</span></span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">40</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Electron-ion interaction pot.</span></span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">41</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Entropy</span></span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">42</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Free energy (folding)</span></span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">43</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Free energy (unfolding)</span></span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">44</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Isoelectric point</span></span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">45</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Non-bonded energy</span></span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">46</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Energy : Unclassified (Energy)</span></span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">47</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : Mutability</span></span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">48</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : PC 1</span></span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">49</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : PC 2</span></span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">50</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : PC 3</span></span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">51</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : PC 4</span></span>
<span id="cb34-52"><a href="#cb34-52" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">52</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : PC 5</span></span>
<span id="cb34-53"><a href="#cb34-53" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">53</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Others : Unclassified (Others)</span></span>
<span id="cb34-54"><a href="#cb34-54" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">54</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Amphiphilicity</span></span>
<span id="cb34-55"><a href="#cb34-55" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">55</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Amphiphilicity (α-helix)</span></span>
<span id="cb34-56"><a href="#cb34-56" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">56</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Hydrophilicity</span></span>
<span id="cb34-57"><a href="#cb34-57" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">57</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Hydrophobicity</span></span>
<span id="cb34-58"><a href="#cb34-58" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">58</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Hydrophobicity (interface)</span></span>
<span id="cb34-59"><a href="#cb34-59" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">59</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Hydrophobicity (surrounding)</span></span>
<span id="cb34-60"><a href="#cb34-60" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">60</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Polarity : Unclassified (Polarity)</span></span>
<span id="cb34-61"><a href="#cb34-61" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">61</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Shape : Graph (1. eigenvalue)</span></span>
<span id="cb34-62"><a href="#cb34-62" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">62</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Shape : Graph (2. eigenvalue)</span></span>
<span id="cb34-63"><a href="#cb34-63" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">63</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Shape : Reduced distance</span></span>
<span id="cb34-64"><a href="#cb34-64" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">64</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Shape : Shape and Surface</span></span>
<span id="cb34-65"><a href="#cb34-65" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">65</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Shape : Side chain length</span></span>
<span id="cb34-66"><a href="#cb34-66" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">66</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Shape : Steric parameter</span></span>
<span id="cb34-67"><a href="#cb34-67" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">67</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Shape : Unclassified (Shape)</span></span>
<span id="cb34-68"><a href="#cb34-68" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">68</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Backbone-dynamics (-CH)</span></span>
<span id="cb34-69"><a href="#cb34-69" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">69</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Backbone-dynamics (-NH)</span></span>
<span id="cb34-70"><a href="#cb34-70" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">70</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Flexibility</span></span>
<span id="cb34-71"><a href="#cb34-71" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">71</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Flexibility (2 rigid neighbors)</span></span>
<span id="cb34-72"><a href="#cb34-72" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">72</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Stability</span></span>
<span id="cb34-73"><a href="#cb34-73" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">73</span>] <span class="ot">&lt;-</span> <span class="dv">1</span>    <span class="co"># Structure-Activity : Stability (helix-coil)</span></span>
<span id="cb34-74"><a href="#cb34-74" aria-hidden="true" tabindex="-1"></a>  mySubCats<span class="sc">$</span>use[<span class="dv">74</span>] <span class="ot">&lt;-</span> <span class="dv">0</span>    <span class="co"># Structure-Activity : Unclassified (Structure-Activity)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we define a selection vector that selects only the subcategories we want in the aaOntology table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>  sel <span class="ot">&lt;-</span> <span class="fu">logical</span>(<span class="fu">nrow</span>(aaOntology))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(aaOntology))) {</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    idx <span class="ot">&lt;-</span> <span class="fu">which</span>(mySubCats<span class="sc">$</span>subcat <span class="sc">==</span> aaOntology<span class="sc">$</span>subcategory[i])  <span class="co"># which()</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    sel[i] <span class="ot">&lt;-</span> (mySubCats<span class="sc">$</span>use[idx] <span class="sc">==</span> <span class="dv">1</span>)  <span class="co"># This is a bit of idiomatic R style:</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                                         <span class="co"># wrapping the expression in</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                                         <span class="co"># parentheses returns the result,</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                                         <span class="co"># which can then be assigned. Here, the</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                                         <span class="co"># result is either TRUE or FALSE</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The above selection could be written both more “pedestrian”, by running a loop that checks every single row, or more idiomatic, with this single expression:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>  sel <span class="ot">&lt;-</span> aaOntology<span class="sc">$</span>subcategory <span class="sc">%in%</span> mySubCats<span class="sc">$</span>subcat[mySubCats<span class="sc">$</span>use <span class="sc">==</span> <span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>… which coding style is preferred? Prefer the version that is most explicit for YOU.</p>
</section>
</section>
<section id="principal-component-analysis-pca" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Principal Component Analysis (PCA)</h1>
<p>PCA extracts the uncorrelated dimensions (principal components, PCs) along which the categories of observations in a large data set vary most strongly.</p>
<p>In our context:</p>
<ul>
<li>Observations = the 20 amino acids - as categories</li>
<li>Variables = the selected scales (hundreds of them)</li>
</ul>
<p>PCA gives us a reduced-dimensional “feature space” for amino acids:</p>
<ul>
<li>Each <em>axis</em> (principal component, PC) is orthogonal and ordered by how much variance it explains.</li>
<li>Each <em>amino acid</em> is projected into this space as a point with new coordinates (its “scores”).</li>
<li>Each <em>original scale</em> contributes to these PCs with a weight (the “loadings”), which we can use for interpretation.</li>
</ul>
<p>The coordinates of the amino acids allow us to compute distances between them - this becomes a computed “similarity”.</p>
<p>Comparing the axes with the original data-sets allows us to interpret them as e.g.&nbsp;hydrophobicity, polarity, size, etc.</p>
<section id="computing-the-pca" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="computing-the-pca"><span class="header-section-number">4.1</span> Computing the PCA</h2>
<p>Computing PCAs in R can be done with a single command. We apply it to the selected rows. Remember that the data columns of each amino acid are named according to the aaLabels…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>   aaPCA <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(aaOntology[sel, aaLabels]), <span class="at">scale. =</span> <span class="cn">FALSE</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>   <span class="co"># confirm the correct dimensions of aaPCA$x: should be 20 x 20</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>   <span class="fu">dim</span>(aaPCA<span class="sc">$</span>x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20 20</code></pre>
</div>
</div>
<p>Note: Normally we would scale each variable to unit variance, (PCA is sensitive to the magnitude of its input data), but here we already <code>scale()</code>’d the raw inputs, so we don’t scale them again. The parameter <code>scale. = TRUE</code> of <code>prcomp()</code> is set to <code>TRUE</code> by default, once again: be wary of hidden behaviour. For example, in the case that some of your data is meaningfully more “important” you might actually want its variance to be larger and not to scale implicitly.</p>
</section>
<section id="selecting-important-components" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="selecting-important-components"><span class="header-section-number">4.2</span> Selecting “important” components</h2>
<p>We get as many dimensions as the number of categories in the contributing data-sets, but the dimensions are (a) orthogonal, i.e.&nbsp;un-correlated, and (b) ordered by importance, i.e.&nbsp;by the amount of “variance” in the original data that each of them explains.</p>
<p>We can now keep only the most important ones, and discard all others. “Most important” could mean, those that together explain 95% of the variance in the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>  myColorGradient <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#02b597"</span>, <span class="st">"#549097"</span>, <span class="st">"#edf5f5"</span>))(<span class="dv">20</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  cumulativeProportions <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(aaPCA<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> <span class="fu">sum</span>(aaPCA<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">barplot</span>(cumulativeProportions,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">main=</span><span class="st">"Cumulative Proportion of Variance Explained"</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">xlab=</span><span class="st">"Principal Component"</span>,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylab=</span><span class="st">"Cumulative Proportion"</span>,</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">border=</span><span class="st">"#AAAAAA"</span>,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">col=</span>myColorGradient)</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  myCutoff <span class="ot">&lt;-</span> <span class="fl">0.95</span> <span class="co"># we define a cutoff for "most important"</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">abline</span>(<span class="at">h=</span>myCutoff, <span class="at">col=</span><span class="st">"#AA0000"</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)  <span class="co"># mark importance threshold</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="computingAminoAcidSimilarity_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># How many PCs to retain? (first PC where we pass 95% of variance)</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  numPCsToRetain <span class="ot">&lt;-</span> <span class="fu">which</span>(cumulativeProportions <span class="sc">&gt;=</span> myCutoff)[<span class="dv">1</span>]</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">"Retaining %i principal components (%3.0f%% cutoff ).</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>              numPCsToRetain,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>              myCutoff <span class="sc">*</span> <span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Retaining 12 principal components ( 95% cutoff ).</code></pre>
</div>
</div>
</section>
<section id="constructing-a-feature-space" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="constructing-a-feature-space"><span class="header-section-number">4.3</span> Constructing a Feature Space</h2>
<p>Simply use only columns <code>1:numPCsToRetain</code> and assign that to the variable name we want:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>  aaFeatureSpace <span class="ot">&lt;-</span> aaPCA<span class="sc">$</span>x[ , <span class="dv">1</span><span class="sc">:</span>numPCsToRetain]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="relationships-between-amino-acids" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Relationships between amino acids</h1>
<p>A natural way to explore such relationships is a “heatmap”. Here we compute a heatmap from the distances (note: smaller distances mean greater similarity). Above we have defined a feature space. If we measure Euclidean distances of all pairs of positions of amino acids in that space, we get an overview of all similarities between amino acids. A heatmap is a commmon way to visualize that: bright spots mean amino acids are far apart, dark spots mean they’re close, and block structures identify groups. Notice how biochemically similar residues cluster together - that is a signal that our abstract space lines up with biology.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>  dmat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">dist</span>(aaFeatureSpace))</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  pinkPeaksPal <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#ff295f"</span>, <span class="st">"#f0297f"</span>,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"#99999b"</span>, <span class="st">"#cccccd"</span>), <span class="at">bias =</span> <span class="fl">1.4</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heatmap</span>(dmat,</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">symm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#888888"</span>, <span class="fu">pinkPeaksPal</span>(<span class="dv">99</span>)),</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">na.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">main =</span> <span class="st">"Amino Acid Similarity (PCA Feature Space)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="computingAminoAcidSimilarity_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Figure:</strong> Heatmap of amino acid distances in Feature Space. Pink colours are closer distance. The dendrograms show the rleationships through hierarchical clustering. This clustering controls the order of rows and columns, which results in the visible block-structure that makes inherent groupings visible.</figcaption>
</figure>
</div>
</div>
</div>
<section id="summarize-and-save" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="summarize-and-save"><span class="header-section-number">5.1</span> Summarize and Save</h2>
<p>In summary our “Feature Space” for amino acids:</p>
<ul>
<li>… takes all of the original data into account, each scale according to its contribution to the variance;</li>
<li>… recognizes that with 20 categories of observations (the 20 amino acids), there are only 19 independent dimensions of variance;</li>
<li>… defines those independent dimensions along which the contributions are orthogonal, i.e.&nbsp;uncorrelated;</li>
<li>… selects from those only the most “important” ones, i.e.&nbsp;those that jointly explain most of the variance;</li>
<li>… contains each amino acid as a unique “point”, its position given by its coordinates - which are the values along the dimensions of the feature space.</li>
</ul>
<p>Save the feature space for re-use using saveRDS() … (see below)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>  myFN <span class="ot">&lt;-</span> <span class="st">"dat/aaFeatureSpace.2025.Rds"</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span> <span class="fu">file.exists</span>(myFN)) {</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">saveRDS</span>(aaFeatureSpace, <span class="at">file =</span> myFN)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># If we wish to reload later:</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># aaFeatureSpace &lt;- readRDS("dat/aaFeatureSpace.2025.Rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note: The pair <code>saveRDS()</code> / <code>readRDS()</code> save and <em>recreate</em> single compressed R objects. You can assign the “value” of <code>readRDS()</code> to a new variable, if you want, or to the same variable name you used when creating the object.</p>
<p>In contrast the pair <code>save()</code> / <code>load()</code>, <em>restores</em> an R object to its original name. This might silently overwrite an object that currently exists in your workspace.</p>
</section>
</section>
<section id="milestone-amino-acid-feature-space" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> MILESTONE: Amino Acid Feature Space</h1>
<p>At this point:</p>
<ul>
<li>we have taken a large set of biophysical and statistical observations of amino acids;</li>
<li>we have scaled them to be numerically comparable among each other;</li>
<li>we have selected categories of observations that we believe are independent of the genetic code itself, and thus can be used to evaluate the genetic code;</li>
<li>we have used PCA to remove any sampling bias and correlations between the various sets of observations;</li>
<li>we have constructed a “feature space” constructed from the most important “Principal Components” of the dataset, which represents the whole information contained in the original observations.</li>
<li>we have defined the position of each amino acid in this space.</li>
</ul>
<p>Next, we need to consider what these numbers all mean.</p>
</section>
<section id="interpreting-the-principal-components" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Interpreting the Principal Components</h1>
<p>Each of the original scales contributed a little to every dimension of our feature space - but this does not mean that the feature space is now a purely abstract mathematical artefact. We can interpret the dimensions by comparing them back to the original scales.</p>
<section id="corelations-between-pcs-and-observed-scales" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="corelations-between-pcs-and-observed-scales"><span class="header-section-number">7.1</span> Corelations between PCs and observed scales</h2>
<p>For example, let’s take the third-most important PC, the third column in our <code>aaFeatureSpace</code> matrix, and calculate a coefficient of correlation with each of the original scales. I write this as a completely “pedestrian” for-loop. Why?</p>
<p>In R, the <code>cor()</code> function is “vectorised” and can take whole matrices at once. That can look elegant, but it relies on a set of implicit assumptions: - Are rows or columns the “observations”? - What happens if one argument is a vector, and the other is a matrix? - Will <code>cor()</code> silently drop dimensions or recycle values?</p>
<p>For an expert user, those details can be managed. But for learners, <strong>“implicit” almost always means “unexpected”</strong>. A one-liner with <code>cor()</code> might be concise, but it is also brittle: change the orientation of a table, and suddenly you get “incompatible dimensions” errors - or worse: plausible numbers that are wrong, but hard to spot.</p>
<p>By contrast, the explicit loop below states exactly what we intend: “For each scale (row), compute its correlation with this PC.”</p>
<p>This may be more pedestrian, but it is:</p>
<ul>
<li>Readable: the intent is obvious.</li>
<li>Robust: no hidden conventions about rows vs.&nbsp;columns.</li>
<li>Fast enough: looping over a few hundred scales × 20 amino acids is trivial for R.</li>
</ul>
<p>For teaching, the clarity of intent matters more than saving one line of code. In fact, this is a good example of the broader principle: in data analysis, <strong>explicit beats implicit</strong>, especially when code is read by humans (or AIs) who need to understand the logic, not just the result.</p>
<p><strong>Defensive coding: explicitly align amino acid types:</strong></p>
<p>We want the scores of amino acids along some Principal Component. But: row order can sometimes change (e.g.&nbsp;one-letter and three-letter codes are sorted differently) and all of a sudden, without noticing, you are comparing values for Ala - A (Apples) with Arg - R (oRanges), which is the kind of silent error that is as awful in its consequences as it is hard to spot.</p>
<p>By indexing with the order defined in <code>aaLabels</code>, we guarantee the rows to be in the same order as our amino acid labels. (Mind you, they are already sorted that way, because this is how we constructed them. But that is something we only <em>know</em>, not something we can <em>see</em> at this point in the code.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>  thisPC <span class="ot">&lt;-</span> aaFeatureSpace[aaLabels, <span class="dv">3</span>] <span class="co"># define which PC to use</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  myCors <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="fu">nrow</span>(aaOntology))   <span class="co"># a vector to store the results</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(myCors) <span class="ot">&lt;-</span> aaOntology<span class="sc">$</span>scaleID   <span class="co"># named, to keep IDs with results</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(myCors)) {        <span class="co"># iterate over each row</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cor() the PC with the aaLabel-ordered values, coerced into a vector</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>    myCors[i] <span class="ot">&lt;-</span> <span class="fu">cor</span>(thisPC, <span class="fu">as.numeric</span>(aaOntology[i, aaLabels]))</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print out the 10 most positive and most negative correlations</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>  (topPos <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">sort</span>(myCors, <span class="at">decreasing =</span> <span class="cn">TRUE</span> ), <span class="dv">10</span>))   <span class="co"># strongest positive</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>RACS820112 BUNA790101 ROBB760104 BLAM930101 ONEK900101 FAUJ880108 FASG760103 
 0.7492448  0.7463910  0.7275348  0.6898095  0.6181188  0.6004750  0.5957423 
PALJ810108 FAUJ880113 QIAN880108 
 0.5816101  0.5795863  0.5654517 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>  (topNeg <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">sort</span>(myCors, <span class="at">decreasing =</span> <span class="cn">FALSE</span>), <span class="dv">10</span>))   <span class="co"># strongest negative</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FASG760104 FINA910102 ISOY800104 MUNV940104 ONEK900102 MUNV940105 GEOR030109 
-0.7325075 -0.7210725 -0.7156222 -0.6972615 -0.6711438 -0.6707951 -0.6629569 
LAWE840101 AURR980119 TANS770104 
-0.6521145 -0.6441270 -0.6266386 </code></pre>
</div>
</div>
<p>… and of course we can retrieve the actual scale-definitions, like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(topPos[<span class="dv">1</span>], aaOntology[<span class="fu">names</span>(topPos[<span class="dv">1</span>]), <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RACS820112
[1] 0.7492448

$category
[1] "Conformation"

$subcategory
[1] "Unclassified (Conformation)"

$scaleName
[1] "Extended (right-handed)"

$scaleDescription
[1] "Average relative fractional occurrence in ER(i-1) (Rackovsky-Scheraga, 1982)"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(topNeg[<span class="dv">1</span>], aaOntology[<span class="fu">names</span>(topPos[<span class="dv">1</span>]), <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$FASG760104
[1] -0.7325075

$category
[1] "Conformation"

$subcategory
[1] "Unclassified (Conformation)"

$scaleName
[1] "Extended (right-handed)"

$scaleDescription
[1] "Average relative fractional occurrence in ER(i-1) (Rackovsky-Scheraga, 1982)"</code></pre>
</div>
</div>
</section>
<section id="examine-the-dimensions" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="examine-the-dimensions"><span class="header-section-number">7.2</span> Examine the dimensions</h2>
<p>We can also plot two dimensions of the feature space as a scatterplot, with one-letter codes. This is a great way to get an intuitive sense of the similarity. These are just the two first principal components, but jointly they already explain about 70% of the variance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>  myDim1 <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  myDim2 <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(aaFeatureSpace[ , myDim1], aaFeatureSpace[ , myDim2],</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">asp =</span> <span class="fl">1.0</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">type =</span> <span class="st">"n"</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="st">"PC scatterplot"</span>,</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="fu">sprintf</span>(<span class="st">"principal component %i"</span>, myDim1),</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">ylab =</span> <span class="fu">sprintf</span>(<span class="st">"principal component %i"</span>, myDim2))</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">text</span>(aaFeatureSpace[, myDim1], aaFeatureSpace[,myDim2],</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>       <span class="fu">rownames</span>(aaFeatureSpace),</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> AACOLS[<span class="fu">rownames</span>(aaFeatureSpace)],</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">cex =</span> <span class="fl">2.0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="computingAminoAcidSimilarity_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>But looking at this plot, we have to ask ourselves - what do these <em>principal components</em> actually mean? Is this just a mathematical trick, or does this connect to the “real world”? To examine that, we need to take a more global view of the original scales, and their relationship to the PCs we use.</p>
</section>
<section id="global-interpretation-of-axes-and-scales" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="global-interpretation-of-axes-and-scales"><span class="header-section-number">7.3</span> Global Interpretation of Axes and Scales</h2>
<section id="interpreting-in-2d" class="level3" data-number="7.3.1">
<h3 data-number="7.3.1" class="anchored" data-anchor-id="interpreting-in-2d"><span class="header-section-number">7.3.1</span> Interpreting in 2D</h3>
<p>The original 586 scales for 20 amino acids occupy a 20-dimensional subspace of a 586-dimensional space. But even after using PCA, the resulting space is too high-dimensional to properly visualize. This is a common problem for data science and one of the most popular solutions was contributed in 2008 by UofT’s Geoffrey Hinton, with Laurens van der Maaten (<a href="https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf"><strong>PDF</strong></a>): this is “t-Stochastic Neighbour Embedding”, or t-SNE. This is an excellent dimensionality reduction algorithm that works by embedding in a low dimensional space while preserving neighborhood relations. Essentially, this is a projection that is not based on algebra, but on semantics, the meaning and significance of the data. The only drawback was its <em>O</em>(<em>N</em><sup>2</sup>) time complexity, that made computational requirements very significant for large datasets. However, van der Maaten published an improvement in 2014 (<a href="https://www.jmlr.org/papers/volume15/vandermaaten14a/vandermaaten14a.pdf"><strong>PDF</strong></a>) which used a tree-based algorithm to achieve an <em>O</em>( <em>N</em> log <em>N</em> ) complexity, which is practical even for “large” datasets. This is the version that is implemented in R.</p>
<p>Below, I am showcasing how to prepare our data and run it through tSNE. It’s actually quite simple. We prepare a joint matrix of scales and feature Space PCs, run that through t-SNE, and plot the results. The difficulties are only in the details of data preparation and plotting:</p>
<ul>
<li>We need to ensure none of our points overlap exactly;</li>
<li>We need to keep the names and definitions attached to the data, so we can interpret the plot;</li>
<li>We need to color code the points to identify and distinguish scales and PCs;</li>
<li>We would like some interactivity - the ability to zoom into the plot, or get detailed information;</li>
<li>…</li>
</ul>
<p>… in short, all those things that allow us to tell a story with our data. And that is what data science is really about.</p>
</section>
<section id="joint-t-sne-embedding-of-scales-pcs-2d" class="level3" data-number="7.3.2">
<h3 data-number="7.3.2" class="anchored" data-anchor-id="joint-t-sne-embedding-of-scales-pcs-2d"><span class="header-section-number">7.3.2</span> Joint t-SNE embedding of scales + PCs (2D)</h3>
<section id="first-step-organising-our-data" class="level4" data-number="7.3.2.1">
<h4 data-number="7.3.2.1" class="anchored" data-anchor-id="first-step-organising-our-data"><span class="header-section-number">7.3.2.1</span> First step: Organising our Data</h4>
<p>We define a lookup table for clear naming of PCs. PCs don’t come with globally unique IDs, some of the AA ontology scales are also called “PC1”, “PC2” etc. so we <em>namespace</em> them: we add a prefix that makes sure we won’t have any collision between names. The lookup table allows us to switch between our “private” names, and the names we used in the <code>aaFeatureSpace</code> table, as needed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>pcLookup <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">myID   =</span> <span class="fu">paste0</span>(<span class="st">"my."</span>, <span class="fu">colnames</span>(aaFeatureSpace)),  <span class="co"># e.g. "my.PC1"</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pcID   =</span> <span class="fu">colnames</span>(aaFeatureSpace)                  <span class="co"># e.g. "PC1"</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Scales already have unique and meaningful IDs, so we leave them as-is.</p>
<p>However, some scales have duplicated values. That will cause a problem later, because the t-SNE embedding method will not work if points with a distance of zero are present in the dataset. Let’s quickly build a vector of scale-IDs whose values (i.e positions in the 20- dimensional space) are unique: <code>myScID</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># identify duplicated rownames</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>isDup <span class="ot">&lt;-</span> <span class="fu">duplicated</span>(aaOntology[, aaLabels])   <span class="co"># a vector of TRUE and FALSE</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which scales these are:</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(aaOntology)[isDup]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PRAM900102" "RICJ880102" "PRAM900103" "KOEH090105" "KOEH090107"
[6] "KOEH090108"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define IDs of unique scales</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>myScID <span class="ot">&lt;-</span> <span class="fu">rownames</span>(aaOntology)[<span class="sc">!</span> isDup]</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ... as well, for symmetry of reference:</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>myPcID <span class="ot">&lt;-</span> pcLookup<span class="sc">$</span>myID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These two vectors define our scales and PCs, each as a vector of rownames in the respective matrices that store the originals. Referencing them thrrough names rather than row numbers guarantees we will always use the correct rows (or columns) in exactly the right order. Errors about row numbers are notoriously hard to identify and may lead to subtly wrong results. Using names as labels of the contents is the safer way.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the joint matrix of scales and PCs</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>jointMat <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  aaOntology[ myScID, aaLabels],              <span class="co"># rows: scales, cols: amino acids</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>(aaFeatureSpace[aaLabels, pcLookup<span class="sc">$</span>pcID])  <span class="co"># rows: PCs, cols: amino acids</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Row names: preserve scale IDs, add namespaced PC IDs</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(jointMat) <span class="ot">&lt;-</span> <span class="fu">c</span>(myScID, myPcID)</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Define hover text for scales: include ID, name, and description</span></span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>scaleHover <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Scale: "</span>, aaOntology[myScID, <span class="st">"scaleID"</span>], <span class="st">"&lt;br&gt;"</span>,</span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Name: "</span>, aaOntology[myScID, <span class="st">"scaleName"</span>], <span class="st">"&lt;br&gt;"</span>,</span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Description: "</span>, aaOntology[myScID, <span class="st">"scaleDescription"</span>]</span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-17"><a href="#cb58-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Add hover text for PCs: minimal but interpretable</span></span>
<span id="cb58-18"><a href="#cb58-18" aria-hidden="true" tabindex="-1"></a>pcHover <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb58-19"><a href="#cb58-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Feature-space axis "</span>, pcLookup<span class="sc">$</span>pcID,</span>
<span id="cb58-20"><a href="#cb58-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">" ("</span>, pcLookup<span class="sc">$</span>myID, <span class="st">")"</span></span>
<span id="cb58-21"><a href="#cb58-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-22"><a href="#cb58-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-23"><a href="#cb58-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine into a single vector aligned with the jointMat rows</span></span>
<span id="cb58-24"><a href="#cb58-24" aria-hidden="true" tabindex="-1"></a>hoverText <span class="ot">&lt;-</span> <span class="fu">c</span>(scaleHover, pcHover)</span>
<span id="cb58-25"><a href="#cb58-25" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(hoverText) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(jointMat)</span>
<span id="cb58-26"><a href="#cb58-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-27"><a href="#cb58-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Final sanity checks </span></span>
<span id="cb58-28"><a href="#cb58-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-29"><a href="#cb58-29" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">ncol</span>(jointMat) <span class="sc">==</span> <span class="fu">length</span>(aaLabels))     <span class="co"># 20 amino acids</span></span>
<span id="cb58-30"><a href="#cb58-30" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">nrow</span>(jointMat) <span class="sc">==</span> <span class="fu">length</span>(myScID) <span class="sc">+</span></span>
<span id="cb58-31"><a href="#cb58-31" aria-hidden="true" tabindex="-1"></a>                            <span class="fu">nrow</span>(pcLookup))       <span class="co"># unique scales + PCs</span></span>
<span id="cb58-32"><a href="#cb58-32" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(aaLabels <span class="sc">%in%</span> <span class="fu">colnames</span>(jointMat)))  <span class="co"># column names intact</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="computing-the-t-sne-embeddings" class="level4" data-number="7.3.2.2">
<h4 data-number="7.3.2.2" class="anchored" data-anchor-id="computing-the-t-sne-embeddings"><span class="header-section-number">7.3.2.2</span> Computing the t-SNE embeddings</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the Rtsne package if needed ...</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span> <span class="fu">requireNamespace</span>(<span class="st">"Rtsne"</span>, <span class="at">quietly=</span><span class="cn">TRUE</span>)) {</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"Rtsne"</span>, <span class="at">repos =</span> <span class="st">"http://cran.us.r-project.org"</span>)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into '/opt/homebrew/lib/R/4.5/site-library'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ready to run t-SNE ...</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13</span>)  <span class="co"># reproducible layout</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>tsneRes <span class="ot">&lt;-</span> Rtsne<span class="sc">::</span><span class="fu">Rtsne</span>(<span class="fu">as.matrix</span>(jointMat),</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dims =</span> <span class="dv">2</span>,</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">perplexity =</span> <span class="dv">30</span>,</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">max_iter =</span> <span class="dv">3000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-the-data-for-plotting" class="level4" data-number="7.3.2.3">
<h4 data-number="7.3.2.3" class="anchored" data-anchor-id="prepare-the-data-for-plotting"><span class="header-section-number">7.3.2.3</span> Prepare the data for plotting</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the plotly package if needed ...</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span> <span class="fu">requireNamespace</span>(<span class="st">"plotly"</span>, <span class="at">quietly=</span><span class="cn">TRUE</span>)) {</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"plotly"</span>, <span class="at">repos =</span> <span class="st">"http://cran.us.r-project.org"</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into '/opt/homebrew/lib/R/4.5/site-library'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>also installing the dependencies 'gtable', 'isoband', 'S7', 'farver', 'labeling', 'tidyselect', 'generics', 'later', 'ggplot2', 'scales', 'viridisLite', 'htmlwidgets', 'tidyr', 'RColorBrewer', 'dplyr', 'lazyeval', 'crosstalk', 'purrr', 'data.table', 'promises'</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define category colors (explicit, ordered)</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>catColors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Energy               =</span> <span class="st">"#f2003c"</span>,</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Shape                =</span> <span class="st">"#F0A200"</span>,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">ASA/Volume</span><span class="st">`</span>         <span class="ot">=</span> <span class="st">"#f0ea00"</span>,</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">Conformation         =</span> <span class="st">"#62C923"</span>,</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Composition          =</span> <span class="st">"#0A9A9B"</span>,</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">Polarity             =</span> <span class="st">"#1958C3"</span>,</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">Structure-Activity</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"#8000D3"</span>,</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">Others               =</span> <span class="st">"#999999"</span></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-14"><a href="#cb65-14" aria-hidden="true" tabindex="-1"></a><span class="co">#  Extract the embedded coordinates from the t-SNE result object</span></span>
<span id="cb65-15"><a href="#cb65-15" aria-hidden="true" tabindex="-1"></a>tsneCoords <span class="ot">&lt;-</span> tsneRes<span class="sc">$</span>Y</span>
<span id="cb65-16"><a href="#cb65-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tsneCoords) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(jointMat)</span>
<span id="cb65-17"><a href="#cb65-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-18"><a href="#cb65-18" aria-hidden="true" tabindex="-1"></a>scaleCoords <span class="ot">&lt;-</span> tsneCoords[myScID, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb65-19"><a href="#cb65-19" aria-hidden="true" tabindex="-1"></a>pcCoords    <span class="ot">&lt;-</span> tsneCoords[myPcID, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb65-20"><a href="#cb65-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-21"><a href="#cb65-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Build data frames with coordinates and annotations</span></span>
<span id="cb65-22"><a href="#cb65-22" aria-hidden="true" tabindex="-1"></a>scaleDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb65-23"><a href="#cb65-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">x        =</span> tsneCoords[myScID, <span class="dv">1</span>],</span>
<span id="cb65-24"><a href="#cb65-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">y        =</span> tsneCoords[myScID, <span class="dv">2</span>],</span>
<span id="cb65-25"><a href="#cb65-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">category =</span> aaOntology[myScID, <span class="st">"category"</span>],   <span class="co"># bring in category explicitly</span></span>
<span id="cb65-26"><a href="#cb65-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">hover    =</span> hoverText[myScID]</span>
<span id="cb65-27"><a href="#cb65-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb65-28"><a href="#cb65-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-29"><a href="#cb65-29" aria-hidden="true" tabindex="-1"></a>pcDF <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb65-30"><a href="#cb65-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">x     =</span> tsneCoords[myPcID, <span class="dv">1</span>],</span>
<span id="cb65-31"><a href="#cb65-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">y     =</span> tsneCoords[myPcID, <span class="dv">2</span>],</span>
<span id="cb65-32"><a href="#cb65-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">myID  =</span> myPcID,                 <span class="co"># short ID for plot labels</span></span>
<span id="cb65-33"><a href="#cb65-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">hover =</span> hoverText[myPcID]       <span class="co"># full hover text</span></span>
<span id="cb65-34"><a href="#cb65-34" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-the-actual-interactive-plot" class="level4" data-number="7.3.2.4">
<h4 data-number="7.3.2.4" class="anchored" data-anchor-id="create-the-actual-interactive-plot"><span class="header-section-number">7.3.2.4</span> Create the actual interactive plot</h4>
<p>We went through a fair bit of assembling and bookkeeping to have everything prepared prior to creating the plots. We could have just pushed the respective data into the statements that produce the plots themselves, but uniting it all in one data structure makes it easier:</p>
<ul>
<li>to check for correctness;</li>
<li>to establish a shared understanding about our data and annotations with the pair-programming AI;</li>
<li>to <strong>make our implicit intent explicit</strong>.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize the figure as an empty container</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>fig <span class="ot">&lt;-</span> plotly<span class="sc">::</span><span class="fu">plot_ly</span>()</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the observed scales (colored by category) with the add_trace() function</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="co"># of the plotly:: package</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>fig <span class="ot">&lt;-</span> fig <span class="sc">|&gt;</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  plotly<span class="sc">::</span><span class="fu">add_trace</span>(</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> scaleDF,</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="sc">~</span>x, <span class="at">y =</span> <span class="sc">~</span>y,</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"scatter"</span>,</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">mode =</span> <span class="st">"markers"</span>,</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="sc">~</span><span class="fu">factor</span>(category, <span class="at">levels =</span> <span class="fu">names</span>(catColors)),</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> catColors,</span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">hoverinfo =</span> <span class="st">"text"</span>,</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">hovertext =</span> <span class="sc">~</span>hover,</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">marker =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">6</span>)</span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the PCs (red squares, single legend entry)</span></span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>fig <span class="ot">&lt;-</span> fig <span class="sc">|&gt;</span></span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>  plotly<span class="sc">::</span><span class="fu">add_trace</span>(</span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> pcDF,</span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="sc">~</span>x, <span class="at">y =</span> <span class="sc">~</span>y,</span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"scatter"</span>,</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">mode =</span> <span class="st">"markers+text"</span>,</span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">marker =</span> <span class="fu">list</span>(<span class="at">symbol =</span> <span class="st">"square"</span>, <span class="at">size =</span> <span class="dv">10</span>, <span class="at">color =</span> <span class="st">"#ff295f"</span>),</span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="sc">~</span>myID,</span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">textposition =</span> <span class="st">"top center"</span>,</span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">hoverinfo =</span> <span class="st">"text"</span>,</span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">hovertext =</span> <span class="sc">~</span>hover,</span>
<span id="cb66-31"><a href="#cb66-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Feature-space PCs"</span>,</span>
<span id="cb66-32"><a href="#cb66-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">showlegend =</span> <span class="cn">TRUE</span></span>
<span id="cb66-33"><a href="#cb66-33" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb66-34"><a href="#cb66-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-35"><a href="#cb66-35" aria-hidden="true" tabindex="-1"></a><span class="co"># We are done preparing the figure: show it</span></span>
<span id="cb66-36"><a href="#cb66-36" aria-hidden="true" tabindex="-1"></a>fig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="plotly html-widget html-fill-item" id="htmlwidget-9563bd8cc14e402ee614" style="width:100%;height:672px;"></div>
<script type="application/json" data-for="htmlwidget-9563bd8cc14e402ee614">{"x":{"visdat":{"106b52ec51541":["function () ","plotlyVisDat"],"106b5cf67664":["function () ","data"],"106b58d6a5e2":["function () ","data"]},"cur_data":"106b58d6a5e2","attrs":{"106b5cf67664":{"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"x":{},"y":{},"type":"scatter","mode":"markers","color":{},"colors":["#f2003c","#F0A200","#f0ea00","#62C923","#0A9A9B","#1958C3","#8000D3","#999999"],"hoverinfo":"text","hovertext":{},"marker":{"size":6},"inherit":true},"106b58d6a5e2":{"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"x":{},"y":{},"type":"scatter","mode":"markers+text","marker":{"symbol":"square","size":10,"color":"#ff295f"},"text":{},"textposition":"top center","hoverinfo":"text","hovertext":{},"name":"Feature-space PCs","showlegend":true,"inherit":true}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"x"},"yaxis":{"domain":[0,1],"automargin":true,"title":"y"},"hovermode":"closest","showlegend":true},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"x":[14.468178138971792,12.565929181505814,7.0672853937510824,7.8302240246420025,14.137401081217739,-5.4416952718092828,5.7377865395133512,5.737777768158673,12.638155361009275,13.062171646412285,11.131596962206068,-2.5685878229441998,-2.5318810880960965,-2.821264266002439,-0.99702904603096831,-1.013565750883122,-13.706034842364806,-14.143386879473546,-15.008238988415116,-15.000417855605905,-12.312706334067775,-14.697249819521357,-15.131600358266603,-13.998393344233216,11.009318825233425,14.270308123180559,10.339434048715436,-19.035543179402406,-17.42108053893206,-17.403904063814764,-17.510483584520866,-6.4866336561312616,-14.03768531798773,-10.336049985459036,5.2474399384218877,-11.017728015990798],"y":[-9.6781225175844661,9.6723000156283181,-10.528715960412841,-13.685998575279237,-10.39907754018528,-2.7535052230318273,0.53390020889478695,0.53371243996206164,8.4757446097757292,7.8566726556280697,6.946068373649922,-6.1984912284372058,-6.1200405723584117,-10.123442693177235,-7.9760163711738059,-8.0141344641314323,10.353969782138028,10.124879328981701,9.8732598725667611,9.8729619206570991,14.022576268585718,11.772403730123804,11.996955449778371,12.194226066523509,-18.453381882808191,-9.822924454170364,-18.02650977674287,-3.1563536322623573,-3.421522586595767,-3.3301482745456634,-3.0673853462503406,9.2628062090238021,14.823027220281013,-11.065330523792966,8.0361721303834237,8.8113381841911078],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: KLEP840101<br>Name: Charge<br>Description: Net charge (Klein et al., 1984)","Scale: CHAM830108<br>Name: Charge (donor)<br>Description: A parameter of charge transfer donor capability (Charton-Charton, 1983)","Scale: FAUJ880112<br>Name: Charge (negative)<br>Description: Negative charge (Fauchere et al., 1988)","Scale: CHAM830107<br>Name: Charge (transfer)<br>Description: A parameter of charge transfer capability (Charton-Charton, 1983)","Scale: FAUJ880111<br>Name: Charge (positive)<br>Description: Positive charge (Fauchere et al., 1988)","Scale: FAUJ880108<br>Name: Electrical Effect<br>Description: Localized Electrical Effect (Fauchere et al., 1988)","Scale: COSI940101<br>Name: Electron-ion interaction potential<br>Description: Electron-ion interaction potential values (Cosic, 1994)","Scale: VELV850101<br>Name: Electron-ion interaction potential<br>Description: Electron-ion interaction potential (Veljkovic et al., 1985)","Scale: HUTJ700102<br>Name: Entropy<br>Description: Absolute entropy (Hutchens, 1970)","Scale: HUTJ700103<br>Name: Entropy<br>Description: Entropy of formation (Hutchens, 1970)","Scale: OOBM770102<br>Name: Non-bonded energy per atom<br>Description: Short and medium range non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: MUNV940101<br>Name: Free energy (α-helix)<br>Description: Free energy in alpha-helical conformation (Munoz-Serrano, 1994)","Scale: MUNV940102<br>Name: Free energy (α-helix)<br>Description: Free energy in alpha-helical region (Munoz-Serrano, 1994)","Scale: MUNV940103<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand conformation (Munoz-Serrano, 1994)","Scale: MUNV940104<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand region (Munoz-Serrano, 1994)","Scale: MUNV940105<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand region (Munoz-Serrano, 1994)","Scale: YUTK870101<br>Name: Free energy (unfolding)<br>Description: Unfolding Gibbs energy in water, pH7.0 (Yutani et al., 1987)","Scale: YUTK870102<br>Name: Free energy (unfolding)<br>Description: Unfolding Gibbs energy in water, pH9.0 (Yutani et al., 1987)","Scale: YUTK870103<br>Name: Free energy (unfolding)<br>Description: Activation Gibbs energy of unfolding, pH7.0 (Yutani et al., 1987)","Scale: YUTK870104<br>Name: Free energy (unfolding)<br>Description: Activation Gibbs energy of unfolding, pH9.0 (Yutani et al., 1987)","Scale: JANJ790102<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy (Janin, 1979)","Scale: RADA880104<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy from chx to oct (Radzicka-Wolfenden, 1988)","Scale: RADA880105<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy from vap to oct (Radzicka-Wolfenden, 1988)","Scale: RADA880107<br>Name: Transfer free energy (TFE) to inside<br>Description: Energy transfer from out to in (95% buried) (Radzicka-Wolfenden, 1988)","Scale: EISD860102<br>Name: Atom-based hydrophobic moment<br>Description: Atom-based hydrophobic moment (Eisenberg-McLachlan, 1986)","Scale: ZIMJ680104<br>Name: Isoelectric point<br>Description: Isoelectric point (Zimmerman et al., 1968)","Scale: FAUJ880109<br>Name: Number hydrogen bond donors<br>Description: Number of hydrogen bond donors (Fauchere et al., 1988)","Scale: VASM830102<br>Name: Free energy (Extended)<br>Description: Relative population of conformational state C (Vasquez et al., 1983)","Scale: OOBM770104<br>Name: Non-bonded energy per residue<br>Description: Average non-bonded energy per residue (Oobatake-Ooi, 1977)","Scale: OOBM770105<br>Name: Non-bonded energy per residue<br>Description: Short and medium range non-bonded energy per residue (Oobatake-Ooi, 1977)","Scale: RADA880103<br>Name: TFE to inside<br>Description: Transfer free energy from vap to chx (Radzicka-Wolfenden, 1988)","Scale: WERD780103<br>Name: Free energy (α-helix)<br>Description: Free energy change of alpha(Ri) to alpha(Rh) (Wertz-Scheraga, 1978)","Scale: WERD780104<br>Name: Free energy (α-helix)<br>Description: Free energy change of epsilon(i) to alpha(Rh) (Wertz-Scheraga, 1978)","Scale: WERD780102<br>Name: Free energy change<br>Description: Free energy change of epsilon(i) to epsilon(ex) (Wertz-Scheraga, 1978)","Scale: HUTJ700101<br>Name: Heat capactiy<br>Description: Heat capacity (Hutchens, 1970)","Scale: OOBM850103<br>Name: Transfer free energy (TFE)<br>Description: Optimized transfer energy parameter (Oobatake et al., 1985)"],"marker":{"color":"rgba(242,0,60,1)","size":6,"line":{"color":"rgba(242,0,60,1)"}},"name":"Energy","textfont":{"color":"rgba(242,0,60,1)"},"error_y":{"color":"rgba(242,0,60,1)"},"error_x":{"color":"rgba(242,0,60,1)"},"line":{"color":"rgba(242,0,60,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[8.7033376570175598,7.3873469167500758,8.685448270024331,7.3566040741307699,8.3306366446884716,-18.810041332302127,-0.95639677823330205,-2.0246570641429371,-3.8345365381309771,2.2284586564061208,2.0680465461728446,3.226525325906763,2.9868011577929861,-14.529783877804263,-14.122944994932867,-14.518490782404887,9.2548918627844152,9.7638198684120301,11.538864046521189,9.9560634717011691,11.725961440183193,8.6218439867516175,11.217188279772058,8.7790227177071287,9.3381809036892047,9.0612581647645936,8.4096251958829704,8.329516692261377,13.894724197895959,11.652517324884247,10.915248212922846,10.594443156885053,10.88002021915637,12.246029331016475,10.190751577727932,1.6478193518433433,1.6572246532324055,17.421363935621891,17.531872464505089,17.578164401758283,1.7268935456606451,-4.065400992500515,-17.018983846952597,1.9229997782301769,-10.510933043651841],"y":[5.6114750228752879,5.9926631991600141,5.8182369817675852,5.9094507971654577,5.6442216552067057,-3.1645951076201162,-5.4255305231847935,-6.1421506415961016,-11.303035900540349,-20.009306452316437,-20.271179147761309,-19.980529916813513,-20.076235720378797,-4.7056405559413257,-5.7764203444416031,-4.8824861198611726,8.3422133143055834,8.1463639113189696,8.7646560608357547,8.6711508989500263,8.9678240697804004,7.7168362086681164,8.1712491799178437,7.9379759500351819,7.5590594444076329,8.5140946404873237,8.5443011624380389,8.3985416335359435,5.2575552699653025,10.04321186810734,9.4757756039294989,8.4958426289531896,8.7313957861656135,7.9269914629575142,7.4711832677019796,7.3193826334056942,9.2415192546950369,1.0828350777692644,0.72114835551243772,0.47767742138895486,7.7787936259058394,-1.9711321719115142,-3.980273511190934,-9.3811436245720152,-6.8144889859788389],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: KARS160111<br>Name: Eigenvalue (average)<br>Description: Average eigenvalue of the Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160121<br>Name: Eigenvalue (average)<br>Description: Weighted average eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: KARS160109<br>Name: Eigenvalue (maximum)<br>Description: Maximum eigenvalue of the weighted Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160119<br>Name: Eigenvalue (maximum)<br>Description: Weighted maximum eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: KARS160108<br>Name: Graph (weighted degree)<br>Description: Average weighted degree (total degree, divided by the number of vertices) (Kakraba-Knisley, 2016)","Scale: KARS160112<br>Name: Eigenvalue (2. smallest)<br>Description: Second smallest eigenvalue of the Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160122<br>Name: Eigenvalue (2. smallest)<br>Description: Weighted second smallest eigenvalue of the weighted Laplacian matrix (Kakraba-Knisley, 2016)","Scale: RACS820114<br>Name: Side chain angle (Theta)<br>Description: Value of theta(i-1) (Rackovsky-Scheraga, 1982)","Scale: PRAM820101<br>Name: Intercept in regression<br>Description: Intercept in regression analysis (Prabhakaran-Ponnuswamy, 1982)","Scale: MEIH800101<br>Name: Reduced distance (C-α)<br>Description: Average reduced distance for C-alpha (Meirovitch et al., 1980)","Scale: RACS770101<br>Name: Reduced distance (C-α)<br>Description: Average reduced distance for C-alpha (Rackovsky-Scheraga, 1977)","Scale: MEIH800102<br>Name: Reduced distance (side chain)<br>Description: Average reduced distance for side chain (Meirovitch et al., 1980)","Scale: RACS770102<br>Name: Reduced distance (side chain)<br>Description: Average reduced distance for side chain (Rackovsky-Scheraga, 1977)","Scale: PRAM820103<br>Name: Correlation coefficient in regression<br>Description: Correlation coefficient in regression analysis (Prabhakaran-Ponnuswamy, 1982)","Scale: LEVM760104<br>Name: Side chain angle (Phi)<br>Description: Side chain torsion angle phi(AAAR) (Levitt, 1976)","Scale: PRAM820102<br>Name: Slope in Regression<br>Description: Slope in Regression analysis x 1.0E1 (Prabhakaran-Ponnuswamy, 1982)","Scale: KARS160117<br>Name: Atomic number (total)<br>Description: Total weighted atomic number of the graph (obtained by summing all the atomic number of each of the vertices in the graph) (Kakraba-Knisley, 2016)","Scale: KARS160105<br>Name: Eccentricity (average)<br>Description: Average eccentricity (Kakraba-Knisley, 2016)","Scale: KARS160114<br>Name: Eccentricity (average)<br>Description: Average weighted eccentricity based on the atomic number (Kakraba-Knisley, 2016)","Scale: KARS160107<br>Name: Eccentricity (maximum)<br>Description: Diameter (maximum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160116<br>Name: Eccentricity (maximum)<br>Description: Weighted diameter based on the atomic number (maximum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160106<br>Name: Eccentricity (minimum)<br>Description: Radius (minimum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160115<br>Name: Eccentricity (minimum)<br>Description: Weighted radius based on the atomic number (minimum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160104<br>Name: Graph (domination number)<br>Description: Weighted domination number (Kakraba-Knisley, 2016)","Scale: KARS160113<br>Name: Graph (domination number)<br>Description: Weighted domination number using the atomic number (Kakraba-Knisley, 2016)","Scale: KARS160101<br>Name: Graph (order)<br>Description: Number of vertices (order of the graph) (Kakraba-Knisley, 2016)","Scale: KARS160102<br>Name: Graph (size)<br>Description: Number of edges (size of the graph) (Kakraba-Knisley, 2016)","Scale: KARS160103<br>Name: Graph (weighted degree)<br>Description: Total weighted degree of the graph (obtained by adding all the weights of all the vertices) (Kakraba-Knisley, 2016)","Scale: CHAM830104<br>Name: n atoms in side chain (2+1)<br>Description: The number of atoms in the side chain labelled 2+1 (Charton-Charton, 1983)","Scale: CHAM830105<br>Name: n atoms in side chain (3+1)<br>Description: The number of atoms in the side chain labelled 3+1 (Charton-Charton, 1983)","Scale: CHAM830106<br>Name: n bonds in longest chain<br>Description: The number of bonds in the longest chain (Charton-Charton, 1983)","Scale: LEVM760102<br>Name: Side chain length<br>Description: Distance between C-alpha and centroid of side chain (Levitt, 1976)","Scale: LEVM760105<br>Name: Side chain length<br>Description: Radius of gyration of side chain (Levitt, 1976)","Scale: FAUJ880104<br>Name: Steric parameter<br>Description: STERIMOL length of the side chain (Fauchere et al., 1988)","Scale: FAUJ880106<br>Name: Steric parameter (maximum width R)<br>Description: STERIMOL maximum width of the side chain (Fauchere et al., 1988)","Scale: CHAM830103<br>Name: n atoms in side chain (1+1)<br>Description: The number of atoms in the side chain labelled 1+1 (Charton-Charton, 1983)","Scale: FAUJ880101<br>Name: Shape Index<br>Description: Graph shape index (Fauchere et al., 1988)","Scale: LEVM760103<br>Name: Side chain angle (theta)<br>Description: Side chain angle theta(AAR) (Levitt, 1976)","Scale: CHAM810101<br>Name: Steric parameter<br>Description: Steric parameter (Charton, 1981)","Scale: FAUJ880102<br>Name: Steric parameter<br>Description: Smoothed upsilon steric parameter (Fauchere et al., 1988)","Scale: FAUJ880105<br>Name: Steric parameter (minimum width R)<br>Description: STERIMOL minimum width of the side chain (Fauchere et al., 1988)","Scale: KARS160118<br>Name: Atomic number (average)<br>Description: Average weighted atomic number or degree based on atomic number in the graph (Kakraba-Knisley, 2016)","Scale: KARS160110<br>Name: Eigenvalue (minimum)<br>Description: Minimum eigenvalue of the weighted Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160120<br>Name: Eigenvalue (minimum)<br>Description: Weighted minimum eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: RACS820113<br>Name: Side chain angle (theta)<br>Description: Value of theta(i) (Rackovsky-Scheraga, 1982)"],"marker":{"color":"rgba(240,162,0,1)","size":6,"line":{"color":"rgba(240,162,0,1)"}},"name":"Shape","textfont":{"color":"rgba(240,162,0,1)"},"error_y":{"color":"rgba(240,162,0,1)"},"error_x":{"color":"rgba(240,162,0,1)"},"line":{"color":"rgba(240,162,0,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[7.8263030338565196,2.2764109485088175,8.7139018338751999,8.2998618958404151,7.6481503504104431,3.8477971293033368,9.2454432072920749,8.5675802108035732,4.6575182344266191,9.2005605976830314,8.4468120963422617,4.5125894057019069,8.4151041825816204,6.390811712134524,8.7525129255214651,8.6797756685856431,1.9388893269031615,3.0406135279615656,3.5173413105407842,8.5132690429729045,9.8471065385716372,4.2510374623001894,7.107371234983094,-7.8353907767795121,-8.1422887718885075,-7.2608787890548969,-5.8877922654102077,-10.882754631502985,-12.840726120994946,-13.615119626952009,-12.134705077749198,-11.614261202571992,-13.052852270848236,-8.7908910394459845,-12.316206273110462,13.522885055676664,13.460067512568468,13.51085049828295,-6.6157652385968131,1.6375373936541779,-3.4278727125539463,-5.875553258599493,-6.1620447594034538,-7.2270653918822196,-0.77034206839600816,-0.82635058092800695,-4.5281706691689712,9.7271534336119938,10.640985776718818,10.741610544198846,4.0063289301981708,8.5153274107061296,9.0526580264433658,9.4231146581900393,9.0768723392943844,8.4429991148744481,8.8772879251461418,9.6082252223310505,8.519703459373261,2.3912170802308892,9.7429782390174644,8.8341812120387022,9.0482812800829109,9.6794500293775982],"y":[-18.61200177667709,-17.644204867135937,-19.117917720010638,-19.343478455460314,-18.726933772043576,-18.395139010071237,-19.35361552902755,-18.479347779216894,-18.192586012913235,-16.656154275261194,-16.736613998995772,-17.465265007708123,-17.119443322598375,-17.780981459942179,-17.288481347721284,-17.919844872236908,-17.695307763295293,-18.346121216260297,-18.039259269814824,-19.407589059259298,-18.332711696393016,-20.408339260427539,-16.550484443733946,14.155359533950685,13.824049579267362,13.773134463653472,13.691865053405101,10.922738041512957,13.159209896378883,13.946700683823853,13.868100748976675,13.946224404356213,13.037803336202341,15.459282997954215,15.047981759276297,-14.767932855153807,-14.816124574259325,-14.781792757647009,15.070476014624216,9.6228893076437334,13.866721387608422,15.852500517067853,15.571331796649853,14.539383163412284,8.9022484382693357,8.7640728529934133,12.994790808910199,9.2331506712885556,9.1058119396864274,10.437218858153694,11.642497839571414,10.257876738092536,9.7538827536796777,11.243604409063758,9.3330211111161141,10.249642778060107,10.531730260697566,10.387421187564009,10.455287619824954,10.803557897961644,9.8577931975889044,10.044393550720535,10.035979080136206,8.647721084090005],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: LINS030110<br>Name: ASA (folded coil/turn)<br>Description: Total median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030113<br>Name: ASA (folded coil/turn)<br>Description: % total accessible surfaces of whole residues in random coil/turn structures (Lins et al., 2003)","Scale: JANJ780101<br>Name: ASA (folded protein)<br>Description: Average accessible surface area (Janin et al., 1978)","Scale: JANJ780103<br>Name: ASA (folded protein)<br>Description: Percentage of exposed residues (Janin et al., 1978)","Scale: LINS030104<br>Name: ASA (folded protein)<br>Description: Total median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030107<br>Name: ASA (folded protein)<br>Description: % total accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: CHOC760102<br>Name: ASA (folded proteins)<br>Description: Residue accessible surface area in folded protein (Chothia, 1976)","Scale: LINS030116<br>Name: ASA (folded β-strand)<br>Description: Total median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: LINS030119<br>Name: ASA (folded β-strand)<br>Description: % total accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: LINS030103<br>Name: Hydrophilic ASA<br>Description: Hydrophilic accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: LINS030112<br>Name: Hydrophilic ASA (folded coil/turn)<br>Description: Hydrophilic median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030115<br>Name: Hydrophilic ASA (folded coil/turn)<br>Description: % Hydrophilic accessible surfaces of whole residues in random coil/turn structures (Lins et al., 2003)","Scale: LINS030106<br>Name: Hydrophilic ASA (folded proteins)<br>Description: Hydrophilic median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030109<br>Name: Hydrophilic ASA (folded proteins)<br>Description: % Hydrophilic accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: LINS030118<br>Name: Hydrophilic ASA (folded β-strand)<br>Description: Hydrophilic median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: LINS030121<br>Name: Hydrophilic ASA (folded β-strand)<br>Description: % Hydrophilic accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: LINS030114<br>Name: Hydrophobic ASA (folded coil/turn)<br>Description: % Hydrophobic accessible surfaces of whole residues in random coil/turn structures(Lins et al., 2003)","Scale: LINS030108<br>Name: Hydrophobic ASA (folded protein)<br>Description: % Hydrophobic accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: LINS030120<br>Name: Hydrophobic ASA (folded β-strand)<br>Description: % Hydrophobic accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: GUYH850104<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Janin index (Guy, 1985)","Scale: GUYH850105<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Chothia index (Guy, 1985)","Scale: RACS770103<br>Name: Side chain orientation<br>Description: Side chain orientational preference (Rackovsky-Scheraga, 1977)","Scale: VHEG790101<br>Name: TFE to lipophilic phase<br>Description: Transfer free energy to lipophilic phase (von Heijne-Blomberg, 1979)","Scale: BIOV880101<br>Name: Buriability<br>Description: Information value for accessibility; average fraction 35% (Biou et al., 1988)","Scale: BIOV880102<br>Name: Buriability<br>Description: Information value for accessibility; average fraction 23% (Biou et al., 1988)","Scale: WERD780101<br>Name: Buriability<br>Description: Propensity to be buried inside (Wertz-Scheraga, 1978)","Scale: ZHOH040103<br>Name: Buriability<br>Description: Buriability (Zhou-Zhou, 2004)","Scale: ARGP820103<br>Name: Buried<br>Description: Membrane-buried preference parameters (Argos et al., 1982)","Scale: CHOC760103<br>Name: Buried<br>Description: Proportion of residues 95% buried (Chothia, 1976)","Scale: CHOC760104<br>Name: Buried<br>Description: Proportion of residues 100% buried (Chothia, 1976)","Scale: JANJ780102<br>Name: Buried<br>Description: Percentage of buried residues (Janin et al., 1978)","Scale: JANJ790101<br>Name: Buried<br>Description: Ratio of buried and accessible molar fractions (Janin, 1979)","Scale: OLSK800101<br>Name: Buried<br>Description: Average internal preferences (Olsen, 1980)","Scale: NISK800101<br>Name: Interactivity<br>Description: 8 A contact number (Nishikawa-Ooi, 1980)","Scale: WARP780101<br>Name: Interactivity<br>Description: Average interactions per side chain atom (Warme-Morgan, 1978)","Scale: LINS030111<br>Name: Hydrophobic ASA (folded coil/turn)<br>Description: Hydrophobic median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030105<br>Name: Hydrophobic ASA (folded proteins)<br>Description: Hydrophobic median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030117<br>Name: Hydrophobic ASA (folded β-strand)<br>Description: Hydrophobic median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: PONP800107<br>Name: Accessibility reduction ratio<br>Description: Accessibility reduction ratio (Ponnuswamy et al., 1980)","Scale: ZIMJ680102<br>Name: Bulkiness<br>Description: Bulkiness (Zimmerman et al., 1968)","Scale: LINS030102<br>Name: Hydrophobic ASA<br>Description: Hydrophobic accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: BASU050101<br>Name: Interactivity<br>Description: Interactivity scale obtained from the contact matrix (Bastolla et al., 2005)","Scale: BASU050103<br>Name: Interactivity<br>Description: Interactivity scale obtained by maximizing the mean of correlation Coefficient over pairs of sequences sharing the TIM barrel fold (Bastolla et al., 2005)","Scale: NISK860101<br>Name: Interactivity<br>Description: 14 A contact number (Nishikawa-Ooi, 1986)","Scale: BULH740102<br>Name: Partial specific volume<br>Description: Apparent partial specific volume (Bull-Breese, 1974)","Scale: COHE430101<br>Name: Partial specific volume<br>Description: Partial specific volume (Cohn-Edsall, 1943)","Scale: PLIV810101<br>Name: Partition coefficient<br>Description: Partition Coefficient (Pliska et al., 1981)","Scale: CHOC760101<br>Name: Accessible surface area (ASA)<br>Description: Residue accessible surface area in tripeptide (Chothia, 1976)","Scale: LINS030101<br>Name: Accessible surface area (ASA)<br>Description: Total accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: RADA880106<br>Name: Accessible surface area (ASA)<br>Description: Accessible surface area (Radzicka-Wolfenden, 1988)","Scale: ROSG850101<br>Name: Mean area buried<br>Description: Mean area buried on transfer (Rose et al., 1985)","Scale: BIGC670101<br>Name: Volume<br>Description: Residue volume (Bigelow, 1967)","Scale: CHOC750101<br>Name: Volume<br>Description: Average volume of buried residue (Chothia, 1975)","Scale: DAWD720101<br>Name: Volume<br>Description: Size (Dawson, 1972)","Scale: FAUJ880103<br>Name: Volume<br>Description: Normalized van der Waals volume (Fauchere et al., 1988)","Scale: GOLD730102<br>Name: Volume<br>Description: Residue volume (Goldsack-Chalifoux, 1973)","Scale: GRAR740103<br>Name: Volume<br>Description: Volume (Grantham, 1974)","Scale: HARY940101<br>Name: Volume<br>Description: Mean volumes of residues buried in protein interiors (Harpaz et al., 1994)","Scale: KRIW790103<br>Name: Volume<br>Description: Side chain volume (Krigbaum-Komoriya, 1979)","Scale: LEVM760106<br>Name: Volume<br>Description: van der Waals parameter R0 (Levitt, 1976)","Scale: PONJ960101<br>Name: Volume<br>Description: Average volumes of residues (Pontius et al., 1996)","Scale: TSAJ990101<br>Name: Volume<br>Description: Volumes including the crystallographic waters using the ProtOr (Tsai et al., 1999)","Scale: TSAJ990102<br>Name: Volume<br>Description: Volumes not including the crystallographic waters using the ProtOr (Tsai et al., 1999)","Scale: FASG760101<br>Name: Weight<br>Description: Molecular weight (Fasman, 1976)"],"marker":{"color":"rgba(240,234,0,1)","size":6,"line":{"color":"rgba(240,234,0,1)"}},"name":"ASA/Volume","textfont":{"color":"rgba(240,234,0,1)"},"error_y":{"color":"rgba(240,234,0,1)"},"error_x":{"color":"rgba(240,234,0,1)"},"line":{"color":"rgba(240,234,0,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[-5.5679141366417237,-5.2184407330656777,-6.024746368806869,-3.2476994524228222,-7.4018639146112477,-6.7273990648874191,-4.9874981157155629,-5.1663716052154056,-4.7958446634782508,-3.5576583088790876,-3.5388542091264075,-5.8207519851604212,-5.5197875988024574,-2.8611506926713681,0.39041368433889873,0.29293750544855812,-1.4399531171058286,-3.3856724877986943,-10.282876050750925,-12.438558471370996,2.2345731024400082,1.1881119018996542,1.0293502987158587,-0.9707371802273721,-0.91397496486945295,-1.2762557310585321,2.4879892475325134,2.5942255491268225,2.4935628688676892,6.6182769536261468,5.9198045040088498,0.85592448557729828,-4.2261814507536775,8.3904517538294634,8.9444680915585124,0.86566427363776444,-5.2496803048650742,-4.9733944106227863,19.53408212491021,17.962429126166978,1.5086051762042436,5.6642114520303544,3.7235594741811324,19.58613227261138,21.575119748214064,4.9082398039740447,24.257630561497354,-7.6060042302655395,0.70154687022274176,-10.20910617648159,-3.7639155020371651,20.690461055074483,22.512245324465262,21.773484979950254,20.107203447988656,20.503015123890588,24.30113187837879,22.903655880449875,23.503792151923179,23.975761562698175,22.353513192253558,22.876107141420032,20.988510835495081,22.972566177587908,23.237390741807157,22.829142763637016,23.966303451488233,23.468276841446947,22.812740394465123,23.915449170736473,21.725807619511258,18.356615516865737,18.350697488666423,22.215395341936606,19.846266239705628,22.748236770042606,23.697432125738132,22.629374357678039,24.162780474475166,21.386126957787344,20.596192104210434,18.565978169863858,20.264480691746094,19.381576056860496,21.74239445744389,21.48256722154057,19.157861811379483,14.558161310062879,17.913295125159102,15.272955116941295,17.179037054229166,20.441375957209107,22.750443924199132,18.338579780531301,17.988279767394687,17.400022793008699,17.956409850408974,21.097073831669341,2.2608130692937207,0.28886187108939021,2.3140211607156762,-0.5616529208600336,5.6566752975949894,-11.938659905071145,-12.100888700923093,-13.677398676887558,-12.761043858139418,-12.699123770783341,-11.052017355919036,-12.112520316376715,-12.526603369848024,-11.30066094595554,-12.352262716513462,-8.888995710252539,-9.119029028832081,-2.2814626440931667,-2.2884263263696951,-9.8013994820022443,-2.0985192358565672,5.5504103748336631,5.6374043399920879,7.4801881505307266,7.0964639272803991,5.5509997730456941,6.851034692149363,25.095110894034192,6.5719238987106188,-10.012369361578902,23.186737455991267,22.2296201317093,25.158256512005668,22.875634575408476,22.886984517524144,-10.36902511850705,5.9185812700639069,2.5855982435721723,-5.4988381285244632,-5.6148367935822039,-6.6941418280874405,-4.1450269264685247,-7.5120377828163676,-5.8122455788940908,-4.3576382900763733,-5.2218482991014499,-6.8470791359626846,-5.4905666952841123,-5.6701604263975911,-3.5609161825224436,-2.3709915762665834,-4.4762267273322056,-5.0143633369470342,-4.4228224902222983,-4.7696044503311361,-5.3834014603597664,-6.5241236461843997,-10.262837470773899,-10.303844358080816,-10.21127338989576,-6.5817833504207286,-6.591895108801018,4.1806757916050143,9.8665681215472727,-7.4811387602317048,-7.2649445170188489,-6.5949170837207367,-7.9546046183985588,-7.0136180573692615,-6.9881877494724129,-6.3511904874836738,-6.9541140805828636,-6.9136866072128411,-7.9236382500159825,-6.6229712481795611,-9.3749734956824593,-5.82494118114617,-3.613793709356349,-5.214944758317813,-5.0767314630552205,-8.0479558016382935,-4.5992651014110626,-2.0984251327457746,-11.532455115378124,-6.2654935727547167,-7.8240542935826776,-6.7565636011139754,-6.0604769186051968,-7.3557410795644476,-6.1168629188639789,-4.9193582474407336,-4.956418687172949,-6.9399442478153404,-7.1027172289953198,-6.7868992158796688,-4.9528271165079341,-5.7632298764509517,-4.9248148888152521,-7.8399207599378169,-6.2842767648326809,-5.6599108104217661,-7.9121915356216936,-7.6219021762168699,-7.5421593324184988,-7.9559431040873507,-10.13449225129964,-10.256750639005045,-9.1626399291830101,-9.3088009480867004,-8.4961247671700235,-6.5501929877577378,-4.9283529700465758,-1.732994897284905,-1.8082129722620399,-1.5357654191320793,5.6722216845234401,5.7620498641461664,10.730678432721639,-7.7750999060993813,3.9565854581839739,-2.9080632681084788,17.061507232736147,3.434205665253681],"y":[-8.8394648754993419,-8.6670711985835567,-7.580378026292502,-9.8448444166786491,-8.3026934638932186,-5.2834556021622259,-7.1818190495474852,-8.0907758889540435,-8.4468530514067233,-8.6180279023322726,-8.5122443831098504,-10.771989901300053,-6.4738075275274545,-8.3109500875675089,-8.7393494171558395,-8.8150746409512504,-4.3299220287756057,8.6486587107021453,17.189420208544657,16.182704430203778,-4.266257123401747,-5.1027949852485746,-5.2152149457832904,-6.1741280291726595,-6.8720771117294994,-6.5508635742121744,-3.8376867354924227,-4.0931295984502301,-4.0142505815754932,-6.1659492599947265,-7.0402514218010559,-2.4981290673863565,5.7303407465525682,12.865508062314388,-13.350283955332653,11.335342511038748,-3.4268851716688413,-0.70386231854324399,1.5123775081545663,-7.6064418995377325,-9.009927710145675,-3.3978972414436712,-13.814342814368311,-5.2778588114331866,-6.7102313179209414,-8.5077769954158917,-5.8356573204126754,3.0850767026862149,-1.956072507584298,-2.5723604700499143,-16.376363892428657,-2.2786299969053947,-1.1439885583445109,-2.0126967269982043,-2.4252690138169308,-2.9951878148711586,-3.4941150655703885,-3.0433888404843366,-3.6167934797087287,-2.6644193765134618,-3.0263992185040887,-3.3413581280178075,-2.739796951953247,-4.2643278250043215,-4.1941717074645171,-2.8690357202315142,-4.2138431811557693,-4.2560369531820923,-2.8356581944107457,-2.3324359063753346,-4.6331445011349395,-3.8128024796679085,-3.735735853546152,-3.8495806245850024,-0.78166513113903746,-3.6821799684766954,-3.4263122018221188,-1.5142895647900538,-4.2702594837187045,-3.3927402867719656,-3.406500312621223,-4.098232838498209,-3.6469325340072802,-3.5868833607898427,-2.3968698291814796,-4.3141869755584494,-6.3169159393375827,-9.9235696815335057,-7.9061853608627279,-8.0355631216613492,-4.7861656891134112,-1.7449872932802124,-1.7213264704661808,-6.6792860057399634,-3.1746566524362327,-2.6746885372556934,-4.9618719370117175,-0.088413223652034675,-6.729865872227224,-7.5817809227331932,-6.6756378236846778,-10.379715851479874,-3.5433133354246076,-7.5380411016132882,-6.5400509458651861,-6.3552160315545816,-7.5799466995458191,-7.4381283005274366,-6.2221508189772559,-7.4565673632736278,-6.4488720199071166,-7.3866168663535063,-7.9992115384921174,-9.4127483659388158,-11.419245695203124,-12.76959988626632,-12.818919470071684,-10.984455946249613,-12.790927868664221,-8.7465785869509727,-9.4278377439867072,-10.89513616083428,-10.104761905501245,-9.5506437853562574,-9.7416528339072812,-4.4167274036441944,-6.1256090592752903,-9.2163711609152301,-5.3506527258808498,-5.6655223217961854,-2.4988633761759336,2.3147403751846789,2.3328784469064883,-11.324626283570533,-2.2015688895302952,12.860816142123412,18.042383220437195,20.43877655824863,19.796247585690537,18.265249106075053,20.035471778904611,20.915403341145094,21.682541602283866,21.350897784679354,19.925496530764025,21.647362383768613,20.284453947927894,20.931092543746271,17.822753118320932,21.849906662792783,22.60799883558343,20.200841089349062,19.942749990803051,19.608823371560685,20.63944748130697,-8.5429570069849508,-3.1595266137876106,-3.3531650983544412,-13.093758485885814,-3.6425964276448366,-16.691361524499658,-14.999139560580804,-13.434127786865174,-13.340668969334565,-13.17771014560399,21.982663025300646,23.29729262954282,23.389962806335863,23.541378567883861,21.469458784263288,23.228903211132472,12.404015868297732,21.294160597972926,18.225972376803142,21.998727970029346,21.12428243562174,21.089527631238727,19.99669809794409,18.672866609099632,19.881034504702754,-8.9988243361711255,-7.784602062954816,-10.574382074223662,-6.8987977102081484,-7.6823246205333842,-8.4503339906428145,-6.5732142231109139,-9.2250195273917104,-9.0007608877893652,-8.9090243910621272,-10.425072253999804,-9.2786851178608831,-7.1978281327580316,-10.048542178325903,-10.895097119136357,-9.0139984656369094,-6.6147587172367217,-7.7059591066358264,-8.3489718988637023,-6.3809958422022515,-9.3540088985025118,-3.7163985997424516,-6.489079180059572,-8.3421689265902454,-8.2801310501784258,-5.5806424556481202,-5.3890544171296355,-9.6007309798021296,-8.1614617785799819,-6.46729057292898,-7.6397848247416951,-7.6380981584016174,-7.55704327188535,-19.578154941288219,-19.545622349633316,-16.714819437921154,11.908162265964592,12.085407694644465,15.874605370305099,-2.4438102877680166,13.988571170291145],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: GEIM800108<br>Name: Aperiodic indices<br>Description: Aperiodic indices (Geisow-Roberts, 1980)","Scale: GEIM800111<br>Name: Aperiodic indices<br>Description: Aperiodic indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: CHAM830101<br>Name: Coil<br>Description: The Chou-Fasman parameter of the coil conformation (Charton-Charton, 1983)","Scale: GEIM800110<br>Name: Coil<br>Description: Aperiodic indices for beta-proteins (Geisow-Roberts, 1980)","Scale: NAGK730103<br>Name: Coil<br>Description: Normalized frequency of coil (Nagano, 1973)","Scale: QIAN880130<br>Name: Coil<br>Description: Weights for coil at the window position of -3 (Qian-Sejnowski, 1988)","Scale: QIAN880131<br>Name: Coil<br>Description: Weights for coil at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880132<br>Name: Coil<br>Description: Weights for coil at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880133<br>Name: Coil<br>Description: Weights for coil at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880134<br>Name: Coil<br>Description: Weights for coil at the window position of 1 (Qian-Sejnowski, 1988)","Scale: QIAN880135<br>Name: Coil<br>Description: Weights for coil at the window position of 2 (Qian-Sejnowski, 1988)","Scale: ROBB760112<br>Name: Coil<br>Description: Information measure for coil (Robson-Suzuki, 1976)","Scale: CHOP780207<br>Name: Non helical region (C-terminal)<br>Description: Normalized frequency of C-terminal non helical region (Chou-Fasman, 1978b)","Scale: QIAN880136<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 3 (Qian-Sejnowski, 1988)","Scale: QIAN880137<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880138<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 5 (Qian-Sejnowski, 1988)","Scale: ROBB760107<br>Name: Coil (C-terminal)<br>Description: Information measure for extended without H-bond (Robson-Suzuki, 1976)","Scale: QIAN880127<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -6 (Qian-Sejnowski, 1988)","Scale: QIAN880128<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -5 (Qian-Sejnowski, 1988)","Scale: GEOR030105<br>Name: Linker (small)<br>Description: Linker propensity from small dataset (linker length is less than six residues) (George-Heringa, 2003)","Scale: GEOR030101<br>Name: Linker<br>Description: Linker propensity from all dataset (George-Heringa, 2003)","Scale: GEOR030102<br>Name: Linker (1)<br>Description: Linker propensity from 1-linker dataset (George-Heringa, 2003)","Scale: GEOR030107<br>Name: Linker (Long)<br>Description: Linker propensity from long dataset (linker length is greater than 14 residues) (George-Heringa, 2003)","Scale: GEOR030109<br>Name: Linker (Non-Helical)<br>Description: Linker propensity from non-helical (annotated by DSSP) dataset (George-Heringa, 2003)","Scale: FINA910102<br>Name: α-helix (C-term, out)<br>Description: Helix initiation parameter at position i,i+1,i+2 (Finkelstein et al., 1991)","Scale: ONEK900102<br>Name: α-helix (C-term, out)<br>Description: Helix formation parameters (delta delta G) (O'Neil-DeGrado, 1990)","Scale: GEOR030103<br>Name: Linker (2)<br>Description: Linker propensity from 2-linker dataset (George-Heringa, 2003)","Scale: GEOR030104<br>Name: Linker (3)<br>Description: Linker propensity from 3-linker dataset (George-Heringa, 2003)","Scale: GEOR030106<br>Name: Linker (medium)<br>Description: Linker propensity from medium dataset (linker length is between six and 14 residues) (George-Heringa, 2003)","Scale: AURR980118<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C\" (Aurora-Rose, 1998)","Scale: AURR980101<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N4'(Aurora-Rose, 1998)","Scale: AURR980102<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N\"' (Aurora-Rose, 1998)","Scale: QIAN880139<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 6 (Qian-Sejnowski, 1988)","Scale: QIAN880129<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -4 (Qian-Sejnowski, 1988)","Scale: ISOY800107<br>Name: Double bend<br>Description: Normalized relative frequency of double bend (Isogai et al., 1980)","Scale: VASM830103<br>Name: Extended<br>Description: Relative population of conformational state E (Vasquez et al., 1983)","Scale: RACS820104<br>Name: Extended (left-handed)<br>Description: Average relative fractional occurrence in EL(i) (Rackovsky-Scheraga, 1982)","Scale: RACS820112<br>Name: Extended (right-handed)<br>Description: Average relative fractional occurrence in ER(i-1) (Rackovsky-Scheraga, 1982)","Scale: VASM830101<br>Name: α-helix<br>Description: Relative population of conformational state A (Vasquez et al., 1983)","Scale: RICJ880114<br>Name: α-helix (C-terminal, inside)<br>Description: Relative preference value at C1 (Richardson-Richardson, 1988)","Scale: AURR980120<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C4' (Aurora-Rose, 1998)","Scale: RICJ880117<br>Name: α-helix (C-terminal, outside)<br>Description: Relative preference value at C\" (Richardson-Richardson, 1988)","Scale: RACS820103<br>Name: α-helix (left-handed)<br>Description: Average relative fractional occurrence in AL(i) (Rackovsky-Scheraga, 1982)","Scale: RICJ880108<br>Name: α-helix (middle)<br>Description: Relative preference value at N5 (Richardson-Richardson, 1988)","Scale: QIAN880101<br>Name: α-helix (N-terminal)<br>Description: Weights for alpha-helix at the window position of -6 (Qian-Sejnowski, 1988)","Scale: RICJ880104<br>Name: α-helix (N-terminal, inside)<br>Description: Relative preference value at N1 (Richardson-Richardson, 1988)","Scale: GEIM800103<br>Name: α-helix (β-proteins)<br>Description: Alpha-helix indices for beta-proteins (Geisow-Roberts, 1980)","Scale: CHAM830102<br>Name: β-sheet<br>Description: A parameter defined from the residuals obtained from the best correlation of the Chou-Fasman parameter of beta-sheet (Charton-Charton, 1983)","Scale: QIAN880117<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -3 (Qian-Sejnowski, 1988)","Scale: QIAN880123<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 3 (Qian-Sejnowski, 1988)","Scale: OOBM850102<br>Name: β-turn (optimized propensity)<br>Description: Optimized propensity to form reverse turn (Oobatake et al., 1985)","Scale: AURR980109<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini N4 (Aurora-Rose, 1998)","Scale: AURR980111<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C5 (Aurora-Rose, 1998)","Scale: AURR980112<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C4 (Aurora-Rose, 1998)","Scale: AURR980113<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C3 (Aurora-Rose, 1998)","Scale: BEGF750101<br>Name: α-helix<br>Description: Conformational parameter of inner helix (Beghin-Dirkx, 1975)","Scale: BURA740101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Burgess et al., 1974)","Scale: CHOP780201<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Chou-Fasman, 1978b)","Scale: GEIM800101<br>Name: α-helix<br>Description: Alpha-helix indices (Geisow-Roberts, 1980)","Scale: GEIM800104<br>Name: α-helix<br>Description: Alpha-helix indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: ISOY800101<br>Name: α-helix<br>Description: Normalized relative frequency of alpha-helix (Isogai et al., 1980)","Scale: KANM800101<br>Name: α-helix<br>Description: Average relative probability of helix (Kanehisa-Tsong, 1980)","Scale: KANM800103<br>Name: α-helix<br>Description: Average relative probability of inner helix (Kanehisa-Tsong, 1980)","Scale: LEVM780101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix, with weights (Levitt, 1978)","Scale: LEVM780104<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix, unweighted (Levitt, 1978)","Scale: MAXF760101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Maxfield-Scheraga, 1976)","Scale: NAGK730101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Nagano, 1973)","Scale: PALJ810101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix from LG (Palau et al., 1981)","Scale: PALJ810102<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix from CF (Palau et al., 1981)","Scale: PALJ810109<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix in alpha/beta class (Palau et al., 1981)","Scale: QIAN880105<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880109<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of 2 (Qian-Sejnowski, 1988)","Scale: QIAN880110<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of 3 (Qian-Sejnowski, 1988)","Scale: RACS820108<br>Name: α-helix<br>Description: Average relative fractional occurrence in AR(i-1) (Rackovsky-Scheraga, 1982)","Scale: RICJ880112<br>Name: α-helix<br>Description: Relative preference value at C3 (Richardson-Richardson, 1988)","Scale: ROBB760101<br>Name: α-helix<br>Description: Information measure for alpha-helix (Robson-Suzuki, 1976)","Scale: TANS770101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Tanaka-Scheraga, 1977)","Scale: AURR980110<br>Name: α-helix (middle)<br>Description: Normalized positional residue frequency at helix termini N5 (Aurora-Rose, 1998)","Scale: CRAJ730101<br>Name: α-helix (middle)<br>Description: Normalized frequency of middle helix (Crawford et al., 1973)","Scale: QIAN880106<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880107<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880108<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of 1 (Qian-Sejnowski, 1988)","Scale: RICJ880109<br>Name: α-helix (middle)<br>Description: Relative preference value at Mid (Richardson-Richardson, 1988)","Scale: ROBB760103<br>Name: α-helix (middle)<br>Description: Information measure for middle helix (Robson-Suzuki, 1976)","Scale: AURR980108<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized positional residue frequency at helix termini N3 (Aurora-Rose, 1998)","Scale: QIAN880104<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -3 (Qian-Sejnowski, 1988)","Scale: AURR980116<br>Name: α-helix (C-terminal, C-cap)<br>Description: Normalized positional residue frequency at helix termini Cc (Aurora-Rose, 1998)","Scale: FINA910103<br>Name: α-helix (C-terminal, inside)<br>Description: Helix termination parameter at position j-2,j-1,j (Finkelstein et al., 1991)","Scale: RICJ880113<br>Name: α-helix (C-terminal, inside)<br>Description: Relative preference value at C2 (Richardson-Richardson, 1988)","Scale: FINA910104<br>Name: α-helix termination<br>Description: Helix termination parameter at position j+1 (Finkelstein et al., 1991)","Scale: BLAM930101<br>Name: α-helix<br>Description: Alpha-helix propensity of position 44 in T4 lysozyme (Blaber et al., 1993)","Scale: AURR980114<br>Name: α-helix (C-terminal)<br>Description: Normalized positional residue frequency at helix termini C2 (Aurora-Rose, 1998)","Scale: AURR980115<br>Name: α-helix (C-terminal)<br>Description: Normalized positional residue frequency at helix termini C1 (Aurora-Rose, 1998)","Scale: CHOP780205<br>Name: α-helix (C-terminal)<br>Description: Normalized frequency of C-terminal helix (Chou-Fasman, 1978b)","Scale: QIAN880111<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880112<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 5 (Qian-Sejnowski, 1988)","Scale: ROBB760104<br>Name: α-helix (C-terminal)<br>Description: Information measure for C-terminal helix (Robson-Suzuki, 1976)","Scale: RICJ880110<br>Name: α-helix (middle)<br>Description: Relative preference value at C5 (Richardson-Richardson, 1988)","Scale: MAXF760106<br>Name: α region<br>Description: Normalized frequency of alpha region (Maxfield-Scheraga, 1976)","Scale: AURR980119<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C\"' (Aurora-Rose, 1998)","Scale: ISOY800106<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized relative frequency of helix end (Isogai et al., 1980)","Scale: RICJ880116<br>Name: α-helix (C-terminal, outside)<br>Description: Relative preference value at C' (Richardson-Richardson, 1988)","Scale: TANS770102<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized frequency of isolated helix (Tanaka-Scheraga, 1977)","Scale: ISOY800108<br>Name: Coil (normalized by low Prolin value)<br>Description: Normalized relative frequency of coil (Isogai et al., 1980)","Scale: TANS770109<br>Name: Coil (normalized by low Prolin value)<br>Description: Normalized frequency of coil (Tanaka-Scheraga, 1977)","Scale: KIMC930101<br>Name: Free energy (β-sheet)<br>Description: Thermodynamic beta-sheet propensity (Kim-Berg, 1993)","Scale: AURR980117<br>Name: α-helix (C-terminal, C-cap)<br>Description: Normalized positional residue frequency at helix termini C' (Aurora-Rose, 1998)","Scale: RICJ880115<br>Name: α-helix (C-terminal, C-cap)<br>Description: Relative preference value at C-cap (Richardson-Richardson, 1988)","Scale: KOEP990101<br>Name: α-helix (left-handed)<br>Description: Alpha-helix propensity derived from designed sequences (Koehl-Levitt, 1999)","Scale: MAXF760104<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of left-handed alpha-helix (Maxfield-Scheraga, 1976)","Scale: MAXF760105<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of zeta L (Maxfield-Scheraga, 1976)","Scale: RACS820109<br>Name: α-helix (left-handed)<br>Description: Average relative fractional occurrence in AL(i-1) (Rackovsky-Scheraga, 1982)","Scale: TANS770107<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of left-handed helix (Tanaka-Scheraga, 1977)","Scale: PALJ810113<br>Name: β-turn (α class)<br>Description: Normalized frequency of turn in all-alpha class (Palau et al., 1981)","Scale: CHOP780206<br>Name: Non helical region (N-terminal)<br>Description: Normalized frequency of N-terminal non helical region (Chou-Fasman, 1978b)","Scale: AURR980104<br>Name: α-helix (N-terminal, N-cap)<br>Description: Normalized positional residue frequency at helix termini N'(Aurora-Rose, 1998)","Scale: AURR980105<br>Name: α-helix (N-terminal, N-cap)<br>Description: Normalized positional residue frequency at helix termini Nc (Aurora-Rose, 1998)","Scale: RICJ880103<br>Name: α-helix (N-terminal, N-cap)<br>Description: Relative preference value at N-cap (Richardson-Richardson, 1988)","Scale: FINA910101<br>Name: α-helix initiation<br>Description: Helix initiation parameter at position i-1 (Finkelstein et al., 1991)","Scale: AURR980106<br>Name: α-helix (N-terminal)<br>Description: Normalized positional residue frequency at helix termini N1 (Aurora-Rose, 1998)","Scale: CHOP780204<br>Name: α-helix (N-terminal)<br>Description: Normalized frequency of N-terminal helix (Chou-Fasman, 1978b)","Scale: RICJ880105<br>Name: α-helix (N-terminal)<br>Description: Relative preference value at N2 (Richardson-Richardson, 1988)","Scale: RICJ880106<br>Name: α-helix (N-terminal)<br>Description: Relative preference value at N3 (Richardson-Richardson, 1988)","Scale: ROBB760102<br>Name: α-helix (N-terminal)<br>Description: Information measure for N-terminal helix (Robson-Suzuki, 1976)","Scale: AURR980107<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized positional residue frequency at helix termini N2 (Aurora-Rose, 1998)","Scale: QIAN880102<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -5 (Qian-Sejnowski, 1988)","Scale: AURR980103<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N\" (Aurora-Rose, 1998)","Scale: RICJ880101<br>Name: α-helix (N-terminal, outside)<br>Description: Relative preference value at N\" (Richardson-Richardson, 1988)","Scale: PALJ810108<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized frequency of alpha-helix in alpha+beta class (Palau et al., 1981)","Scale: QIAN880103<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -4 (Qian-Sejnowski, 1988)","Scale: RACS820102<br>Name: α-helix (right-handed)<br>Description: Average relative fractional occurrence in AR(i) (Rackovsky-Scheraga, 1982)","Scale: GEIM800102<br>Name: α-helix (α-proteins)<br>Description: Alpha-helix indices for alpha-proteins (Geisow-Roberts, 1980)","Scale: PALJ810107<br>Name: α-helix (α-proteins)<br>Description: Normalized frequency of alpha-helix in all-alpha class (Palau et al., 1981)","Scale: MAXF760103<br>Name: β/α-bridge<br>Description: Normalized frequency of zeta R (Maxfield-Scheraga, 1976)","Scale: TANS770108<br>Name: β/α-bridge<br>Description: Normalized frequency of zeta R (Tanaka-Scheraga, 1977)","Scale: RACS820105<br>Name: Extended<br>Description: Average relative fractional occurrence in E0(i) (Rackovsky-Scheraga, 1982)","Scale: BEGF750102<br>Name: β-sheet<br>Description: Conformational parameter of beta-structure (Beghin-Dirkx, 1975)","Scale: CHOP780202<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet (Chou-Fasman, 1978b)","Scale: CHOP780208<br>Name: β-sheet<br>Description: Normalized frequency of N-terminal beta-sheet (Chou-Fasman, 1978b)","Scale: CHOP780209<br>Name: β-sheet<br>Description: Normalized frequency of C-terminal beta-sheet (Chou-Fasman, 1978b)","Scale: CRAJ730102<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet (Crawford et al., 1973)","Scale: KANM800102<br>Name: β-sheet<br>Description: Average relative probability of beta-sheet (Kanehisa-Tsong, 1980)","Scale: LEVM780102<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet, with weights (Levitt, 1978)","Scale: LEVM780105<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet, unweighted (Levitt, 1978)","Scale: NAGK730102<br>Name: β-sheet<br>Description: Normalized frequency of beta-structure (Nagano, 1973)","Scale: PALJ810103<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet from LG (Palau et al., 1981)","Scale: PALJ810104<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet from CF (Palau et al., 1981)","Scale: PALJ810110<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in all-beta class (Palau et al., 1981)","Scale: PALJ810111<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in alpha+beta class (Palau et al., 1981)","Scale: PALJ810112<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in alpha/beta class (Palau et al., 1981)","Scale: QIAN880118<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880119<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880120<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880121<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 1 (Qian-Sejnowski, 1988)","Scale: ROBB760106<br>Name: β-sheet<br>Description: Information measure for pleated-sheet (Robson-Suzuki, 1976)","Scale: ISOY800105<br>Name: β-sheet (C-terminal)<br>Description: Normalized relative frequency of bend S (Isogai et al., 1980)","Scale: QIAN880124<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880125<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 5 (Qian-Sejnowski, 1988)","Scale: QIAN880126<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 6 (Qian-Sejnowski, 1988)","Scale: CHOP780212<br>Name: β-turn (1st residue)<br>Description: Frequency of the 1st residue in turn (Chou-Fasman, 1978b)","Scale: KOEP990102<br>Name: Extended (designed β-sheet)<br>Description: Beta-sheet propensity derived from designed sequences (Koehl-Levitt, 1999)","Scale: RACS820101<br>Name: α-helix with flat 4-Cα structure (i)<br>Description: Average relative fractional occurrence in A0(i) (Rackovsky-Scheraga, 1982)","Scale: QIAN880114<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -6 (Qian-Sejnowski, 1988)","Scale: QIAN880115<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -5 (Qian-Sejnowski, 1988)","Scale: QIAN880116<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -4 (Qian-Sejnowski, 1988)","Scale: BURA740102<br>Name: Extended<br>Description: Normalized frequency of extended structure (Burgess et al., 1974)","Scale: ISOY800102<br>Name: Extended<br>Description: Normalized relative frequency of extended structure (Isogai et al., 1980)","Scale: MAXF760102<br>Name: Extended<br>Description: Normalized frequency of extended structure (Maxfield-Scheraga, 1976)","Scale: RACS820111<br>Name: Extended<br>Description: Average relative fractional occurrence in E0(i-1) (Rackovsky-Scheraga, 1982)","Scale: ROBB760105<br>Name: Extended<br>Description: Information measure for extended (Robson-Suzuki, 1976)","Scale: TANS770103<br>Name: Extended<br>Description: Normalized frequency of extended structure (Tanaka-Scheraga, 1977)","Scale: BAEK050101<br>Name: Linker index (negative linker propensity, 10-100 AA long region)<br>Description: Linker index (Bae et al., 2005)","Scale: KANM800104<br>Name: β-sheet<br>Description: Average relative probability of inner beta-sheet (Kanehisa-Tsong, 1980)","Scale: QIAN880122<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 2 (Qian-Sejnowski, 1988)","Scale: GEIM800105<br>Name: β-strand<br>Description: Beta-strand indices (Geisow-Roberts, 1980)","Scale: GEIM800106<br>Name: β-strand<br>Description: Beta-strand indices for beta-proteins (Geisow-Roberts, 1980)","Scale: GEIM800107<br>Name: β-strand<br>Description: Beta-strand indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: LIFS790101<br>Name: β-strand<br>Description: Conformational preference for all beta-strands (Lifson-Sander, 1979)","Scale: LIFS790102<br>Name: β-strand<br>Description: Conformational preference for parallel beta-strands (Lifson-Sander, 1979)","Scale: LIFS790103<br>Name: β-strand<br>Description: Conformational preference for antiparallel beta-strands (Lifson-Sander, 1979)","Scale: RACS820110<br>Name: Extended (left-handed)<br>Description: Average relative fractional occurrence in EL(i-1) (Rackovsky-Scheraga, 1982)","Scale: RACS820106<br>Name: Extended (right-handed)<br>Description: Average relative fractional occurrence in ER(i) (Rackovsky-Scheraga, 1982)","Scale: CHOP780211<br>Name: Non β region (C-terminal)<br>Description: Normalized frequency of C-terminal non beta region (Chou-Fasman, 1978b)","Scale: BEGF750103<br>Name: β-turn<br>Description: Conformational parameter of beta-turn (Beghin-Dirkx, 1975)","Scale: CHOP780101<br>Name: β-turn<br>Description: Normalized frequency of beta-turn (Chou-Fasman, 1978a)","Scale: CHOP780203<br>Name: β-turn<br>Description: Normalized frequency of beta-turn (Chou-Fasman, 1978b)","Scale: CRAJ730103<br>Name: β-turn<br>Description: Normalized frequency of turn (Crawford et al., 1973)","Scale: ISOY800103<br>Name: β-turn<br>Description: Normalized relative frequency of bend (Isogai et al., 1980)","Scale: LEVM780103<br>Name: β-turn<br>Description: Normalized frequency of reverse turn, with weights (Levitt, 1978)","Scale: LEVM780106<br>Name: β-turn<br>Description: Normalized frequency of reverse turn, unweighted (Levitt, 1978)","Scale: LEWP710101<br>Name: β-turn<br>Description: Frequency of occurrence in beta-bends (Lewis et al., 1971)","Scale: PALJ810105<br>Name: β-turn<br>Description: Normalized frequency of turn from LG (Palau et al., 1981)","Scale: PALJ810106<br>Name: β-turn<br>Description: Normalized frequency of turn from CF (Palau et al., 1981)","Scale: PALJ810114<br>Name: β-turn<br>Description: Normalized frequency of turn in all-beta class (Palau et al., 1981)","Scale: PALJ810115<br>Name: β-turn<br>Description: Normalized frequency of turn in alpha+beta class (Palau et al., 1981)","Scale: PRAM900104<br>Name: β-turn<br>Description: Relative frequency in reverse-turn (Prabhakaran, 1990)","Scale: ROBB760113<br>Name: β-turn<br>Description: Information measure for loop (Robson-Suzuki, 1976)","Scale: TANS770110<br>Name: β-turn<br>Description: Normalized frequency of chain reversal (Tanaka-Scheraga, 1977)","Scale: CHOP780216<br>Name: β-turn (middle)<br>Description: Normalized frequency of the 2nd and 3rd residues in turn (Chou-Fasman, 1978b)","Scale: ROBB760110<br>Name: β-turn (middle)<br>Description: Information measure for middle turn (Robson-Suzuki, 1976)","Scale: PALJ810116<br>Name: β-turn (α/β class)<br>Description: Normalized frequency of turn in alpha/beta class (Palau et al., 1981)","Scale: RACS820107<br>Name: α-helix with flat 4-Cα structure (i+1)<br>Description: Average relative fractional occurrence in A0(i-1) (Rackovsky-Scheraga, 1982)","Scale: ROBB760108<br>Name: β-turn<br>Description: Information measure for turn (Robson-Suzuki, 1976)","Scale: CHOP780214<br>Name: β-turn (3rd residue)<br>Description: Frequency of the 3rd residue in turn (Chou-Fasman, 1978b)","Scale: TANS770105<br>Name: β-turn (3rd residue)<br>Description: Normalized frequency of chain reversal S (Tanaka-Scheraga, 1977)","Scale: CHOP780215<br>Name: β-turn (4th residue)<br>Description: Frequency of the 4th residue in turn (Chou-Fasman, 1978b)","Scale: ROBB760111<br>Name: β-turn (C-terminal)<br>Description: Information measure for C-terminal turn (Robson-Suzuki, 1976)","Scale: GEIM800109<br>Name: Aperiodic indices<br>Description: Aperiodic indices for alpha-proteins (Geisow-Roberts, 1980)","Scale: CHOP780210<br>Name: Non β region (N-terminal)<br>Description: Normalized frequency of N-terminal non beta region (Chou-Fasman, 1978b)","Scale: ROBB760109<br>Name: β-turn (1st residue)<br>Description: Information measure for N-terminal turn (Robson-Suzuki, 1976)","Scale: CHOP780213<br>Name: β-turn (2nd residue)<br>Description: Frequency of the 2nd residue in turn (Chou-Fasman, 1978b)","Scale: ISOY800104<br>Name: β-turn (2nd residue)<br>Description: Normalized relative frequency of bend R (Isogai et al., 1980)","Scale: TANS770104<br>Name: β-turn (2nd residue)<br>Description: Normalized frequency of chain reversal R (Tanaka-Scheraga, 1977)","Scale: MONM990101<br>Name: β-turn (TM helix)<br>Description: Turn propensity scale for transmembrane helices (Monne et al., 1999)","Scale: MONM990201<br>Name: β-turn (TM helix)<br>Description: Averaged turn propensities in a transmembrane helix (Monne et al., 1999)","Scale: TANS770106<br>Name: β-turn in double bend<br>Description: Normalized frequency of chain reversal D (Tanaka-Scheraga, 1977)","Scale: SUYM030101<br>Name: Linker index (negative linker propensity, 10-100 AA long region)<br>Description: Linker propensity index (Suyama-Ohara, 2003)","Scale: RICJ880107<br>Name: α-helix<br>Description: Relative preference value at N4 (Richardson-Richardson, 1988)","Scale: RICJ880111<br>Name: α-helix<br>Description: Relative preference value at C4 (Richardson-Richardson, 1988)","Scale: QIAN880113<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 6 (Qian-Sejnowski, 1988)","Scale: FODM020101<br>Name: π-helix<br>Description: Propensity of amino acids within pi-helices (Fodje-Al-Karadaghi, 2002)"],"marker":{"color":"rgba(98,201,35,1)","size":6,"line":{"color":"rgba(98,201,35,1)"}},"name":"Conformation","textfont":{"color":"rgba(98,201,35,1)"},"error_y":{"color":"rgba(98,201,35,1)"},"error_x":{"color":"rgba(98,201,35,1)"},"line":{"color":"rgba(98,201,35,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[-20.975308466004396,-20.541165558258427,-20.919417478007123,-21.096379354455525,-20.338255940276547,-19.878151363434146,-20.822741731457054,-20.907676246792747,-20.351325946427057,-21.892966514188991,-20.462031378060544,-21.780848653093827,-21.784054581167887,-20.810723498470264,-20.913764210008761,-20.442107264268877,-19.915443770511541,-20.400000070914032,-21.413833002883589,-19.760777467247234,-20.015488309530916,-18.337824943070437,-19.436644932843553,-21.266007955583458,2.3146110009080734,-0.45866361791068944,2.6354149817620587,2.7432933478142569,-19.81272691327742,-19.728061352206179,-18.507867960724923,-18.959341808497133,-18.361969168640758,-18.537706324056625,-19.21489306497805,-19.570156796051265,-18.756849027325011,-18.426242582174115,-19.556270991576657,-19.110089262540185,-19.01461003653796,-1.0568281047329346,-19.136266488568289,-0.98619334989434471,-1.3864753831372008,5.5867111475849942,5.4649103471621876,2.2312965053242428,3.1764957485567518,0.96326070163361466,1.6457078042423863,1.6792563219353924,2.1565718640869167,2.0477114049816425,2.555088160898709,0.41970889213967361,-15.225036158001963,-5.3359635683969433],"y":[2.1037383741004634,2.0300356486004008,2.6976873452963219,2.5851248715529924,1.9224398184096159,0.6430296663537185,1.8004074604546152,2.4823250986998064,1.7150472717894796,2.3163082888936257,0.82974104053966291,1.6298818656383223,1.684073252927788,0.52526456128308008,0.17842824663036069,0.36418068093543687,1.6891443131850807,3.1729111498017843,3.2698788657225117,2.6600703179705967,3.2042777407331884,2.8124047696921384,2.3670070424826863,-0.72930691815489745,-15.007822221048796,-17.916225134819364,-14.798597838871592,-14.529825489051387,4.533322621365719,4.6884277565224899,6.1630082379712627,7.3102362813308206,6.3253687414746622,6.1849886864087189,7.2905556585183078,6.9089188039642471,6.1542098438201256,2.8428834591761429,5.2948743743367714,5.4324436671051233,5.4094500817998004,16.463653572682034,7.4128473730873381,16.386651930455994,16.836173691633036,-18.473645550728296,-17.343358044432232,-17.092479263463876,-19.604915497296897,-17.722519147741284,-19.806404955101303,-19.771071348053393,-18.545044271266388,-18.250973905457666,-18.148086863710041,-20.968953028945712,7.626196627978584,-2.900450110932721],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: DAYM780101<br>Name: AA composition<br>Description: Amino acid composition (Dayhoff et al., 1978a)","Scale: JOND920101<br>Name: AA composition<br>Description: Relative frequency of occurrence (Jones et al., 1992)","Scale: JUKT750101<br>Name: AA composition<br>Description: Amino acid distribution (Jukes et al., 1975)","Scale: JUNJ780101<br>Name: AA composition<br>Description: Sequence frequency (Jungck, 1978)","Scale: NAKH900101<br>Name: AA composition<br>Description: AA composition of total proteins (Nakashima et al., 1990)","Scale: NAKH900102<br>Name: AA composition<br>Description: SD of AA composition of total proteins (Nakashima et al., 1990)","Scale: CEDJ970102<br>Name: AA composition (anchored)<br>Description: Composition of amino acids in anchored proteins [%] (Cedano et al., 1997)","Scale: CEDJ970101<br>Name: AA composition (EXT)<br>Description: Composition of amino acids in extracellular proteins [%] (Cedano et al., 1997)","Scale: CEDJ970104<br>Name: AA composition (INT)<br>Description: Composition of amino acids in intracellular proteins [%] (Cedano et al., 1997)","Scale: NAKH920107<br>Name: Membrane proteins (multi-spanning, EXT)<br>Description: AA composition of EXT of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920106<br>Name: Membrane proteins (multi-spanning, INT)<br>Description: AA composition of CYT of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920103<br>Name: Membrane proteins (single-spanning, EXT)<br>Description: AA composition of EXT of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920104<br>Name: Membrane proteins (single-spanning, EXT)<br>Description: AA composition of EXT2 of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920101<br>Name: Membrane proteins (single-spanning, INT)<br>Description: AA composition of CYT of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920102<br>Name: Membrane proteins (single-spanning, INT)<br>Description: AA composition of CYT2 of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: CEDJ970105<br>Name: Nuclear proteins<br>Description: Composition of amino acids in nuclear proteins [%] (Cedano et al., 1997)","Scale: FUKS010112<br>Name: Nuclear proteins<br>Description: Entire chain composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000102<br>Name: Proteins of mesophiles<br>Description: Distribution of amino acid residues in the 18 non-redundant families of mesophilic proteins (Kumar et al., 2000)","Scale: FUKS010111<br>Name: Proteins of mesophiles (EXT)<br>Description: Entire chain composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010110<br>Name: Proteins of mesophiles (INT)<br>Description: Entire chain composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000101<br>Name: Proteins of thermophiles<br>Description: Distribution of amino acid residues in the 18 non-redundant families of thermophilic proteins (Kumar et al., 2000)","Scale: KUMS000103<br>Name: Proteins of thermophiles<br>Description: Distribution of amino acid residues in the alpha-helices in thermophilic proteins (Kumar et al., 2000)","Scale: FUKS010109<br>Name: Proteins of thermophiles (INT)<br>Description: Entire chain composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: DIGM050101<br>Name: Hydrostatic pressure<br>Description: Hydrostatic Pressure asymmetry index, PAI (Di Giulio, 2005)","Scale: FUKS010104<br>Name: Nuclear proteins<br>Description: Surface composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010103<br>Name: Proteins of mesophiles (EXT)<br>Description: Surface composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010102<br>Name: Proteins of mesophiles (INT)<br>Description: Surface composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010101<br>Name: Proteins of thermophiles (INT)<br>Description: Surface composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: CEDJ970103<br>Name: Membrane proteins<br>Description: Composition of amino acids in membrane proteins [%] (Cedano et al., 1997)","Scale: NAKH900109<br>Name: Membrane proteins<br>Description: AA composition of membrane proteins (Nakashima et al., 1990)","Scale: NAKH900111<br>Name: Membrane proteins<br>Description: Transmembrane regions of non-mt-proteins (Nakashima et al., 1990)","Scale: NAKH900112<br>Name: Membrane proteins<br>Description: Transmembrane regions of mt-proteins (Nakashima et al., 1990)","Scale: NAKH920108<br>Name: Membrane proteins (multi-spanning)<br>Description: AA composition of MEM of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920105<br>Name: Membrane proteins (single-spanning)<br>Description: AA composition of MEM of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH900103<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins (Nakashima et al., 1990)","Scale: NAKH900107<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins from fungi and plant (Nakashima et al., 1990)","Scale: FUKS010108<br>Name: Nuclear proteins (interior)<br>Description: Interior composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000104<br>Name: Proteins of mesophiles<br>Description: Distribution of amino acid residues in the alpha-helices in mesophilic proteins (Kumar et al., 2000)","Scale: FUKS010107<br>Name: Proteins of mesophiles (EXT)<br>Description: Interior composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010106<br>Name: Proteins of mesophiles (INT)<br>Description: Interior composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010105<br>Name: Proteins of thermophiles (INT)<br>Description: Interior composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: NAKH900104<br>Name: Mitochondrial proteins<br>Description: Normalized composition of mt-proteins (Nakashima et al., 1990)","Scale: NAKH900105<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins from animal (Nakashima et al., 1990)","Scale: NAKH900106<br>Name: Mitochondrial proteins<br>Description: Normalized composition from animal (Nakashima et al., 1990)","Scale: NAKH900108<br>Name: Mitochondrial proteins<br>Description: Normalized composition from fungi and plant (Nakashima et al., 1990)","Scale: PUNT030101<br>Name: Membrane proteins (anchor)<br>Description: Knowledge-based membrane-propensity scale from 1D_Helix in MPtopo databases (Punta-Maritan, 2003)","Scale: PUNT030102<br>Name: Membrane proteins (anchor)<br>Description: Knowledge-based membrane-propensity scale from 3D_Helix in MPtopo databases (Punta-Maritan, 2003)","Scale: OOBM770103<br>Name: Non-bonded Energy per Atom<br>Description: Long range non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: GUYH850101<br>Name: Partition energy<br>Description: Partition energy (Guy, 1985)","Scale: GUYH850102<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Wertz-Scheraga index (Guy, 1985)","Scale: MIYS990101<br>Name: Partition energy<br>Description: Relative partition energies derived by the Bethe approximation (Miyazawa-Jernigan, 1999)","Scale: MIYS990102<br>Name: Partition energy<br>Description: Optimized relative partition energies - method A (Miyazawa-Jernigan, 1999)","Scale: MIYS990103<br>Name: Partition energy<br>Description: Optimized relative partition energies - method B (Miyazawa-Jernigan, 1999)","Scale: MIYS990104<br>Name: Partition energy<br>Description: Optimized relative partition energies - method C (Miyazawa-Jernigan, 1999)","Scale: MIYS990105<br>Name: Partition energy<br>Description: Optimized relative partition energies - method D (Miyazawa-Jernigan, 1999)","Scale: BULH740101<br>Name: TFE to surface<br>Description: Transfer free energy to surface (Bull-Breese, 1974)","Scale: NAKH900113<br>Name: Average Composition<br>Description: Ratio of average and computed composition (Nakashima et al., 1990)","Scale: GRAR740101<br>Name: Substitution Frequency<br>Description: Composition (Grantham, 1974)"],"marker":{"color":"rgba(10,154,155,1)","size":6,"line":{"color":"rgba(10,154,155,1)"}},"name":"Composition","textfont":{"color":"rgba(10,154,155,1)"},"error_y":{"color":"rgba(10,154,155,1)"},"error_x":{"color":"rgba(10,154,155,1)"},"line":{"color":"rgba(10,154,155,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[10.964668481791355,5.3519419813977001,0.9715376111553109,8.3843902403470985,7.0810628230561656,6.1576438413429182,-9.0552485421968907,-4.5428398018645764,-7.4652204998808216,-7.3606340333196201,-6.3338221285479843,-6.3859437398404371,-6.9658156178779755,-8.8191382453722493,-11.155341864842093,-11.300625641645789,-6.1345165416082041,-10.081159976758334,-10.896531630506637,6.3813811571361523,7.7175911931498806,-4.042853894375626,8.6748301516457627,7.6189662533427196,2.5940135918768821,4.8639892671913669,5.282791279254746,7.784973111864864,5.0819469368078618,6.0349401146438719,9.4074528240146602,6.5981262127701532,8.5524325273867685,6.3406452896363783,6.8983773176408114,4.6678295303047053,9.0791439104301102,5.3967161500028569,7.6215534931827289,5.343403918139269,6.7972655339298287,8.7176928472630344,8.698181920486638,1.4145181604036301,-6.8898734209880459,-4.9123769568666225,-0.70457058907211834,-3.8706581748663593,-3.7196755711591796,-4.982571934728421,-5.5858141325385242,-4.57862344812537,-13.678299528031104,-11.849507130908552,-5.7165838543227423,-0.68892436797115175,-0.70261309323520627,-12.485486784304832,-12.546066431907247,-10.669119308737363,-9.9992445699081731,-9.7623194151478554,-7.8054005913104954,-8.0456666147201421,-4.5507369391822694,-4.1936643126521949,-3.7158613116348946,-1.1297938136697716,-0.85104338281050129,-4.2430714612623897,-0.46196730894423826,-1.3801533877335763,-3.612651868794869,-3.6278593087249162,-0.54953255827268788,-2.337057401568384,-2.3152512927567725,-1.5928706910395556,-13.738997361278846,-2.1509408492096904,-0.71933418686895578,0.65546737741718619,-8.1666404323944199,-4.3898127066298382,-6.9288906701334358,-14.645302822023114,-14.532675851204985,-8.4998142168495079,-5.1706820237438951,-11.49891647409842,-9.5907251082560627,-9.3762071669840239,-14.092076079492982,-5.5190875631759688,-7.9491851788345533,-8.7998242863131804,-9.2318181859058175,-9.5891498804356523,-14.00431695893097,-10.498505908786958,-8.9639987140526323,-8.8204335761991732,-10.801824263403601,-9.1147572880305905,-15.414888927652381,16.729823185231549,-15.389653909321133,0.35738679381946697],"y":[11.676383976374943,11.000470268039964,14.774869084735341,9.4538233704628123,9.3820201974560753,10.641775584370913,15.252613798166534,16.236982885519428,16.68707283541762,16.71788974836246,17.214194504978487,17.155311208497626,16.820959705017479,14.329808226806033,15.223394043827234,14.774726073548237,10.21486885327848,14.652690108463275,10.928871159660847,-14.14908012016685,-19.184077248694191,-15.59007590833617,-14.89338842987204,-17.119431325789193,-18.924825722421076,-16.659286886466258,-14.953232335107012,-15.885379000572909,-15.403555927125884,-18.522205952776702,-17.31979383849967,-19.281529528266041,-19.36578825874917,-18.528490254320044,-18.089437680186649,-15.140679311894127,-15.780790256365131,-14.898378324486105,-17.119231369213985,-16.010080229942272,-15.1694812140182,-16.128319690938287,-16.229678249669067,-19.285946418115881,14.346944237278361,14.044144181703647,11.192311961095628,14.833977970703387,14.196727489641617,14.75171271387917,14.527357078792031,14.447237508162736,11.714932811072046,12.430916170132814,12.814968584056395,11.06807232207737,11.191231503515683,12.936102077264669,12.983691293295692,13.878522243659798,13.729802019664978,13.532629469719213,15.750220968423521,14.369095204926035,16.237339455134798,11.097047831762739,10.952366535991105,10.135680033404279,13.044921001689838,11.968490006526986,13.871797096779623,13.169842217713033,12.314937880981946,12.105577183679307,13.126789359911886,13.890557710659003,12.769750975902229,10.732370268171008,11.679924558520062,13.622009353900422,11.082783425140503,14.73627233810843,10.727738314753894,-2.4948936947005143,-1.3954297816612697,12.315481684904901,12.388706020232323,13.548652386906543,11.252525666988351,12.822196365526793,12.474546053981312,11.993219461174931,11.516599436616721,12.093537266072765,15.911627405216255,16.072187200981279,15.886334154816423,15.788731877913156,14.288956851048466,16.40647349285619,16.900357595700438,15.496435024588145,8.6163572055278266,11.111953056055754,8.4504936852605432,-5.2514110026410181,8.734616509803816,-7.3352076729288225],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: MITS020101<br>Name: Amphiphilicity<br>Description: Amphiphilicity index (Mitaku et al., 2002)","Scale: GARJ730101<br>Name: Partition coefficient<br>Description: Partition Coefficient (Garel et al., 1973)","Scale: ZASB820101<br>Name: Partition coefficient<br>Description: Dependence of Partition Coefficient on ionic strength (Zaslavsky et al., 1982)","Scale: CHAM820101<br>Name: Polarizability<br>Description: Polarizability parameter (Charton-Charton, 1982)","Scale: MCMT640101<br>Name: Refractivity<br>Description: Refractivity (McMeekin et al., 1964), Cited by Jones (1975)","Scale: LEVM760107<br>Name: van der Waals parameter epsilon<br>Description: van der Waals parameter epsilon (Levitt, 1976)","Scale: CORJ870101<br>Name: Amphiphilicity (α-helix)<br>Description: NNEIG index (Cornette et al., 1987)","Scale: CORJ870102<br>Name: Amphiphilicity (α-helix)<br>Description: SWEIG index (Cornette et al., 1987)","Scale: CORJ870103<br>Name: Amphiphilicity (α-helix)<br>Description: PRIFT index (Cornette et al., 1987)","Scale: CORJ870104<br>Name: Amphiphilicity (α-helix)<br>Description: PRILS index (Cornette et al., 1987)","Scale: CORJ870105<br>Name: Amphiphilicity (α-helix)<br>Description: ALTFT index (Cornette et al., 1987)","Scale: CORJ870106<br>Name: Amphiphilicity (α-helix)<br>Description: ALTLS index (Cornette et al., 1987)","Scale: CORJ870107<br>Name: Amphiphilicity (α-helix)<br>Description: TOTFT index (Cornette et al., 1987)","Scale: ROSG850102<br>Name: Mean fractional area loss<br>Description: Mean fractional area loss (Rose et al., 1985)","Scale: DESM900101<br>Name: Membrane preference<br>Description: Membrane preference for cytochrome b: MPH89 (Degli Esposti et al., 1990)","Scale: DESM900102<br>Name: Membrane preference<br>Description: Average membrane preference: AMP07 (Degli Esposti et al., 1990)","Scale: NAKH900110<br>Name: Membrane preference<br>Description: Normalized composition of Membrane proteins (Nakashima et al., 1990)","Scale: MEIH800103<br>Name: Side chain angle (phi)<br>Description: Average side chain orientation angle (Meirovitch et al., 1980)","Scale: ARGP820102<br>Name: Signal sequence helical potential<br>Description: Signal sequence helical potential (Argos et al., 1982)","Scale: HOPA770101<br>Name: Hydration Number<br>Description: Hydration Number (Hopfinger, 1971), Cited by Charton-Charton (1982)","Scale: OOBM770101<br>Name: Non-bonded energy per atom<br>Description: Average non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: OOBM850104<br>Name: Non-bonded energy per atom<br>Description: Optimized average non-bonded energy per atom (Oobatake et al., 1985)","Scale: FAUJ880110<br>Name: Number full non-bonding orbitals<br>Description: Number of full nonbonding orbitals (Fauchere et al., 1988)","Scale: ENGD860101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity index (Engelman et al., 1986)","Scale: FASG890101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity index (Fasman, 1989)","Scale: GRAR740102<br>Name: Polarity (hydrophilicity)<br>Description: Polarity (Grantham, 1974)","Scale: HOPT810101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophilicity value (Hopp-Woods, 1981)","Scale: KIDA850101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity-related index (Kidera et al., 1985)","Scale: KOEH090101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Wimley-White, 1996), Cited by Koehler et al. (2009)","Scale: KOEH090102<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Hessa et al., 2007), Cited by Koehler et al. (2009)","Scale: KOEH090103<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Eisenberg-Weiss, 1984), Inverted to match the direction of transfer from water to bilayer by Koehler et al. (2009)","Scale: KOEH090104<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Kyte-Doolittle, 1982), Inverted to match the direction of transfer from water to bilayer by Koehler et al. (2009)","Scale: KOEH090106<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Janin, 1979), Cited by Koehler et al. (2009)","Scale: KOEH090109<br>Name: Polarity (hydrophilicity)<br>Description: Unified Hydrophobicity Scale, UHS  (Koehler et al., 2009)","Scale: KOEH090110<br>Name: Polarity (hydrophilicity)<br>Description: Mammalian Hydrophobicity Scale, MHS (Koehler et al., 2009)","Scale: KOEH090111<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale transition between water-interface from (Wimley-White, 1998), Cited by Koehler et al. (2009)","Scale: KUHL950101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophilicity scale (Kuhn et al., 1995)","Scale: LEVM760101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobic parameter (Levitt, 1976)","Scale: PRAM900101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity (Prabhakaran, 1990)","Scale: WOEC730101<br>Name: Polarity (hydrophilicity)<br>Description: Polar requirement (Woese, 1973)","Scale: ZIMJ680103<br>Name: Polarity (hydrophilicity)<br>Description: Polarity (Zimmerman et al., 1968)","Scale: ROSM880101<br>Name: Side chain hydropathy<br>Description: Side chain hydropathy, uncorrected for solvation (Roseman, 1988)","Scale: ROSM880102<br>Name: Side chain hydropathy<br>Description: Side chain hydropathy, corrected for solvation (Roseman, 1988)","Scale: CORJ870108<br>Name: TOTLS index<br>Description: TOTLS index (Cornette et al., 1987)","Scale: MIYS850101<br>Name: Effective partition energy<br>Description: Effective partition energy (Miyazawa-Jernigan, 1985)","Scale: ROBB790101<br>Name: Hydration free energy<br>Description: Hydration free energy (Robson-Osguthorpe, 1979)","Scale: ARGP820101<br>Name: Hydrophobicity<br>Description: Hydrophobicity index (Argos et al., 1982)","Scale: CIDH920101<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha-proteins (Cid et al., 1992)","Scale: CIDH920102<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for beta-proteins (Cid et al., 1992)","Scale: CIDH920103<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha+beta-proteins (Cid et al., 1992)","Scale: CIDH920104<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha/beta-proteins (Cid et al., 1992)","Scale: CIDH920105<br>Name: Hydrophobicity<br>Description: Normalized average hydrophobicity scales (Cid et al., 1992)","Scale: EISD840101<br>Name: Hydrophobicity<br>Description: Consensus normalized hydrophobicity scale (Eisenberg, 1984)","Scale: EISD860103<br>Name: Hydrophobicity<br>Description: Direction of hydrophobic moment (Eisenberg-McLachlan, 1986)","Scale: FAUJ830101<br>Name: Hydrophobicity<br>Description: Hydrophobic parameter π (Fauchere-Pliska, 1983)","Scale: GOLD730101<br>Name: Hydrophobicity<br>Description: Hydrophobicity factor (Goldsack-Chalifoux, 1973)","Scale: JOND750101<br>Name: Hydrophobicity<br>Description: Hydrophobicity (Jones, 1975)","Scale: JURD980101<br>Name: Hydrophobicity<br>Description: Modified Kyte-Doolittle hydrophobicity scale (Juretic et al., 1998)","Scale: KYTJ820101<br>Name: Hydrophobicity<br>Description: Hydropathy index (Kyte-Doolittle, 1982)","Scale: NADH010102<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (9% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010103<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (16% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010104<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (20% accessibility) (Naderi-Manesh et al., 2001)","Scale: PONP930101<br>Name: Hydrophobicity<br>Description: Hydrophobicity scales (Ponnuswamy, 1993)","Scale: RADA880108<br>Name: Hydrophobicity<br>Description: Mean polarity (Radzicka-Wolfenden, 1988)","Scale: SWER830101<br>Name: Hydrophobicity<br>Description: Optimal matching hydrophobicity (Sweet-Eisenberg, 1983)","Scale: WILM950101<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C18 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: WILM950102<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C8 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: ZIMJ680101<br>Name: Hydrophobicity<br>Description: Hydrophobicity (Zimmerman et al., 1968)","Scale: BROC820101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in TFA (Browne et al., 1982)","Scale: GUOD860101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient at pH 2 (Guo et al., 1986)","Scale: MEEJ800101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in HPLC, pH7.4 (Meek, 1980)","Scale: MEEJ800102<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in HPLC, pH2.1 (Meek, 1980)","Scale: MEEJ810101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in NaClO4 (Meek-Rossetti, 1981)","Scale: MEEJ810102<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in NaH2PO4 (Meek-Rossetti, 1981)","Scale: BROC820102<br>Name: Retention coefficient<br>Description: Retention Coefficient in HFBA (Browne et al., 1982)","Scale: EISD860101<br>Name: Solvation free energy<br>Description: Solvation free energy (Eisenberg-McLachlan, 1986)","Scale: NOZY710101<br>Name: Transfer free energy (TFE)<br>Description: Transfer energy, organic solvent/water (Nozaki-Tanford, 1971)","Scale: LAWE840101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy, CHP/water (Lawson et al., 1984)","Scale: RADA880101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy from chx to wat (Radzicka-Wolfenden, 1988)","Scale: RADA880102<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy from oct to wat (Radzicka-Wolfenden, 1988)","Scale: SIMZ760101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy (Simon, 1976), Cited by Charton-Charton (1982)","Scale: WIMW960101<br>Name: Transfer free energy (TFE) to outside<br>Description: Free energies of transfer of AcWl-X-LL peptides from bilayer interface to water (Wimley-White, 1996)","Scale: CHAM820102<br>Name: Free energy (interface)<br>Description: Free energy of solution in water, kcal/mole (Charton-Charton, 1982)","Scale: KOEH090112<br>Name: Hydrophobicity (interface)<br>Description: Hydrophobicity scale transition between water-interface, Unified Hydrophobicity Scale interface, UHSint  (Koehler et al., 2009)","Scale: WILM950103<br>Name: Hydrophobicity (interface)<br>Description: Hydrophobicity Coefficient in RP-HPLC, C4 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: WOLR790101<br>Name: Hydration potential<br>Description: Hydrophobicity index (Wolfenden et al., 1979)","Scale: WOLR810101<br>Name: Hydration potential<br>Description: Hydration potential (Wolfenden et al., 1981)","Scale: CASG920101<br>Name: Hydrophobicity<br>Description: Hydrophobicity scale from native protein structures (Casari-Sippl, 1992)","Scale: COWR900101<br>Name: Hydrophobicity<br>Description: Hydrophobicity index, 3.0 pH (Cowan-Whittaker, 1990)","Scale: NADH010101<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (5% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010105<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (25% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010106<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (36% accessibility) (Naderi-Manesh et al., 2001)","Scale: JACR890101<br>Name: Hydrophobicity (surrounding)<br>Description: Weights from the IFH scale (Jacobs-White, 1989)","Scale: BLAS910101<br>Name: Side chain hydrophobicity<br>Description: Scaled side chain hydrophobicity values (Black-Mould, 1991)","Scale: MANP780101<br>Name: Surrounding hydrophobicity<br>Description: Average surrounding hydrophobicity (Manavalan-Ponnuswamy, 1978)","Scale: PONP800101<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in folded form (Ponnuswamy et al., 1980)","Scale: PONP800102<br>Name: Surrounding hydrophobicity<br>Description: Average gain in surrounding hydrophobicity (Ponnuswamy et al., 1980)","Scale: PONP800103<br>Name: Surrounding hydrophobicity<br>Description: Average gain ratio in surrounding hydrophobicity (Ponnuswamy et al., 1980)","Scale: PONP800104<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in alpha-helix (Ponnuswamy et al., 1980)","Scale: PONP800105<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in beta-sheet (Ponnuswamy et al., 1980)","Scale: PONP800106<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in turn (Ponnuswamy et al., 1980)","Scale: PONP800108<br>Name: Surrounding hydrophobicity<br>Description: Average number of surrounding residues (Ponnuswamy et al., 1980)","Scale: WILM950104<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C18 with 0.1%TFA/2-PrOH/MeCN/H2O (Wilce et al., 1995)","Scale: NADH010107<br>Name: Hydrophobicity (based on self-information values)<br>Description: Hydropathy scale based on self-information values in the two-state model (50% accessibility) (Naderi-Manesh et al., 2001)","Scale: FASG760105<br>Name: pK-C<br>Description: pK-C (Fasman, 1976)","Scale: FAUJ880113<br>Name: pK-C<br>Description: pK-a(RCOOH) (Fauchere et al., 1988)","Scale: JOND750102<br>Name: pK-C<br>Description: pK (-COOH) (Jones, 1975)","Scale: FASG760104<br>Name: pK-N<br>Description: pK-N (Fasman, 1976)"],"marker":{"color":"rgba(25,88,195,1)","size":6,"line":{"color":"rgba(25,88,195,1)"}},"name":"Polarity","textfont":{"color":"rgba(25,88,195,1)"},"error_y":{"color":"rgba(25,88,195,1)"},"error_x":{"color":"rgba(25,88,195,1)"},"line":{"color":"rgba(25,88,195,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[8.7290129871178834,6.2546888154273708,6.2800522102286918,15.440523982984347,16.660832410975424,-0.25674399378316209,1.1183354376012511,2.461584528314066,1.7238761256795996,-0.38232187082218183,0.82847398922932447,0.80334498660110332,0.41992851365108214,0.1423104659597855,1.3994380100353754,-0.6479871873637979,1.2494601828681231,3.7446957235415721,6.2540408461079116,-5.7331716313491645,-0.60040981268579707,-3.1994242918314559,2.7367756288313245,-3.4090826391797968,-5.7403904442709885,1.7642095000470925,21.958410647558626,19.626748588346654,18.719113786231915,19.495346493269132,-3.2774121969338741],"y":[4.5449257025458207,3.6810212554724404,3.7288377215004638,-5.7434360600473404,-4.9639153649951373,-16.233516152192465,-16.784829110558395,-15.828262237201562,-17.078158498131632,-17.196081969845444,-17.168430376576755,-18.349840195295187,-16.573579738496651,-18.935398243867862,-18.65437794347379,-18.740846955665479,-14.73226172967199,-19.684851061139582,-21.118948357669797,13.99198462139549,11.866527221112127,13.018503575822541,11.280079198605875,19.681318913848791,19.684032654827469,8.174891376412118,-1.051420470322695,-1.76047597376304,-3.0120871546236545,-2.0063452131416177,8.6580324360450085],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: ROSM880103<br>Name: Loss of hydropathy by α-helix formation<br>Description: Loss of Side chain hydropathy by helix formation (Roseman, 1988)","Scale: ANDN920101<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: alpha-CH chemical shifts (Andersen et al., 1992)","Scale: BUNA790102<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: alpha-CH chemical shifts (Bundi-Wuthrich, 1979)","Scale: BUNA790103<br>Name: Spin-spin coupling<br>Description: Spin-spin coupling constants 3JHalpha-NH (Bundi-Wuthrich, 1979)","Scale: BUNA790101<br>Name: α-NH chemical shifts (backbone-dynamics)<br>Description: alpha-NH chemical shifts (Bundi-Wuthrich, 1979)","Scale: BHAR880101<br>Name: Flexibility<br>Description: Average flexibility indices (Bhaskaran-Ponnuswamy, 1988)","Scale: PARS000101<br>Name: Flexibility<br>Description: p-Values of mesophilic proteins based on the distributions of B values (Parthasarathy-Murthy, 2000)","Scale: PARS000102<br>Name: Flexibility<br>Description: p-Values of thermophilic proteins based on the distributions of B values (Parthasarathy-Murthy, 2000)","Scale: VINM940101<br>Name: Flexibility<br>Description: Normalized flexibility parameters (B-values), average (Vihinen et al., 1994)","Scale: KARP850101<br>Name: Flexibility (0 rigid neighbors)<br>Description: Flexibility parameter for no rigid neighbors (Karplus-Schulz, 1985)","Scale: VINM940102<br>Name: Flexibility (0 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by none rigid neighbours (Vihinen et al., 1994)","Scale: KARP850102<br>Name: Flexibility (1 rigid neighbors)<br>Description: Flexibility parameter for one rigid neighbor (Karplus-Schulz, 1985)","Scale: VINM940103<br>Name: Flexibility (1 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by one rigid neighbours (Vihinen et al., 1994)","Scale: KRIW710101<br>Name: Side chain interaction<br>Description: Side chain interaction parameter (Krigbaum-Rubin, 1971)","Scale: KRIW790101<br>Name: Side chain interaction<br>Description: Side chain interaction parameter (Krigbaum-Komoriya, 1979)","Scale: OOBM850105<br>Name: Side chain interaction<br>Description: Optimized side chain interaction parameter (Oobatake et al., 1985)","Scale: KARP850103<br>Name: Flexibility (2 rigid neighbors)<br>Description: Flexibility parameter for two rigid neighbors (Karplus-Schulz, 1985)","Scale: KRIW790102<br>Name: Flexibility (2 rigid neighbors)<br>Description: Fraction of site occupied by water (Krigbaum-Komoriya, 1979)","Scale: VINM940104<br>Name: Flexibility (2 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by two rigid neighbours (Vihinen et al., 1994)","Scale: BASU050102<br>Name: Interactivity<br>Description: Interactivity scale obtained by maximizing the mean of correlation Coefficient over single-domain globular proteins (Bastolla et al., 2005)","Scale: TAKK010101<br>Name: Stability<br>Description: Side-chain contribution to protein stability (kJ/mol) (Takano-Yutani, 2001)","Scale: ZHOH040101<br>Name: Stability<br>Description: The stability scale from the knowledge-based atom-atom potential (Zhou-Zhou, 2004)","Scale: ZHOH040102<br>Name: Stability<br>Description: The relative stability scale extracted from mutation experiments (Zhou-Zhou, 2004)","Scale: OOBM850101<br>Name: Stability (extended-coil)<br>Description: Optimized beta-structure-coil equilibrium constant (Oobatake et al., 1985)","Scale: PTIO830102<br>Name: Stability (extended-coil)<br>Description: Beta-coil equilibrium constant (Ptitsyn-Finkelstein, 1983)","Scale: FAUJ880107<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: N.m.r. chemical shift of alpha-carbon (Fauchere et al., 1988)","Scale: GEOR030108<br>Name: Linker (helical)<br>Description: Linker propensity from helical (annotated by DSSP) dataset (George-Heringa, 2003)","Scale: FINA770101<br>Name: Stability (helix-coil)<br>Description: Helix-coil equilibrium constant (Finkelstein-Ptitsyn, 1977)","Scale: PTIO830101<br>Name: Stability (helix-coil)<br>Description: Helix-coil equilibrium constant (Ptitsyn-Finkelstein, 1983)","Scale: SUEM840101<br>Name: Stability (helix-coil)<br>Description: Zimm-Bragg parameter s at 20 C (Sueki et al., 1984)","Scale: SUEM840102<br>Name: Stability (extended-coil)<br>Description: Zimm-Bragg parameter sigma x 1.0E4 (Sueki et al., 1984)"],"marker":{"color":"rgba(128,0,211,1)","size":6,"line":{"color":"rgba(128,0,211,1)"}},"name":"Structure-Activity","textfont":{"color":"rgba(128,0,211,1)"},"error_y":{"color":"rgba(128,0,211,1)"},"error_x":{"color":"rgba(128,0,211,1)"},"line":{"color":"rgba(128,0,211,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[-3.4761874258925398,-3.5087741117521722,-3.6541942906242193,16.961069538621128,-1.2964371728514346,-15.513608936905355,9.6596666790659924,6.8589203285893676,-6.9709494895567516,-9.1063319195029102,0.79698244438880994,0.64529085265678154,-2.5934835828888954,13.477025754219909,-5.9488168547375624,16.467509606403969,17.293096486067054],"y":[-19.083029972931875,-19.079142000664419,-16.869273781619718,-0.45296640895809276,13.219279671037123,13.106512623571078,6.7377533505474103,11.081681480162146,-1.8882646518425703,-3.1326304904940745,-20.598610890232337,-20.828066709720087,12.2130096040837,-9.6103829223042148,7.9108855300376435,-5.5480512046471882,-4.8273650652809774],"type":"scatter","mode":"markers","hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Scale: DAYM780201<br>Name: Mutability<br>Description: Relative mutability (Dayhoff et al., 1978b)","Scale: JOND920102<br>Name: Mutability<br>Description: Relative mutability (Jones et al., 1992)","Scale: WEBA780101<br>Name: RF value<br>Description: RF value in high salt chromatography (Weber-Lacey, 1978)","Scale: SNEP660101<br>Name: Principal Component 1 (Sneath)<br>Description: Principal component I (Sneath, 1966)","Scale: ZIMJ680105<br>Name: Principal Component 1 (Zimmerman)<br>Description: RF rank (Zimmerman et al., 1968)","Scale: SNEP660102<br>Name: Principal Component 2 (Sneath)<br>Description: Principal component II (Sneath, 1966)","Scale: WOLS870102<br>Name: Principal Component 2 (Wold)<br>Description: Principal property value z2 (Wold et al., 1987)","Scale: SNEP660103<br>Name: Principal Component 3 (Sneath)<br>Description: Principal component III (Sneath, 1966)","Scale: WOLS870103<br>Name: Principal Component 3 (Wold)<br>Description: Principal property value z3 (Wold et al., 1987)","Scale: SNEP660104<br>Name: Principal Component 4 (Sneath)<br>Description: Principal component IV (Sneath, 1966)","Scale: PARJ860101<br>Name: HPLC parameter<br>Description: HPLC parameter (Parker et al., 1986)","Scale: WOLS870101<br>Name: Principal Component 1 (Wold)<br>Description: Principal property value z1 (Wold et al., 1987)","Scale: VENT840101<br>Name: Bitterness<br>Description: Bitterness (Venanzi, 1984)","Scale: KHAG800101<br>Name: Kerr-constant<br>Description: The Kerr-constant increments (Khanarian-Moore, 1980)","Scale: FASG760102<br>Name: Melting point<br>Description: Melting Point (Fasman, 1976)","Scale: FASG760103<br>Name: Optical rotation<br>Description: Optical rotation (Fasman, 1976)","Scale: ONEK900101<br>Name: ΔG values in peptides<br>Description: Delta G values for the peptides extrapolated to 0 M urea (O'Neil-DeGrado, 1990)"],"marker":{"color":"rgba(153,153,153,1)","size":6,"line":{"color":"rgba(153,153,153,1)"}},"name":"Others","textfont":{"color":"rgba(153,153,153,1)"},"error_y":{"color":"rgba(153,153,153,1)"},"error_x":{"color":"rgba(153,153,153,1)"},"line":{"color":"rgba(153,153,153,1)"},"xaxis":"x","yaxis":"y","frame":null},{"x":[2.8468087795665937,-17.130287131115871,15.593318088084208,-7.5064340479732117,8.0197788749005205,-12.73259496118254,12.860375097287918,-7.0454088111013666,8.4051368843935546,5.8718581214136814,6.7368576240057285,-6.7723373775604019],"y":[-21.795679183845309,-2.4307194038423003,-4.3432997092592709,-2.2530680533885636,-9.5131570074694665,-4.8111263076568225,-3.3291182433039253,3.2049701255228999,-11.367121697926617,0.36655065197415654,-3.7935050496259279,6.5989002226226212],"type":"scatter","mode":"markers+text","marker":{"color":"#ff295f","symbol":"square","size":10,"line":{"color":"rgba(188,189,34,1)"}},"text":["my.PC1","my.PC2","my.PC3","my.PC4","my.PC5","my.PC6","my.PC7","my.PC8","my.PC9","my.PC10","my.PC11","my.PC12"],"textposition":["top center","top center","top center","top center","top center","top center","top center","top center","top center","top center","top center","top center"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text"],"hovertext":["Feature-space axis PC1 (my.PC1)","Feature-space axis PC2 (my.PC2)","Feature-space axis PC3 (my.PC3)","Feature-space axis PC4 (my.PC4)","Feature-space axis PC5 (my.PC5)","Feature-space axis PC6 (my.PC6)","Feature-space axis PC7 (my.PC7)","Feature-space axis PC8 (my.PC8)","Feature-space axis PC9 (my.PC9)","Feature-space axis PC10 (my.PC10)","Feature-space axis PC11 (my.PC11)","Feature-space axis PC12 (my.PC12)"],"name":"Feature-space PCs","showlegend":true,"error_y":{"color":"rgba(188,189,34,1)"},"error_x":{"color":"rgba(188,189,34,1)"},"line":{"color":"rgba(188,189,34,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p><strong>Figure:</strong> t-SNE of amino acid scales and Feature Space PCs.</p>
</div>
</div>
<p>This plot is interactive: hovering over data points show what the point contains. It uses color to identify the different categories of our scales, and you can zoom in and out for details.</p>
<p><strong>Task: interpret this plot.</strong></p>
</section>
</section>
<section id="joint-t-sne-embedding-of-pcs-and-scales-in-3d" class="level3" data-number="7.3.3">
<h3 data-number="7.3.3" class="anchored" data-anchor-id="joint-t-sne-embedding-of-pcs-and-scales-in-3d"><span class="header-section-number">7.3.3</span> Joint t-SNE embedding of PCs and scales in 3d</h3>
<p>If a 2D embedding of data gives us a good understanding, 3D should be even better.</p>
<p>Well, maybe. That really depends on the data and the question. But see for yourself. Once we have gone though the trouble of our 2d-plot, moving to 3d is just more of the same…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run t-SNE on the prepared data</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">13</span>)  <span class="co"># reproducible layout</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>tsneRes3d <span class="ot">&lt;-</span> Rtsne<span class="sc">::</span><span class="fu">Rtsne</span>(<span class="fu">as.matrix</span>(jointMat),</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">dims =</span> <span class="dv">3</span>,             <span class="co"># 3 dimensions</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">perplexity =</span> <span class="dv">30</span>,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">max_iter =</span> <span class="dv">3000</span>)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract t-SNE coordinates</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>tsneCoords3d <span class="ot">&lt;-</span> tsneRes3d<span class="sc">$</span>Y</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tsneCoords3d) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(jointMat)</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Scales</span></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a>scaleDF3d <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">x         =</span> tsneCoords3d[myScID, <span class="dv">1</span>],</span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">y         =</span> tsneCoords3d[myScID, <span class="dv">2</span>],</span>
<span id="cb67-18"><a href="#cb67-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">z         =</span> tsneCoords3d[myScID, <span class="dv">3</span>],</span>
<span id="cb67-19"><a href="#cb67-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">category  =</span> aaOntology[myScID, <span class="st">"category"</span>],</span>
<span id="cb67-20"><a href="#cb67-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">hover     =</span> hoverText[myScID]</span>
<span id="cb67-21"><a href="#cb67-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-22"><a href="#cb67-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-23"><a href="#cb67-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure category factor order matches catColors</span></span>
<span id="cb67-24"><a href="#cb67-24" aria-hidden="true" tabindex="-1"></a>scaleDF3d<span class="sc">$</span>category <span class="ot">&lt;-</span> <span class="fu">factor</span>(scaleDF3d<span class="sc">$</span>category, <span class="at">levels =</span> <span class="fu">names</span>(catColors))</span>
<span id="cb67-25"><a href="#cb67-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-26"><a href="#cb67-26" aria-hidden="true" tabindex="-1"></a><span class="co"># PCs</span></span>
<span id="cb67-27"><a href="#cb67-27" aria-hidden="true" tabindex="-1"></a>pcDF3d <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb67-28"><a href="#cb67-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">x     =</span> tsneCoords3d[myPcID, <span class="dv">1</span>],</span>
<span id="cb67-29"><a href="#cb67-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">y     =</span> tsneCoords3d[myPcID, <span class="dv">2</span>],</span>
<span id="cb67-30"><a href="#cb67-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">z     =</span> tsneCoords3d[myPcID, <span class="dv">3</span>],</span>
<span id="cb67-31"><a href="#cb67-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">myID  =</span> <span class="fu">seq_along</span>(myPcID),       <span class="co"># 1, 2, … 12 for labels</span></span>
<span id="cb67-32"><a href="#cb67-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">hover =</span> hoverText[myPcID]</span>
<span id="cb67-33"><a href="#cb67-33" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-34"><a href="#cb67-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-35"><a href="#cb67-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-36"><a href="#cb67-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Create 3D plot with explicit category colors </span></span>
<span id="cb67-37"><a href="#cb67-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-38"><a href="#cb67-38" aria-hidden="true" tabindex="-1"></a>fig3d <span class="ot">&lt;-</span> plotly<span class="sc">::</span><span class="fu">plot_ly</span>(<span class="at">type =</span> <span class="st">"scatter3d"</span>, <span class="at">mode =</span> <span class="st">"markers"</span>)</span>
<span id="cb67-39"><a href="#cb67-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-40"><a href="#cb67-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over categories explicitly</span></span>
<span id="cb67-41"><a href="#cb67-41" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (cat <span class="cf">in</span> <span class="fu">names</span>(catColors)) {</span>
<span id="cb67-42"><a href="#cb67-42" aria-hidden="true" tabindex="-1"></a>  df_cat <span class="ot">&lt;-</span> <span class="fu">subset</span>(scaleDF3d, category <span class="sc">==</span> cat)</span>
<span id="cb67-43"><a href="#cb67-43" aria-hidden="true" tabindex="-1"></a>  fig3d <span class="ot">&lt;-</span> fig3d <span class="sc">|&gt;</span></span>
<span id="cb67-44"><a href="#cb67-44" aria-hidden="true" tabindex="-1"></a>    plotly<span class="sc">::</span><span class="fu">add_trace</span>(</span>
<span id="cb67-45"><a href="#cb67-45" aria-hidden="true" tabindex="-1"></a>      <span class="at">data =</span> df_cat,</span>
<span id="cb67-46"><a href="#cb67-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="sc">~</span>x, <span class="at">y =</span> <span class="sc">~</span>y, <span class="at">z =</span> <span class="sc">~</span>z,</span>
<span id="cb67-47"><a href="#cb67-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> cat,                      <span class="co"># legend label</span></span>
<span id="cb67-48"><a href="#cb67-48" aria-hidden="true" tabindex="-1"></a>      <span class="at">text =</span> <span class="sc">~</span>hover,</span>
<span id="cb67-49"><a href="#cb67-49" aria-hidden="true" tabindex="-1"></a>      <span class="at">hoverinfo =</span> <span class="st">"text"</span>,</span>
<span id="cb67-50"><a href="#cb67-50" aria-hidden="true" tabindex="-1"></a>      <span class="at">marker =</span> <span class="fu">list</span>(</span>
<span id="cb67-51"><a href="#cb67-51" aria-hidden="true" tabindex="-1"></a>        <span class="at">size =</span> <span class="dv">4</span>,</span>
<span id="cb67-52"><a href="#cb67-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">opacity =</span> <span class="fl">0.7</span>,</span>
<span id="cb67-53"><a href="#cb67-53" aria-hidden="true" tabindex="-1"></a>        <span class="at">color =</span> catColors[cat],        <span class="co"># assign fixed color</span></span>
<span id="cb67-54"><a href="#cb67-54" aria-hidden="true" tabindex="-1"></a>        <span class="at">symbol =</span> <span class="st">"circle"</span></span>
<span id="cb67-55"><a href="#cb67-55" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb67-56"><a href="#cb67-56" aria-hidden="true" tabindex="-1"></a>      <span class="at">showlegend =</span> <span class="cn">TRUE</span></span>
<span id="cb67-57"><a href="#cb67-57" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb67-58"><a href="#cb67-58" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb67-59"><a href="#cb67-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-60"><a href="#cb67-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PCs (squares with labels 1..12)</span></span>
<span id="cb67-61"><a href="#cb67-61" aria-hidden="true" tabindex="-1"></a>fig3d <span class="ot">&lt;-</span> fig3d <span class="sc">|&gt;</span></span>
<span id="cb67-62"><a href="#cb67-62" aria-hidden="true" tabindex="-1"></a>  plotly<span class="sc">::</span><span class="fu">add_trace</span>(</span>
<span id="cb67-63"><a href="#cb67-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> pcDF3d,</span>
<span id="cb67-64"><a href="#cb67-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="sc">~</span>x, <span class="at">y =</span> <span class="sc">~</span>y, <span class="at">z =</span> <span class="sc">~</span>z,</span>
<span id="cb67-65"><a href="#cb67-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="sc">~</span>hover,</span>
<span id="cb67-66"><a href="#cb67-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">hoverinfo =</span> <span class="st">"text"</span>,</span>
<span id="cb67-67"><a href="#cb67-67" aria-hidden="true" tabindex="-1"></a>    <span class="at">marker =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">6</span>, <span class="at">color =</span> <span class="st">"#ff295f"</span>, <span class="at">symbol =</span> <span class="st">"square"</span>),</span>
<span id="cb67-68"><a href="#cb67-68" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Feature-space PCs"</span>,   <span class="co"># gives a single legend entry for PCs</span></span>
<span id="cb67-69"><a href="#cb67-69" aria-hidden="true" tabindex="-1"></a>    <span class="at">showlegend =</span> <span class="cn">TRUE</span></span>
<span id="cb67-70"><a href="#cb67-70" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb67-71"><a href="#cb67-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-72"><a href="#cb67-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Add PC text labels (numbers 1..12)</span></span>
<span id="cb67-73"><a href="#cb67-73" aria-hidden="true" tabindex="-1"></a>fig3d <span class="ot">&lt;-</span> fig3d <span class="sc">|&gt;</span></span>
<span id="cb67-74"><a href="#cb67-74" aria-hidden="true" tabindex="-1"></a>  plotly<span class="sc">::</span><span class="fu">add_text</span>(</span>
<span id="cb67-75"><a href="#cb67-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> pcDF3d,</span>
<span id="cb67-76"><a href="#cb67-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="sc">~</span>x, <span class="at">y =</span> <span class="sc">~</span>y, <span class="at">z =</span> <span class="sc">~</span>z,</span>
<span id="cb67-77"><a href="#cb67-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">text =</span> <span class="sc">~</span>myID,</span>
<span id="cb67-78"><a href="#cb67-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">textposition =</span> <span class="st">"top center"</span>,</span>
<span id="cb67-79"><a href="#cb67-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">textfont =</span> <span class="fu">list</span>(<span class="at">color =</span> <span class="st">"#ff295f"</span>, <span class="at">size =</span> <span class="dv">12</span>),</span>
<span id="cb67-80"><a href="#cb67-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">showlegend =</span> <span class="cn">FALSE</span></span>
<span id="cb67-81"><a href="#cb67-81" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb67-82"><a href="#cb67-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-83"><a href="#cb67-83" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the plot</span></span>
<span id="cb67-84"><a href="#cb67-84" aria-hidden="true" tabindex="-1"></a>fig3d</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="plotly html-widget html-fill-item" id="htmlwidget-4be102776a27018dbc54" style="width:100%;height:768px;"></div>
<script type="application/json" data-for="htmlwidget-4be102776a27018dbc54">{"x":{"visdat":{"106b510a21fcc":["function () ","plotlyVisDat"],"106b53d5929c":["function () ","data"],"106b53a8c41bb":["function () ","data"],"106b54e2f7a04":["function () ","data"],"106b5ef3c4b6":["function () ","data"],"106b525f79065":["function () ","data"],"106b52028ea5c":["function () ","data"],"106b55e324e82":["function () ","data"],"106b538c0691e":["function () ","data"],"106b560354dad":["function () ","data"],"106b54b7eca33":["function () ","data"]},"cur_data":"106b54b7eca33","attrs":{"106b510a21fcc":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d"},"106b53d5929c":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Energy","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#f2003c","symbol":"circle"},"showlegend":true,"inherit":true},"106b53a8c41bb":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Shape","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#F0A200","symbol":"circle"},"showlegend":true,"inherit":true},"106b54e2f7a04":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"ASA/Volume","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#f0ea00","symbol":"circle"},"showlegend":true,"inherit":true},"106b5ef3c4b6":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Conformation","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#62C923","symbol":"circle"},"showlegend":true,"inherit":true},"106b525f79065":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Composition","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#0A9A9B","symbol":"circle"},"showlegend":true,"inherit":true},"106b52028ea5c":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Polarity","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#1958C3","symbol":"circle"},"showlegend":true,"inherit":true},"106b55e324e82":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Structure-Activity","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#8000D3","symbol":"circle"},"showlegend":true,"inherit":true},"106b538c0691e":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"name":"Others","text":{},"hoverinfo":"text","marker":{"size":4,"opacity":0.69999999999999996,"color":"#999999","symbol":"circle"},"showlegend":true,"inherit":true},"106b560354dad":{"mode":"markers","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"text":{},"hoverinfo":"text","marker":{"size":6,"color":"#ff295f","symbol":"square"},"name":"Feature-space PCs","showlegend":true,"inherit":true},"106b54b7eca33":{"mode":"text","alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","x":{},"y":{},"z":{},"text":{},"textposition":"top center","textfont":{"color":"#ff295f","size":12},"showlegend":false,"inherit":true}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"scene":{"xaxis":{"title":"x"},"yaxis":{"title":"y"},"zaxis":{"title":"z"}},"hovermode":"closest","showlegend":true},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"mode":"markers","type":"scatter3d","marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"error_y":{"color":"rgba(31,119,180,1)"},"error_x":{"color":"rgba(31,119,180,1)"},"line":{"color":"rgba(31,119,180,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[9.5127323380050406,16.182043707776312,3.9560027136066567,3.355244313957841,9.026452094949704,-4.7695567818278715,5.6116883744466257,5.6116283429248108,16.507470361774104,17.213377880565499,13.396334387445398,-1.6093075666315046,-1.4699924705444241,-4.1873194549995372,-0.89712453151642146,-1.0488620977822336,-10.987803450302598,-11.499309348819269,-12.648936774802339,-12.638094027760578,-10.251267761177086,-12.938104834140868,-13.736929005204567,-11.937401846406281,4.3211963508179743,9.2561726390895931,3.3011821713954235,-15.616449650947557,-14.09462967904998,-14.39560124357549,-14.599485061009894,-3.2640904402910502,-12.956272432584925,-13.242786946495725,8.6536352382308692,-12.682909411022997],"y":[13.722300128276483,-10.830872148968567,10.532945270048137,15.72669509397581,13.99273447563697,-8.7039058840609513,-7.6948374215266639,-7.6945889467742896,-10.015924652016363,-9.0794416761907026,-6.4139566350808686,-2.4994704589694847,-2.5365069082050913,2.1244156727043157,0.57866174689188377,0.334331103516608,-6.6668343033694342,-6.0490649389506466,-5.442171892747619,-5.4529675409159077,-12.409151771025122,-8.4294429294741846,-8.8685630076208515,-9.2131364785441203,22.587233892283734,13.64082919803797,21.714271213335376,9.9535929184691678,8.9167748881292237,8.4865987520404644,8.2153851104917397,-5.0037366332412949,-12.90336251042636,-3.1691413839438929,-13.379386034709196,-14.449981295013517],"z":[3.4803780994128952,-2.1436337952466671,13.802291908051274,12.636132179170684,4.7686294444175603,11.137713278928731,6.1236607324807109,6.1238704335582339,-4.2516132722223157,-4.4198440963342058,-3.4230472022790241,18.493779745233116,18.704745354187537,15.412066041537667,17.969152889673865,18.227186427941415,-16.494561811624365,-16.378969442960674,-15.566542297106659,-15.576080123833091,-16.485285230502061,-16.141607098822533,-16.260157353133302,-16.468358308151998,13.033764612028801,3.8076049084427575,13.197650487184188,2.0959934003423455,3.6434391227989171,3.4323472511462665,2.9099400561321436,-16.393373919464779,-17.030340855988044,18.191046594576473,-7.4267784423511136,-13.1172840854393],"name":"Energy","text":["Scale: KLEP840101<br>Name: Charge<br>Description: Net charge (Klein et al., 1984)","Scale: CHAM830108<br>Name: Charge (donor)<br>Description: A parameter of charge transfer donor capability (Charton-Charton, 1983)","Scale: FAUJ880112<br>Name: Charge (negative)<br>Description: Negative charge (Fauchere et al., 1988)","Scale: CHAM830107<br>Name: Charge (transfer)<br>Description: A parameter of charge transfer capability (Charton-Charton, 1983)","Scale: FAUJ880111<br>Name: Charge (positive)<br>Description: Positive charge (Fauchere et al., 1988)","Scale: FAUJ880108<br>Name: Electrical Effect<br>Description: Localized Electrical Effect (Fauchere et al., 1988)","Scale: COSI940101<br>Name: Electron-ion interaction potential<br>Description: Electron-ion interaction potential values (Cosic, 1994)","Scale: VELV850101<br>Name: Electron-ion interaction potential<br>Description: Electron-ion interaction potential (Veljkovic et al., 1985)","Scale: HUTJ700102<br>Name: Entropy<br>Description: Absolute entropy (Hutchens, 1970)","Scale: HUTJ700103<br>Name: Entropy<br>Description: Entropy of formation (Hutchens, 1970)","Scale: OOBM770102<br>Name: Non-bonded energy per atom<br>Description: Short and medium range non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: MUNV940101<br>Name: Free energy (α-helix)<br>Description: Free energy in alpha-helical conformation (Munoz-Serrano, 1994)","Scale: MUNV940102<br>Name: Free energy (α-helix)<br>Description: Free energy in alpha-helical region (Munoz-Serrano, 1994)","Scale: MUNV940103<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand conformation (Munoz-Serrano, 1994)","Scale: MUNV940104<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand region (Munoz-Serrano, 1994)","Scale: MUNV940105<br>Name: Free energy (β-strand)<br>Description: Free energy in beta-strand region (Munoz-Serrano, 1994)","Scale: YUTK870101<br>Name: Free energy (unfolding)<br>Description: Unfolding Gibbs energy in water, pH7.0 (Yutani et al., 1987)","Scale: YUTK870102<br>Name: Free energy (unfolding)<br>Description: Unfolding Gibbs energy in water, pH9.0 (Yutani et al., 1987)","Scale: YUTK870103<br>Name: Free energy (unfolding)<br>Description: Activation Gibbs energy of unfolding, pH7.0 (Yutani et al., 1987)","Scale: YUTK870104<br>Name: Free energy (unfolding)<br>Description: Activation Gibbs energy of unfolding, pH9.0 (Yutani et al., 1987)","Scale: JANJ790102<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy (Janin, 1979)","Scale: RADA880104<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy from chx to oct (Radzicka-Wolfenden, 1988)","Scale: RADA880105<br>Name: Transfer free energy (TFE) to inside<br>Description: Transfer free energy from vap to oct (Radzicka-Wolfenden, 1988)","Scale: RADA880107<br>Name: Transfer free energy (TFE) to inside<br>Description: Energy transfer from out to in (95% buried) (Radzicka-Wolfenden, 1988)","Scale: EISD860102<br>Name: Atom-based hydrophobic moment<br>Description: Atom-based hydrophobic moment (Eisenberg-McLachlan, 1986)","Scale: ZIMJ680104<br>Name: Isoelectric point<br>Description: Isoelectric point (Zimmerman et al., 1968)","Scale: FAUJ880109<br>Name: Number hydrogen bond donors<br>Description: Number of hydrogen bond donors (Fauchere et al., 1988)","Scale: VASM830102<br>Name: Free energy (Extended)<br>Description: Relative population of conformational state C (Vasquez et al., 1983)","Scale: OOBM770104<br>Name: Non-bonded energy per residue<br>Description: Average non-bonded energy per residue (Oobatake-Ooi, 1977)","Scale: OOBM770105<br>Name: Non-bonded energy per residue<br>Description: Short and medium range non-bonded energy per residue (Oobatake-Ooi, 1977)","Scale: RADA880103<br>Name: TFE to inside<br>Description: Transfer free energy from vap to chx (Radzicka-Wolfenden, 1988)","Scale: WERD780103<br>Name: Free energy (α-helix)<br>Description: Free energy change of alpha(Ri) to alpha(Rh) (Wertz-Scheraga, 1978)","Scale: WERD780104<br>Name: Free energy (α-helix)<br>Description: Free energy change of epsilon(i) to alpha(Rh) (Wertz-Scheraga, 1978)","Scale: WERD780102<br>Name: Free energy change<br>Description: Free energy change of epsilon(i) to epsilon(ex) (Wertz-Scheraga, 1978)","Scale: HUTJ700101<br>Name: Heat capactiy<br>Description: Heat capacity (Hutchens, 1970)","Scale: OOBM850103<br>Name: Transfer free energy (TFE)<br>Description: Optimized transfer energy parameter (Oobatake et al., 1985)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#f2003c","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(255,127,14,1)"}},"showlegend":true,"error_y":{"color":"rgba(255,127,14,1)"},"error_x":{"color":"rgba(255,127,14,1)"},"line":{"color":"rgba(255,127,14,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[12.099030081385441,12.146039062860158,12.096284728783854,12.100721478346848,12.140835429356738,-15.455413112025678,0.57479958225232286,-0.76724599511326963,-6.0075042451949496,-7.5856101802118179,-8.1138983017320072,-6.2043888380886418,-6.7650050139276079,-14.379143926625657,-14.273946653949395,-14.368421371873195,12.278102426952028,12.718892077137768,15.233517655969166,12.970105037338971,15.61785964232585,11.457543891157133,14.70981154350296,11.180684914061699,11.891816482976825,11.871660806730471,10.869985118832606,10.781544560296821,16.177835737935283,14.80903809400418,14.190084165550642,13.998079430464747,14.739784211982046,16.221101592992571,13.476731533680249,6.0166922538305965,6.5801842703212055,18.89587320466541,18.613293503622877,18.520667023767079,6.4676776657339365,-3.2169105216234737,-14.94072775896575,1.8593411169136767,-13.113630246163194],"y":[-9.94183156638403,-12.024680737100958,-10.058694752160042,-12.120756242248577,-10.738105038628444,9.6178113605218414,-2.6361744213857743,-1.9875263063304227,0.23278919398383324,18.973121749255935,18.700289197741011,20.573300377636325,20.240968707356107,1.6643776924007485,0.51995515454233121,1.4453822374245962,-9.3547488694420728,-8.37254761740253,-9.3440641518647958,-8.3132426680377609,-8.9503054431017954,-8.0183414631272267,-10.213737576069972,-9.3665150492887523,-10.650336768748701,-8.4695262106944185,-8.6613306640330858,-8.3222932543351789,-3.8432259100966606,-7.2274450327495909,-7.8080796805460508,-8.8447991315992134,-8.8890200228101932,-8.2836552011770053,-11.146302164844652,-18.275235932593521,-15.33042195756142,4.2626366395899344,5.0214450212305337,5.3052978053371538,-17.537537802037022,-10.4142794430699,8.1656362332342809,3.1302303784790109,-4.3169816207124363],"z":[0.7805371099781413,0.24689656455589298,0.20345409314211405,0.46025763159949423,0.83579953497822379,2.2616117798473945,17.217316867979847,19.138504837952848,11.819199622797687,13.212470096158658,13.526338381833222,13.369017922736642,13.543582875951859,8.5478903152097061,10.234599582816523,8.8162750039124642,-2.4661890628261953,-2.4022938040338873,-2.8352610367540327,-3.5137425637439086,-2.3622837937894001,-1.9734770223804747,-2.2558428975558633,-2.376769780623933,-2.6737437386316851,-3.4299670399325497,-3.6389801498687846,-2.9464202899888421,-4.3169414811555482,-2.1437673958548782,-3.9410880611413157,-3.067070355851893,-3.8801314305239538,-4.0935726083961432,-2.6206724371664505,-15.940475768332554,-14.812871117696988,-4.450530290609656,-4.3166776163335774,-4.1923988998629058,-15.578783139356613,10.812048263669695,4.4918539336117833,20.493750280593929,13.409389085783797],"name":"Shape","text":["Scale: KARS160111<br>Name: Eigenvalue (average)<br>Description: Average eigenvalue of the Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160121<br>Name: Eigenvalue (average)<br>Description: Weighted average eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: KARS160109<br>Name: Eigenvalue (maximum)<br>Description: Maximum eigenvalue of the weighted Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160119<br>Name: Eigenvalue (maximum)<br>Description: Weighted maximum eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: KARS160108<br>Name: Graph (weighted degree)<br>Description: Average weighted degree (total degree, divided by the number of vertices) (Kakraba-Knisley, 2016)","Scale: KARS160112<br>Name: Eigenvalue (2. smallest)<br>Description: Second smallest eigenvalue of the Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160122<br>Name: Eigenvalue (2. smallest)<br>Description: Weighted second smallest eigenvalue of the weighted Laplacian matrix (Kakraba-Knisley, 2016)","Scale: RACS820114<br>Name: Side chain angle (Theta)<br>Description: Value of theta(i-1) (Rackovsky-Scheraga, 1982)","Scale: PRAM820101<br>Name: Intercept in regression<br>Description: Intercept in regression analysis (Prabhakaran-Ponnuswamy, 1982)","Scale: MEIH800101<br>Name: Reduced distance (C-α)<br>Description: Average reduced distance for C-alpha (Meirovitch et al., 1980)","Scale: RACS770101<br>Name: Reduced distance (C-α)<br>Description: Average reduced distance for C-alpha (Rackovsky-Scheraga, 1977)","Scale: MEIH800102<br>Name: Reduced distance (side chain)<br>Description: Average reduced distance for side chain (Meirovitch et al., 1980)","Scale: RACS770102<br>Name: Reduced distance (side chain)<br>Description: Average reduced distance for side chain (Rackovsky-Scheraga, 1977)","Scale: PRAM820103<br>Name: Correlation coefficient in regression<br>Description: Correlation coefficient in regression analysis (Prabhakaran-Ponnuswamy, 1982)","Scale: LEVM760104<br>Name: Side chain angle (Phi)<br>Description: Side chain torsion angle phi(AAAR) (Levitt, 1976)","Scale: PRAM820102<br>Name: Slope in Regression<br>Description: Slope in Regression analysis x 1.0E1 (Prabhakaran-Ponnuswamy, 1982)","Scale: KARS160117<br>Name: Atomic number (total)<br>Description: Total weighted atomic number of the graph (obtained by summing all the atomic number of each of the vertices in the graph) (Kakraba-Knisley, 2016)","Scale: KARS160105<br>Name: Eccentricity (average)<br>Description: Average eccentricity (Kakraba-Knisley, 2016)","Scale: KARS160114<br>Name: Eccentricity (average)<br>Description: Average weighted eccentricity based on the atomic number (Kakraba-Knisley, 2016)","Scale: KARS160107<br>Name: Eccentricity (maximum)<br>Description: Diameter (maximum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160116<br>Name: Eccentricity (maximum)<br>Description: Weighted diameter based on the atomic number (maximum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160106<br>Name: Eccentricity (minimum)<br>Description: Radius (minimum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160115<br>Name: Eccentricity (minimum)<br>Description: Weighted radius based on the atomic number (minimum eccentricity) (Kakraba-Knisley, 2016)","Scale: KARS160104<br>Name: Graph (domination number)<br>Description: Weighted domination number (Kakraba-Knisley, 2016)","Scale: KARS160113<br>Name: Graph (domination number)<br>Description: Weighted domination number using the atomic number (Kakraba-Knisley, 2016)","Scale: KARS160101<br>Name: Graph (order)<br>Description: Number of vertices (order of the graph) (Kakraba-Knisley, 2016)","Scale: KARS160102<br>Name: Graph (size)<br>Description: Number of edges (size of the graph) (Kakraba-Knisley, 2016)","Scale: KARS160103<br>Name: Graph (weighted degree)<br>Description: Total weighted degree of the graph (obtained by adding all the weights of all the vertices) (Kakraba-Knisley, 2016)","Scale: CHAM830104<br>Name: n atoms in side chain (2+1)<br>Description: The number of atoms in the side chain labelled 2+1 (Charton-Charton, 1983)","Scale: CHAM830105<br>Name: n atoms in side chain (3+1)<br>Description: The number of atoms in the side chain labelled 3+1 (Charton-Charton, 1983)","Scale: CHAM830106<br>Name: n bonds in longest chain<br>Description: The number of bonds in the longest chain (Charton-Charton, 1983)","Scale: LEVM760102<br>Name: Side chain length<br>Description: Distance between C-alpha and centroid of side chain (Levitt, 1976)","Scale: LEVM760105<br>Name: Side chain length<br>Description: Radius of gyration of side chain (Levitt, 1976)","Scale: FAUJ880104<br>Name: Steric parameter<br>Description: STERIMOL length of the side chain (Fauchere et al., 1988)","Scale: FAUJ880106<br>Name: Steric parameter (maximum width R)<br>Description: STERIMOL maximum width of the side chain (Fauchere et al., 1988)","Scale: CHAM830103<br>Name: n atoms in side chain (1+1)<br>Description: The number of atoms in the side chain labelled 1+1 (Charton-Charton, 1983)","Scale: FAUJ880101<br>Name: Shape Index<br>Description: Graph shape index (Fauchere et al., 1988)","Scale: LEVM760103<br>Name: Side chain angle (theta)<br>Description: Side chain angle theta(AAR) (Levitt, 1976)","Scale: CHAM810101<br>Name: Steric parameter<br>Description: Steric parameter (Charton, 1981)","Scale: FAUJ880102<br>Name: Steric parameter<br>Description: Smoothed upsilon steric parameter (Fauchere et al., 1988)","Scale: FAUJ880105<br>Name: Steric parameter (minimum width R)<br>Description: STERIMOL minimum width of the side chain (Fauchere et al., 1988)","Scale: KARS160118<br>Name: Atomic number (average)<br>Description: Average weighted atomic number or degree based on atomic number in the graph (Kakraba-Knisley, 2016)","Scale: KARS160110<br>Name: Eigenvalue (minimum)<br>Description: Minimum eigenvalue of the weighted Laplacian matrix of the graph (Kakraba-Knisley, 2016)","Scale: KARS160120<br>Name: Eigenvalue (minimum)<br>Description: Weighted minimum eigenvalue based on the atomic numbers (Kakraba-Knisley, 2016)","Scale: RACS820113<br>Name: Side chain angle (theta)<br>Description: Value of theta(i) (Rackovsky-Scheraga, 1982)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#F0A200","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(44,160,44,1)"}},"showlegend":true,"error_y":{"color":"rgba(44,160,44,1)"},"error_x":{"color":"rgba(44,160,44,1)"},"line":{"color":"rgba(44,160,44,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[-0.94215704496312724,-5.6990927711281785,0.62697913568607821,-0.37774690780857662,-0.41834107603446968,-4.11964396965584,1.167622029189346,0.59258449109561029,-2.6957379563701345,2.7420069747806939,1.6876410519628169,-2.737781393231205,1.2987170605623077,-1.4504548245997753,1.8723370360249258,1.4583193874748628,-6.4962943221071754,-5.1294115338246034,-3.9483723154811732,-0.11248431513132824,2.2670320156320272,-5.0601518683767521,1.1214561187707734,-4.9559759476116216,-5.5386926011063027,-4.2268061166990325,-2.7243312993056068,-7.12503832528059,-11.088340478020275,-12.376981365220566,-10.225551732971983,-10.413763236454749,-11.355022300144906,-5.7340080838846479,-9.6244208768718753,0.071884745673550063,-0.17135288790307202,0.0094033150129381859,-2.2471359909029069,6.566395095303089,1.7909300386267228,-2.0180080552626478,-2.5985040739347296,-3.9827147582196964,6.1581297291816641,6.2063170903992129,-1.5413917684525913,13.117347364471083,14.338462994475941,14.147765988890143,6.5751163273525695,12.715899278290104,12.735475144453735,14.197836405481837,11.973614990819753,11.970092164255087,12.727289590169073,12.823429522177245,11.76756960224645,6.1175387542263451,12.882075729463629,11.687390748480521,12.125661041745374,12.798158483691521],"y":[21.484153765322791,18.132736481211591,22.424955027966686,22.572666055033661,21.039086566222341,19.755902902407069,23.032555623346276,21.613732652578491,20.271017029268318,19.868894747979247,19.460460245917549,18.588193892918884,19.809607760004322,19.699533142112198,20.181495182024612,20.919455119670658,17.730901450626618,19.683542303149768,19.557625805858393,22.807507520477024,22.058559440345185,22.116527616846081,17.49715908016756,-12.959359203301664,-12.141729665462051,-13.507346690430293,-12.393296475864119,-9.478614831516083,-11.56365061622761,-12.092473786597608,-12.698262735996167,-13.84334181181198,-11.120652258630074,-15.542656252327143,-12.330282063781636,26.527692644236772,26.151275249310402,26.408848043207502,-16.070980103005407,-14.67117575701533,-12.000444909958107,-14.590406461876105,-14.401416778908946,-13.812418539675832,-12.291159519585337,-12.353904238285939,-10.611444577049655,-9.2529144697278287,-9.401591887891815,-8.0090840638583565,-12.124395335605399,-10.512839664723652,-9.9311831371680377,-11.353774928386938,-9.6356241509445919,-10.894897856730351,-9.4635246643089879,-11.33691203746992,-10.025528573365349,-12.089780524637764,-10.80158330134743,-10.061065262931089,-9.2442683245174582,-9.5013886939017187],"z":[13.956561616700604,10.786788922196861,13.485746889279946,13.599774651583436,13.646760470096138,12.178023207638955,12.891785365916192,14.497252296723261,12.01731880038888,13.723379091524656,13.268522775767346,12.027459747455779,13.831358731696193,12.893438046211557,14.347105915221857,15.205299344719151,10.979115862642054,11.339446018207243,11.309341493125505,12.846965136164528,12.690596186015746,12.862106594866781,14.517154949986594,-15.251065517488714,-15.134962682601421,-14.064975586643996,-13.731179467128346,-19.625878226438957,-15.426585324850704,-15.899174104069104,-15.741614531055916,-15.054296555266168,-15.542398011491763,-15.616560258171233,-17.685996450969487,9.8493681601961089,10.062278439481583,9.9121439929021378,-14.718593540645559,-14.554843329897833,-13.817280400896205,-16.34715767397455,-16.076127803780718,-15.081719539966674,-17.549050216534379,-17.700629085619706,-14.848615823900383,-4.31011489053187,-4.1614980750569188,-6.1049565265279853,-9.2897422615606082,-6.1708330305042036,-5.1872314066060889,-5.6975320636897555,-4.2893433133525569,-6.0139310724073942,-6.6013257790609803,-4.9738744954087499,-6.5822181333648988,-11.815569643027676,-4.1919186545304141,-5.5005963998511129,-5.5901623563053882,-3.2284000084157629],"name":"ASA/Volume","text":["Scale: LINS030110<br>Name: ASA (folded coil/turn)<br>Description: Total median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030113<br>Name: ASA (folded coil/turn)<br>Description: % total accessible surfaces of whole residues in random coil/turn structures (Lins et al., 2003)","Scale: JANJ780101<br>Name: ASA (folded protein)<br>Description: Average accessible surface area (Janin et al., 1978)","Scale: JANJ780103<br>Name: ASA (folded protein)<br>Description: Percentage of exposed residues (Janin et al., 1978)","Scale: LINS030104<br>Name: ASA (folded protein)<br>Description: Total median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030107<br>Name: ASA (folded protein)<br>Description: % total accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: CHOC760102<br>Name: ASA (folded proteins)<br>Description: Residue accessible surface area in folded protein (Chothia, 1976)","Scale: LINS030116<br>Name: ASA (folded β-strand)<br>Description: Total median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: LINS030119<br>Name: ASA (folded β-strand)<br>Description: % total accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: LINS030103<br>Name: Hydrophilic ASA<br>Description: Hydrophilic accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: LINS030112<br>Name: Hydrophilic ASA (folded coil/turn)<br>Description: Hydrophilic median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030115<br>Name: Hydrophilic ASA (folded coil/turn)<br>Description: % Hydrophilic accessible surfaces of whole residues in random coil/turn structures (Lins et al., 2003)","Scale: LINS030106<br>Name: Hydrophilic ASA (folded proteins)<br>Description: Hydrophilic median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030109<br>Name: Hydrophilic ASA (folded proteins)<br>Description: % Hydrophilic accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: LINS030118<br>Name: Hydrophilic ASA (folded β-strand)<br>Description: Hydrophilic median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: LINS030121<br>Name: Hydrophilic ASA (folded β-strand)<br>Description: % Hydrophilic accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: LINS030114<br>Name: Hydrophobic ASA (folded coil/turn)<br>Description: % Hydrophobic accessible surfaces of whole residues in random coil/turn structures(Lins et al., 2003)","Scale: LINS030108<br>Name: Hydrophobic ASA (folded protein)<br>Description: % Hydrophobic accessible surfaces vs win3 of whole residues from folded proteins (Lins et al., 2003)","Scale: LINS030120<br>Name: Hydrophobic ASA (folded β-strand)<br>Description: % Hydrophobic accessible surfaces of whole residues in beta-strands (Lins et al., 2003)","Scale: GUYH850104<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Janin index (Guy, 1985)","Scale: GUYH850105<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Chothia index (Guy, 1985)","Scale: RACS770103<br>Name: Side chain orientation<br>Description: Side chain orientational preference (Rackovsky-Scheraga, 1977)","Scale: VHEG790101<br>Name: TFE to lipophilic phase<br>Description: Transfer free energy to lipophilic phase (von Heijne-Blomberg, 1979)","Scale: BIOV880101<br>Name: Buriability<br>Description: Information value for accessibility; average fraction 35% (Biou et al., 1988)","Scale: BIOV880102<br>Name: Buriability<br>Description: Information value for accessibility; average fraction 23% (Biou et al., 1988)","Scale: WERD780101<br>Name: Buriability<br>Description: Propensity to be buried inside (Wertz-Scheraga, 1978)","Scale: ZHOH040103<br>Name: Buriability<br>Description: Buriability (Zhou-Zhou, 2004)","Scale: ARGP820103<br>Name: Buried<br>Description: Membrane-buried preference parameters (Argos et al., 1982)","Scale: CHOC760103<br>Name: Buried<br>Description: Proportion of residues 95% buried (Chothia, 1976)","Scale: CHOC760104<br>Name: Buried<br>Description: Proportion of residues 100% buried (Chothia, 1976)","Scale: JANJ780102<br>Name: Buried<br>Description: Percentage of buried residues (Janin et al., 1978)","Scale: JANJ790101<br>Name: Buried<br>Description: Ratio of buried and accessible molar fractions (Janin, 1979)","Scale: OLSK800101<br>Name: Buried<br>Description: Average internal preferences (Olsen, 1980)","Scale: NISK800101<br>Name: Interactivity<br>Description: 8 A contact number (Nishikawa-Ooi, 1980)","Scale: WARP780101<br>Name: Interactivity<br>Description: Average interactions per side chain atom (Warme-Morgan, 1978)","Scale: LINS030111<br>Name: Hydrophobic ASA (folded coil/turn)<br>Description: Hydrophobic median accessible surfaces of whole residues in random coil/turn structures, expressed in Å²  (Lins et al., 2003)","Scale: LINS030105<br>Name: Hydrophobic ASA (folded proteins)<br>Description: Hydrophobic median accessible surfaces of whole residues from folded proteins, expressed in Å²  (Lins et al., 2003)","Scale: LINS030117<br>Name: Hydrophobic ASA (folded β-strand)<br>Description: Hydrophobic median accessible surfaces of whole residues in beta-strands, expressed in Å²  (Lins et al., 2003)","Scale: PONP800107<br>Name: Accessibility reduction ratio<br>Description: Accessibility reduction ratio (Ponnuswamy et al., 1980)","Scale: ZIMJ680102<br>Name: Bulkiness<br>Description: Bulkiness (Zimmerman et al., 1968)","Scale: LINS030102<br>Name: Hydrophobic ASA<br>Description: Hydrophobic accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: BASU050101<br>Name: Interactivity<br>Description: Interactivity scale obtained from the contact matrix (Bastolla et al., 2005)","Scale: BASU050103<br>Name: Interactivity<br>Description: Interactivity scale obtained by maximizing the mean of correlation Coefficient over pairs of sequences sharing the TIM barrel fold (Bastolla et al., 2005)","Scale: NISK860101<br>Name: Interactivity<br>Description: 14 A contact number (Nishikawa-Ooi, 1986)","Scale: BULH740102<br>Name: Partial specific volume<br>Description: Apparent partial specific volume (Bull-Breese, 1974)","Scale: COHE430101<br>Name: Partial specific volume<br>Description: Partial specific volume (Cohn-Edsall, 1943)","Scale: PLIV810101<br>Name: Partition coefficient<br>Description: Partition Coefficient (Pliska et al., 1981)","Scale: CHOC760101<br>Name: Accessible surface area (ASA)<br>Description: Residue accessible surface area in tripeptide (Chothia, 1976)","Scale: LINS030101<br>Name: Accessible surface area (ASA)<br>Description: Total accessible surfaces of whole residues (backbone and lateral chain) calculated with a window 3, expressed in Å²  (Lins et al., 2003)","Scale: RADA880106<br>Name: Accessible surface area (ASA)<br>Description: Accessible surface area (Radzicka-Wolfenden, 1988)","Scale: ROSG850101<br>Name: Mean area buried<br>Description: Mean area buried on transfer (Rose et al., 1985)","Scale: BIGC670101<br>Name: Volume<br>Description: Residue volume (Bigelow, 1967)","Scale: CHOC750101<br>Name: Volume<br>Description: Average volume of buried residue (Chothia, 1975)","Scale: DAWD720101<br>Name: Volume<br>Description: Size (Dawson, 1972)","Scale: FAUJ880103<br>Name: Volume<br>Description: Normalized van der Waals volume (Fauchere et al., 1988)","Scale: GOLD730102<br>Name: Volume<br>Description: Residue volume (Goldsack-Chalifoux, 1973)","Scale: GRAR740103<br>Name: Volume<br>Description: Volume (Grantham, 1974)","Scale: HARY940101<br>Name: Volume<br>Description: Mean volumes of residues buried in protein interiors (Harpaz et al., 1994)","Scale: KRIW790103<br>Name: Volume<br>Description: Side chain volume (Krigbaum-Komoriya, 1979)","Scale: LEVM760106<br>Name: Volume<br>Description: van der Waals parameter R0 (Levitt, 1976)","Scale: PONJ960101<br>Name: Volume<br>Description: Average volumes of residues (Pontius et al., 1996)","Scale: TSAJ990101<br>Name: Volume<br>Description: Volumes including the crystallographic waters using the ProtOr (Tsai et al., 1999)","Scale: TSAJ990102<br>Name: Volume<br>Description: Volumes not including the crystallographic waters using the ProtOr (Tsai et al., 1999)","Scale: FASG760101<br>Name: Weight<br>Description: Molecular weight (Fasman, 1976)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#f0ea00","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(214,39,40,1)"}},"showlegend":true,"error_y":{"color":"rgba(214,39,40,1)"},"error_x":{"color":"rgba(214,39,40,1)"},"line":{"color":"rgba(214,39,40,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[-6.4765069121623728,-5.7418711027071945,-6.3192585482478378,-4.5929259058829013,-8.3275367589034861,-6.7516452818535129,-4.7881994814320255,-5.7354573083644746,-5.3621867516008237,-3.9190583637857639,-3.9609350938797898,-7.1757978759988506,-4.7574164484141415,-3.407040893830565,0.25929470585658654,0.19785296848311404,1.0210097521503758,1.0823132948209351,-7.830591628451919,-7.9894413158927158,4.4977678764370799,2.8612201254970313,2.4866977377952559,0.28257305917402725,-0.056678928606008067,-0.35624479628893613,5.3336594038907652,5.0787250994626927,5.0452841118670921,6.3295188284279895,5.6314716094208368,5.2724879964504208,-4.0264294983105859,16.066385800206568,3.5152634495742916,6.4131412066499935,-5.2564264812248975,-6.4451061862020662,22.27346868655313,14.71319528757064,1.480248341940489,3.1795392322719129,-1.6008978431428067,19.00904291695872,22.428403671458643,3.9374297492647852,21.67002648779286,2.2182634543658,5.9967018383138875,-10.994473008898893,-12.638396566280884,20.543148238998615,23.795293158351868,22.398665265829084,19.844354234882758,19.897917520434337,24.417587809190422,23.321561290205754,23.053148037756063,22.496260190034452,21.83845157996992,22.178214683589179,20.695966363841212,21.80316470474218,22.372015395981375,22.59974237080641,23.849408177969512,22.849470876211843,22.835466675321737,22.395064116585438,21.565395128294842,17.005130351043757,17.159487193035762,20.969688835154539,20.339135622063825,22.760642042792849,23.999495112175325,23.219163687357941,23.376241028821852,21.518751718610634,19.632311792471405,17.259895376634496,20.068922866610031,18.134855385272161,22.570902088142763,22.587764809253958,17.172307265908355,9.4762146119815753,14.463804631218904,10.735231512565889,15.663929747406513,20.551149483796841,23.091932617979648,15.585063662835735,16.795152798774691,16.096186608642501,15.794589328314927,19.978053839292979,1.3307986497728137,0.65869463380956383,1.474578751484471,-1.8234553276648857,3.319749340396716,-14.24470573831999,-13.871054328711661,-13.920334589264785,-15.293940221433658,-14.861613764449604,-13.102298443991053,-14.468027069059834,-14.569351736164341,-13.206631993891051,-15.181673792742535,-10.437519682913972,-11.726381656707753,-1.4167119842134814,-1.2924075209978245,-12.223028376144525,-1.0145768310988739,4.4176455982666978,3.7485821244680211,4.0306261821501161,4.4830970007519699,3.5102039572752339,4.8070346396608681,25.237744900056626,6.0215387262549545,-11.789458742165877,22.489569936440947,22.070081345780089,24.964398944659045,19.985304431237832,19.988617487444902,-12.528537949351724,0.045504299096972477,5.2642909933733719,-0.93922431530174166,0.30630412340172192,-0.58160015574519286,0.71326665827396507,-2.2740443526412486,-0.16885770061151692,2.2893038152936058,1.3910405171044784,-1.4626809095903863,1.1221922826272532,0.15108168281408682,0.74178822077370032,-0.15460270220561187,1.7882895443812326,2.1052578664518675,1.4746490201586295,1.2256621430604011,0.77247390250581571,-0.94472488573661662,-12.291495424062559,-10.858845673461861,-10.530938313827297,-9.1926919580661863,-6.2587836997872754,-3.3184818023429585,4.5716095299687218,-10.100012005866661,-9.8789810531190234,-9.2718941401237771,-2.5222240389758896,-0.3044718720669144,-0.18288444747723206,0.67980986351670736,-1.2645230365763545,-0.26387805487878496,-4.9749143431661729,-1.3312282211135609,-5.2070340042736971,0.66820251125022478,1.1648659226465714,0.32928020018504806,0.97656754945907887,-2.515192317731461,2.1487111110355781,-2.8952198273097913,-14.32403790582759,-7.1886176104180022,-8.1576021645492123,-7.5195425647532641,-7.1532047222744843,-7.3591750288583704,-7.5212253318610776,-5.9622009116599601,-6.4121127458090426,-7.5486916701377842,-8.4479189291121131,-7.1932569157579094,-6.935191148900433,-7.3258465134771846,-6.276270600863473,-8.4064332179514309,-7.0357090660473363,-6.3874010872884357,-8.2592240970327762,-8.9476009885502776,-7.5596496820477093,-8.6698302248143815,-12.12914064944235,-12.470148890701916,-9.9931411880654064,-10.156873871457334,-9.7923114234115296,-6.8733819646030199,-4.6633805298157407,-1.1269923175710144,-1.6251337964848176,-0.80995616330995557,-2.0024202572543559,-1.9217606523196726,3.6672881930493229,-4.7465576162374745,5.5878668148242996,-0.29899022085941285,15.783841541485938,3.2869011326098772],"y":[-0.97069080071055036,-1.0603470063097769,-3.1370943940742526,1.4123154536062412,-2.2667922438506514,-6.7682159039653378,-2.8428346120870192,-2.2625961720473318,-1.4609882746781404,-0.48067604522438906,-0.052385803673417962,-1.65053976009808,-4.0206522184144795,0.67633157658904741,0.71973656913809148,0.36273396044880368,-4.0723981518414396,-6.6621477590181595,-17.132628038275072,-11.446267961283354,-1.8515398208771612,-2.0515873392080883,-2.1859298290579194,-1.6608753668002185,-0.65293659507734703,-1.2963466379220614,-1.4387408691433388,-2.0831329433218335,-1.7898143781271385,4.6360754149504828,5.1274935443767582,1.7059931980088101,-5.2105106922236546,-12.551840383904121,16.824842567648375,-9.1909461383817419,-9.22105750418398,-10.888367943980636,7.4477865536862415,12.369007972274849,2.5390727922969845,1.7660982531046967,17.324469754243761,14.559654025930197,8.0968616619034997,6.4810067482361768,14.729256324580117,-22.98558481623925,1.7906411010790568,-6.506064097583419,9.8939206896576053,9.4350979446940073,8.5173690611403607,9.1811298343298784,9.4132369736735537,10.750899719531121,12.205074364067492,11.29546943382954,12.168180860454616,12.816798475336959,10.998114709638132,11.735409266262078,10.280987469573368,13.023916869685269,12.843316854667794,10.850204065267238,13.302702784204023,13.258837733470884,10.813679660467377,12.355446033933925,12.227199010010334,10.271968673749935,9.9807461808737212,12.243503298122812,7.5635776485194963,11.915599834236605,11.952529269529279,9.000376591988509,13.700769616378279,10.583688463920502,10.6920151915803,10.926697528185333,10.103078886560816,10.936674126790223,9.6606442403569144,11.09500731019647,10.508293629740857,14.410208603519907,12.774467120064974,13.209823589588972,11.621717627213497,8.6787487624182589,9.2251421758842902,12.034518416564257,9.1706370046324288,8.1071516328087245,11.62359034053779,9.851896351624811,0.48385441951915853,1.7639808656579086,0.5016195057482048,3.9645218915269744,1.7528764634167044,-2.5068051535699309,-2.3078291978524388,0.094685288196335524,-1.7497238694852546,-1.7703509526332248,-3.7481878296562852,-2.6815562642742483,-1.7548439711060746,-2.4369537026704458,-3.1333545357853696,-0.53784805348078291,-1.7971614519222099,5.0675948130078412,4.9466262605812545,-2.9745728661032418,5.5194406625652652,7.212508729323277,8.1332121051458355,11.182562357512253,10.124296165040967,8.2223166917725514,9.4366817622300427,13.221449528082776,4.6033504554597178,-2.0141687685904097,14.415595002684451,13.553860057710484,12.131503090261132,17.640463988685894,17.668058229777454,-3.8406613432200469,4.3548561490634237,-13.603463068582073,-16.688413942282764,-20.234068014184025,-19.433352500426992,-18.612830581609636,-19.817897038776394,-21.019754376543705,-21.369957674385127,-21.054135114256862,-19.659832939835574,-21.576980947532657,-20.129531432955684,-20.103377846638942,-16.00873686693064,-21.934274878572527,-21.838129519515874,-19.346383294794801,-19.171151071883326,-19.052904267129971,-20.652344046244728,-3.2669996848140022,-5.7819963135939254,-5.5068239675720783,2.8545385373352667,-7.0331753635954994,15.363913215129884,18.40824772101714,4.1614339065337855,3.853068573886258,3.1266475900484232,-22.877190466090827,-24.469841842672089,-24.700111444912352,-24.387762481076244,-21.852566403825726,-24.190933841070017,-14.74631704553612,-21.540685604481592,-18.080600977450832,-22.380695784040338,-20.870890671228423,-21.352805713095584,-19.484989382932351,-18.002644054053039,-19.425243044334586,-0.15294088412261345,-3.2340253631143323,-2.7833126450147074,-5.4673654572966566,-3.5638116079097952,-2.1433045807358484,-5.417561890602915,-0.94097154698808483,-0.67746130244277991,-1.2088052960748128,-3.8433691189878587,-1.2782140316265012,-4.1424016108373314,0.52409993542416089,-1.5522750568589083,-0.53114956104326827,-5.4012459612070414,-3.3251674083735074,-2.0170786315221862,-6.2185755285386772,-1.0594569228137203,-9.1045584023910742,-5.7694697423301546,-3.7197970391818211,-4.2079726285743586,-7.5717190579590765,-7.882932144824462,-0.10225303466875824,-2.5100112311866107,-3.8779505242058265,-0.17434849925977675,-0.43523779080241182,-0.15279870164077791,17.388347605673747,17.781495733395982,20.54171847680924,-14.884421666661508,-11.628581887419164,-14.265602591531231,7.6329505711021568,-15.566250069575485],"z":[14.959807735666427,15.254236555360624,15.041116373428002,15.285175884370203,13.912028624835068,13.752936511136644,14.911392902922136,16.384285671665271,16.298211211693992,16.437231974703131,16.744339518615881,18.715820118495571,16.21297806280915,17.66230108065853,20.034232530178151,20.36114536856763,17.424090062494749,-10.131950796990843,-16.59071334496133,-18.512758961674187,15.254316984010696,15.564245396747857,15.516042187502325,17.597365010970353,17.99178033381482,18.299516719601403,15.300213402736228,14.70859689808082,15.079693366130041,12.250557860687159,13.600201997631004,16.861314254112084,-11.518160050041082,-5.5577107431222323,15.102203920239061,-13.034013959842341,12.535956690978013,10.861080301895747,-5.785196392598472,1.8453075225601887,20.027983925574141,8.7553810301744317,7.3640949389163222,1.2442767120193159,1.3690820392576153,15.296978276239615,-0.17468718027870964,-14.957239662547341,17.065061915196637,9.5553768236096595,4.7624606876589963,-1.8882091014681806,-1.3660280828086702,-2.3789954595036562,-2.8549863427910158,-3.1104854726406646,-0.6995081247664775,-1.7622596302738178,-0.61763869437744023,0.30230175234034523,-1.5959614091701806,-1.0969308842914094,-2.1013285130003365,-0.83948027036555961,-1.4666245150900687,-0.90135342875261826,-1.4341511987792983,-1.6540190202329073,-1.773714808958212,0.9179131150624541,-3.7690147094943978,-1.1267058325709227,-0.52644081621625394,-1.1968378563744335,-0.56023408343987746,-2.3061281040794697,-1.723450205719659,-0.75844450371066818,-0.47865395315342907,-3.374082566755491,-1.8245404552353852,-1.568098053091745,-4.2444764047145132,-2.6071124446255665,-3.2052017323914153,-4.2847423848958144,1.2874739076308135,3.717043186535276,2.5294004411293938,2.1518502764508551,-2.5641191600036222,-1.3152564518980685,-0.15015446168662291,0.96471749967396569,-1.0952085206284976,-0.86611373966243277,-1.0965947510917786,1.6546461357759186,13.549115644900567,17.722896729359142,13.47875432435354,17.533046361811987,8.8537776717830194,13.56787449532966,11.774330879523182,11.106478226972301,13.290028275761623,13.005695858515365,12.122665149142621,13.038849444538219,11.622665821785942,13.56629999985862,13.310100774078659,15.361029543631229,17.912255621230081,10.580365726048017,10.348719065358285,17.836329033950179,10.646332559608759,14.661643161524404,14.66415743693646,14.453042285994679,13.616196217414856,14.698428035017489,13.833661884217944,-2.3627646518254277,12.036812644201648,15.510482223778279,-2.9646766863397529,-4.6790771494929668,0.9383783805035345,-1.0951799404758737,-1.0889294799324238,18.428423881293316,9.2573387094816972,-9.7511732378483025,-20.061406363438291,-17.946702710184695,-19.808380581294447,-14.804569425652671,-18.402661716882797,-17.737598974363312,-16.963003265219534,-18.015099411044673,-19.508287375198886,-18.267254807674441,-17.075241122844147,-20.648915723316055,-20.584184192914663,-16.351492477191673,-19.292475181160832,-19.122578253491103,-18.442304070813474,-16.825370443618443,-18.771783598393394,15.402090556681882,10.07313406669757,10.265655680680291,14.624872022735124,11.356721519515173,14.076542612622172,12.851781585113986,15.4182148930811,15.270992497754991,14.647535040223225,-18.579841757654847,-18.331654037839986,-18.300008673231549,-19.364485823506318,-18.767725619438892,-18.78176137066653,-11.810672487208326,-17.14635623369966,-14.76681121404097,-18.282728859449154,-20.406138767493648,-17.037499947532336,-17.913463510433228,-15.958616462000178,-17.917647206618447,18.868233552142812,14.413568286832129,18.740555851275655,16.04145734843074,15.195202000927052,15.342088673267451,15.143896368558204,15.951399545167696,16.486880155592303,16.478513366540668,18.537626291895151,15.756211350395697,14.584310119924725,17.245345777576802,19.038789122638974,16.11156921782613,14.988623592871056,15.952911880087216,15.400461649118537,14.91259424408705,15.117388607723058,12.19555060011,15.072113859322696,15.416314626670871,15.435435292664655,14.778968573793357,14.63934177238016,15.657271418053616,13.876745064169569,14.293256869749985,16.423589917085462,16.865598343058586,16.749664741640338,16.550462321637188,16.511356152328176,11.06369477539539,-10.916170975619801,-8.9400269462460109,-10.397820469785492,-0.72450144511149595,-9.7975430053773707],"name":"Conformation","text":["Scale: GEIM800108<br>Name: Aperiodic indices<br>Description: Aperiodic indices (Geisow-Roberts, 1980)","Scale: GEIM800111<br>Name: Aperiodic indices<br>Description: Aperiodic indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: CHAM830101<br>Name: Coil<br>Description: The Chou-Fasman parameter of the coil conformation (Charton-Charton, 1983)","Scale: GEIM800110<br>Name: Coil<br>Description: Aperiodic indices for beta-proteins (Geisow-Roberts, 1980)","Scale: NAGK730103<br>Name: Coil<br>Description: Normalized frequency of coil (Nagano, 1973)","Scale: QIAN880130<br>Name: Coil<br>Description: Weights for coil at the window position of -3 (Qian-Sejnowski, 1988)","Scale: QIAN880131<br>Name: Coil<br>Description: Weights for coil at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880132<br>Name: Coil<br>Description: Weights for coil at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880133<br>Name: Coil<br>Description: Weights for coil at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880134<br>Name: Coil<br>Description: Weights for coil at the window position of 1 (Qian-Sejnowski, 1988)","Scale: QIAN880135<br>Name: Coil<br>Description: Weights for coil at the window position of 2 (Qian-Sejnowski, 1988)","Scale: ROBB760112<br>Name: Coil<br>Description: Information measure for coil (Robson-Suzuki, 1976)","Scale: CHOP780207<br>Name: Non helical region (C-terminal)<br>Description: Normalized frequency of C-terminal non helical region (Chou-Fasman, 1978b)","Scale: QIAN880136<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 3 (Qian-Sejnowski, 1988)","Scale: QIAN880137<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880138<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 5 (Qian-Sejnowski, 1988)","Scale: ROBB760107<br>Name: Coil (C-terminal)<br>Description: Information measure for extended without H-bond (Robson-Suzuki, 1976)","Scale: QIAN880127<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -6 (Qian-Sejnowski, 1988)","Scale: QIAN880128<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -5 (Qian-Sejnowski, 1988)","Scale: GEOR030105<br>Name: Linker (small)<br>Description: Linker propensity from small dataset (linker length is less than six residues) (George-Heringa, 2003)","Scale: GEOR030101<br>Name: Linker<br>Description: Linker propensity from all dataset (George-Heringa, 2003)","Scale: GEOR030102<br>Name: Linker (1)<br>Description: Linker propensity from 1-linker dataset (George-Heringa, 2003)","Scale: GEOR030107<br>Name: Linker (Long)<br>Description: Linker propensity from long dataset (linker length is greater than 14 residues) (George-Heringa, 2003)","Scale: GEOR030109<br>Name: Linker (Non-Helical)<br>Description: Linker propensity from non-helical (annotated by DSSP) dataset (George-Heringa, 2003)","Scale: FINA910102<br>Name: α-helix (C-term, out)<br>Description: Helix initiation parameter at position i,i+1,i+2 (Finkelstein et al., 1991)","Scale: ONEK900102<br>Name: α-helix (C-term, out)<br>Description: Helix formation parameters (delta delta G) (O'Neil-DeGrado, 1990)","Scale: GEOR030103<br>Name: Linker (2)<br>Description: Linker propensity from 2-linker dataset (George-Heringa, 2003)","Scale: GEOR030104<br>Name: Linker (3)<br>Description: Linker propensity from 3-linker dataset (George-Heringa, 2003)","Scale: GEOR030106<br>Name: Linker (medium)<br>Description: Linker propensity from medium dataset (linker length is between six and 14 residues) (George-Heringa, 2003)","Scale: AURR980118<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C\" (Aurora-Rose, 1998)","Scale: AURR980101<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N4'(Aurora-Rose, 1998)","Scale: AURR980102<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N\"' (Aurora-Rose, 1998)","Scale: QIAN880139<br>Name: Coil (C-terminal)<br>Description: Weights for coil at the window position of 6 (Qian-Sejnowski, 1988)","Scale: QIAN880129<br>Name: Coil (N-terminal)<br>Description: Weights for coil at the window position of -4 (Qian-Sejnowski, 1988)","Scale: ISOY800107<br>Name: Double bend<br>Description: Normalized relative frequency of double bend (Isogai et al., 1980)","Scale: VASM830103<br>Name: Extended<br>Description: Relative population of conformational state E (Vasquez et al., 1983)","Scale: RACS820104<br>Name: Extended (left-handed)<br>Description: Average relative fractional occurrence in EL(i) (Rackovsky-Scheraga, 1982)","Scale: RACS820112<br>Name: Extended (right-handed)<br>Description: Average relative fractional occurrence in ER(i-1) (Rackovsky-Scheraga, 1982)","Scale: VASM830101<br>Name: α-helix<br>Description: Relative population of conformational state A (Vasquez et al., 1983)","Scale: RICJ880114<br>Name: α-helix (C-terminal, inside)<br>Description: Relative preference value at C1 (Richardson-Richardson, 1988)","Scale: AURR980120<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C4' (Aurora-Rose, 1998)","Scale: RICJ880117<br>Name: α-helix (C-terminal, outside)<br>Description: Relative preference value at C\" (Richardson-Richardson, 1988)","Scale: RACS820103<br>Name: α-helix (left-handed)<br>Description: Average relative fractional occurrence in AL(i) (Rackovsky-Scheraga, 1982)","Scale: RICJ880108<br>Name: α-helix (middle)<br>Description: Relative preference value at N5 (Richardson-Richardson, 1988)","Scale: QIAN880101<br>Name: α-helix (N-terminal)<br>Description: Weights for alpha-helix at the window position of -6 (Qian-Sejnowski, 1988)","Scale: RICJ880104<br>Name: α-helix (N-terminal, inside)<br>Description: Relative preference value at N1 (Richardson-Richardson, 1988)","Scale: GEIM800103<br>Name: α-helix (β-proteins)<br>Description: Alpha-helix indices for beta-proteins (Geisow-Roberts, 1980)","Scale: CHAM830102<br>Name: β-sheet<br>Description: A parameter defined from the residuals obtained from the best correlation of the Chou-Fasman parameter of beta-sheet (Charton-Charton, 1983)","Scale: QIAN880117<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -3 (Qian-Sejnowski, 1988)","Scale: QIAN880123<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 3 (Qian-Sejnowski, 1988)","Scale: OOBM850102<br>Name: β-turn (optimized propensity)<br>Description: Optimized propensity to form reverse turn (Oobatake et al., 1985)","Scale: AURR980109<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini N4 (Aurora-Rose, 1998)","Scale: AURR980111<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C5 (Aurora-Rose, 1998)","Scale: AURR980112<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C4 (Aurora-Rose, 1998)","Scale: AURR980113<br>Name: α-helix<br>Description: Normalized positional residue frequency at helix termini C3 (Aurora-Rose, 1998)","Scale: BEGF750101<br>Name: α-helix<br>Description: Conformational parameter of inner helix (Beghin-Dirkx, 1975)","Scale: BURA740101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Burgess et al., 1974)","Scale: CHOP780201<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Chou-Fasman, 1978b)","Scale: GEIM800101<br>Name: α-helix<br>Description: Alpha-helix indices (Geisow-Roberts, 1980)","Scale: GEIM800104<br>Name: α-helix<br>Description: Alpha-helix indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: ISOY800101<br>Name: α-helix<br>Description: Normalized relative frequency of alpha-helix (Isogai et al., 1980)","Scale: KANM800101<br>Name: α-helix<br>Description: Average relative probability of helix (Kanehisa-Tsong, 1980)","Scale: KANM800103<br>Name: α-helix<br>Description: Average relative probability of inner helix (Kanehisa-Tsong, 1980)","Scale: LEVM780101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix, with weights (Levitt, 1978)","Scale: LEVM780104<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix, unweighted (Levitt, 1978)","Scale: MAXF760101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Maxfield-Scheraga, 1976)","Scale: NAGK730101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Nagano, 1973)","Scale: PALJ810101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix from LG (Palau et al., 1981)","Scale: PALJ810102<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix from CF (Palau et al., 1981)","Scale: PALJ810109<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix in alpha/beta class (Palau et al., 1981)","Scale: QIAN880105<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880109<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of 2 (Qian-Sejnowski, 1988)","Scale: QIAN880110<br>Name: α-helix<br>Description: Weights for alpha-helix at the window position of 3 (Qian-Sejnowski, 1988)","Scale: RACS820108<br>Name: α-helix<br>Description: Average relative fractional occurrence in AR(i-1) (Rackovsky-Scheraga, 1982)","Scale: RICJ880112<br>Name: α-helix<br>Description: Relative preference value at C3 (Richardson-Richardson, 1988)","Scale: ROBB760101<br>Name: α-helix<br>Description: Information measure for alpha-helix (Robson-Suzuki, 1976)","Scale: TANS770101<br>Name: α-helix<br>Description: Normalized frequency of alpha-helix (Tanaka-Scheraga, 1977)","Scale: AURR980110<br>Name: α-helix (middle)<br>Description: Normalized positional residue frequency at helix termini N5 (Aurora-Rose, 1998)","Scale: CRAJ730101<br>Name: α-helix (middle)<br>Description: Normalized frequency of middle helix (Crawford et al., 1973)","Scale: QIAN880106<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880107<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880108<br>Name: α-helix (middle)<br>Description: Weights for alpha-helix at the window position of 1 (Qian-Sejnowski, 1988)","Scale: RICJ880109<br>Name: α-helix (middle)<br>Description: Relative preference value at Mid (Richardson-Richardson, 1988)","Scale: ROBB760103<br>Name: α-helix (middle)<br>Description: Information measure for middle helix (Robson-Suzuki, 1976)","Scale: AURR980108<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized positional residue frequency at helix termini N3 (Aurora-Rose, 1998)","Scale: QIAN880104<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -3 (Qian-Sejnowski, 1988)","Scale: AURR980116<br>Name: α-helix (C-terminal, C-cap)<br>Description: Normalized positional residue frequency at helix termini Cc (Aurora-Rose, 1998)","Scale: FINA910103<br>Name: α-helix (C-terminal, inside)<br>Description: Helix termination parameter at position j-2,j-1,j (Finkelstein et al., 1991)","Scale: RICJ880113<br>Name: α-helix (C-terminal, inside)<br>Description: Relative preference value at C2 (Richardson-Richardson, 1988)","Scale: FINA910104<br>Name: α-helix termination<br>Description: Helix termination parameter at position j+1 (Finkelstein et al., 1991)","Scale: BLAM930101<br>Name: α-helix<br>Description: Alpha-helix propensity of position 44 in T4 lysozyme (Blaber et al., 1993)","Scale: AURR980114<br>Name: α-helix (C-terminal)<br>Description: Normalized positional residue frequency at helix termini C2 (Aurora-Rose, 1998)","Scale: AURR980115<br>Name: α-helix (C-terminal)<br>Description: Normalized positional residue frequency at helix termini C1 (Aurora-Rose, 1998)","Scale: CHOP780205<br>Name: α-helix (C-terminal)<br>Description: Normalized frequency of C-terminal helix (Chou-Fasman, 1978b)","Scale: QIAN880111<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880112<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 5 (Qian-Sejnowski, 1988)","Scale: ROBB760104<br>Name: α-helix (C-terminal)<br>Description: Information measure for C-terminal helix (Robson-Suzuki, 1976)","Scale: RICJ880110<br>Name: α-helix (middle)<br>Description: Relative preference value at C5 (Richardson-Richardson, 1988)","Scale: MAXF760106<br>Name: α region<br>Description: Normalized frequency of alpha region (Maxfield-Scheraga, 1976)","Scale: AURR980119<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini C\"' (Aurora-Rose, 1998)","Scale: ISOY800106<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized relative frequency of helix end (Isogai et al., 1980)","Scale: RICJ880116<br>Name: α-helix (C-terminal, outside)<br>Description: Relative preference value at C' (Richardson-Richardson, 1988)","Scale: TANS770102<br>Name: α-helix (C-terminal, outside)<br>Description: Normalized frequency of isolated helix (Tanaka-Scheraga, 1977)","Scale: ISOY800108<br>Name: Coil (normalized by low Prolin value)<br>Description: Normalized relative frequency of coil (Isogai et al., 1980)","Scale: TANS770109<br>Name: Coil (normalized by low Prolin value)<br>Description: Normalized frequency of coil (Tanaka-Scheraga, 1977)","Scale: KIMC930101<br>Name: Free energy (β-sheet)<br>Description: Thermodynamic beta-sheet propensity (Kim-Berg, 1993)","Scale: AURR980117<br>Name: α-helix (C-terminal, C-cap)<br>Description: Normalized positional residue frequency at helix termini C' (Aurora-Rose, 1998)","Scale: RICJ880115<br>Name: α-helix (C-terminal, C-cap)<br>Description: Relative preference value at C-cap (Richardson-Richardson, 1988)","Scale: KOEP990101<br>Name: α-helix (left-handed)<br>Description: Alpha-helix propensity derived from designed sequences (Koehl-Levitt, 1999)","Scale: MAXF760104<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of left-handed alpha-helix (Maxfield-Scheraga, 1976)","Scale: MAXF760105<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of zeta L (Maxfield-Scheraga, 1976)","Scale: RACS820109<br>Name: α-helix (left-handed)<br>Description: Average relative fractional occurrence in AL(i-1) (Rackovsky-Scheraga, 1982)","Scale: TANS770107<br>Name: α-helix (left-handed)<br>Description: Normalized frequency of left-handed helix (Tanaka-Scheraga, 1977)","Scale: PALJ810113<br>Name: β-turn (α class)<br>Description: Normalized frequency of turn in all-alpha class (Palau et al., 1981)","Scale: CHOP780206<br>Name: Non helical region (N-terminal)<br>Description: Normalized frequency of N-terminal non helical region (Chou-Fasman, 1978b)","Scale: AURR980104<br>Name: α-helix (N-terminal, N-cap)<br>Description: Normalized positional residue frequency at helix termini N'(Aurora-Rose, 1998)","Scale: AURR980105<br>Name: α-helix (N-terminal, N-cap)<br>Description: Normalized positional residue frequency at helix termini Nc (Aurora-Rose, 1998)","Scale: RICJ880103<br>Name: α-helix (N-terminal, N-cap)<br>Description: Relative preference value at N-cap (Richardson-Richardson, 1988)","Scale: FINA910101<br>Name: α-helix initiation<br>Description: Helix initiation parameter at position i-1 (Finkelstein et al., 1991)","Scale: AURR980106<br>Name: α-helix (N-terminal)<br>Description: Normalized positional residue frequency at helix termini N1 (Aurora-Rose, 1998)","Scale: CHOP780204<br>Name: α-helix (N-terminal)<br>Description: Normalized frequency of N-terminal helix (Chou-Fasman, 1978b)","Scale: RICJ880105<br>Name: α-helix (N-terminal)<br>Description: Relative preference value at N2 (Richardson-Richardson, 1988)","Scale: RICJ880106<br>Name: α-helix (N-terminal)<br>Description: Relative preference value at N3 (Richardson-Richardson, 1988)","Scale: ROBB760102<br>Name: α-helix (N-terminal)<br>Description: Information measure for N-terminal helix (Robson-Suzuki, 1976)","Scale: AURR980107<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized positional residue frequency at helix termini N2 (Aurora-Rose, 1998)","Scale: QIAN880102<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -5 (Qian-Sejnowski, 1988)","Scale: AURR980103<br>Name: α-helix (N-terminal, outside)<br>Description: Normalized positional residue frequency at helix termini N\" (Aurora-Rose, 1998)","Scale: RICJ880101<br>Name: α-helix (N-terminal, outside)<br>Description: Relative preference value at N\" (Richardson-Richardson, 1988)","Scale: PALJ810108<br>Name: α-helix (N-terminal, inside)<br>Description: Normalized frequency of alpha-helix in alpha+beta class (Palau et al., 1981)","Scale: QIAN880103<br>Name: α-helix (N-terminal, inside)<br>Description: Weights for alpha-helix at the window position of -4 (Qian-Sejnowski, 1988)","Scale: RACS820102<br>Name: α-helix (right-handed)<br>Description: Average relative fractional occurrence in AR(i) (Rackovsky-Scheraga, 1982)","Scale: GEIM800102<br>Name: α-helix (α-proteins)<br>Description: Alpha-helix indices for alpha-proteins (Geisow-Roberts, 1980)","Scale: PALJ810107<br>Name: α-helix (α-proteins)<br>Description: Normalized frequency of alpha-helix in all-alpha class (Palau et al., 1981)","Scale: MAXF760103<br>Name: β/α-bridge<br>Description: Normalized frequency of zeta R (Maxfield-Scheraga, 1976)","Scale: TANS770108<br>Name: β/α-bridge<br>Description: Normalized frequency of zeta R (Tanaka-Scheraga, 1977)","Scale: RACS820105<br>Name: Extended<br>Description: Average relative fractional occurrence in E0(i) (Rackovsky-Scheraga, 1982)","Scale: BEGF750102<br>Name: β-sheet<br>Description: Conformational parameter of beta-structure (Beghin-Dirkx, 1975)","Scale: CHOP780202<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet (Chou-Fasman, 1978b)","Scale: CHOP780208<br>Name: β-sheet<br>Description: Normalized frequency of N-terminal beta-sheet (Chou-Fasman, 1978b)","Scale: CHOP780209<br>Name: β-sheet<br>Description: Normalized frequency of C-terminal beta-sheet (Chou-Fasman, 1978b)","Scale: CRAJ730102<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet (Crawford et al., 1973)","Scale: KANM800102<br>Name: β-sheet<br>Description: Average relative probability of beta-sheet (Kanehisa-Tsong, 1980)","Scale: LEVM780102<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet, with weights (Levitt, 1978)","Scale: LEVM780105<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet, unweighted (Levitt, 1978)","Scale: NAGK730102<br>Name: β-sheet<br>Description: Normalized frequency of beta-structure (Nagano, 1973)","Scale: PALJ810103<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet from LG (Palau et al., 1981)","Scale: PALJ810104<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet from CF (Palau et al., 1981)","Scale: PALJ810110<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in all-beta class (Palau et al., 1981)","Scale: PALJ810111<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in alpha+beta class (Palau et al., 1981)","Scale: PALJ810112<br>Name: β-sheet<br>Description: Normalized frequency of beta-sheet in alpha/beta class (Palau et al., 1981)","Scale: QIAN880118<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -2 (Qian-Sejnowski, 1988)","Scale: QIAN880119<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of -1 (Qian-Sejnowski, 1988)","Scale: QIAN880120<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 0 (Qian-Sejnowski, 1988)","Scale: QIAN880121<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 1 (Qian-Sejnowski, 1988)","Scale: ROBB760106<br>Name: β-sheet<br>Description: Information measure for pleated-sheet (Robson-Suzuki, 1976)","Scale: ISOY800105<br>Name: β-sheet (C-terminal)<br>Description: Normalized relative frequency of bend S (Isogai et al., 1980)","Scale: QIAN880124<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 4 (Qian-Sejnowski, 1988)","Scale: QIAN880125<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 5 (Qian-Sejnowski, 1988)","Scale: QIAN880126<br>Name: β-sheet (C-terminal)<br>Description: Weights for beta-sheet at the window position of 6 (Qian-Sejnowski, 1988)","Scale: CHOP780212<br>Name: β-turn (1st residue)<br>Description: Frequency of the 1st residue in turn (Chou-Fasman, 1978b)","Scale: KOEP990102<br>Name: Extended (designed β-sheet)<br>Description: Beta-sheet propensity derived from designed sequences (Koehl-Levitt, 1999)","Scale: RACS820101<br>Name: α-helix with flat 4-Cα structure (i)<br>Description: Average relative fractional occurrence in A0(i) (Rackovsky-Scheraga, 1982)","Scale: QIAN880114<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -6 (Qian-Sejnowski, 1988)","Scale: QIAN880115<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -5 (Qian-Sejnowski, 1988)","Scale: QIAN880116<br>Name: β-sheet (N-terminal)<br>Description: Weights for beta-sheet at the window position of -4 (Qian-Sejnowski, 1988)","Scale: BURA740102<br>Name: Extended<br>Description: Normalized frequency of extended structure (Burgess et al., 1974)","Scale: ISOY800102<br>Name: Extended<br>Description: Normalized relative frequency of extended structure (Isogai et al., 1980)","Scale: MAXF760102<br>Name: Extended<br>Description: Normalized frequency of extended structure (Maxfield-Scheraga, 1976)","Scale: RACS820111<br>Name: Extended<br>Description: Average relative fractional occurrence in E0(i-1) (Rackovsky-Scheraga, 1982)","Scale: ROBB760105<br>Name: Extended<br>Description: Information measure for extended (Robson-Suzuki, 1976)","Scale: TANS770103<br>Name: Extended<br>Description: Normalized frequency of extended structure (Tanaka-Scheraga, 1977)","Scale: BAEK050101<br>Name: Linker index (negative linker propensity, 10-100 AA long region)<br>Description: Linker index (Bae et al., 2005)","Scale: KANM800104<br>Name: β-sheet<br>Description: Average relative probability of inner beta-sheet (Kanehisa-Tsong, 1980)","Scale: QIAN880122<br>Name: β-sheet<br>Description: Weights for beta-sheet at the window position of 2 (Qian-Sejnowski, 1988)","Scale: GEIM800105<br>Name: β-strand<br>Description: Beta-strand indices (Geisow-Roberts, 1980)","Scale: GEIM800106<br>Name: β-strand<br>Description: Beta-strand indices for beta-proteins (Geisow-Roberts, 1980)","Scale: GEIM800107<br>Name: β-strand<br>Description: Beta-strand indices for alpha/beta-proteins (Geisow-Roberts, 1980)","Scale: LIFS790101<br>Name: β-strand<br>Description: Conformational preference for all beta-strands (Lifson-Sander, 1979)","Scale: LIFS790102<br>Name: β-strand<br>Description: Conformational preference for parallel beta-strands (Lifson-Sander, 1979)","Scale: LIFS790103<br>Name: β-strand<br>Description: Conformational preference for antiparallel beta-strands (Lifson-Sander, 1979)","Scale: RACS820110<br>Name: Extended (left-handed)<br>Description: Average relative fractional occurrence in EL(i-1) (Rackovsky-Scheraga, 1982)","Scale: RACS820106<br>Name: Extended (right-handed)<br>Description: Average relative fractional occurrence in ER(i) (Rackovsky-Scheraga, 1982)","Scale: CHOP780211<br>Name: Non β region (C-terminal)<br>Description: Normalized frequency of C-terminal non beta region (Chou-Fasman, 1978b)","Scale: BEGF750103<br>Name: β-turn<br>Description: Conformational parameter of beta-turn (Beghin-Dirkx, 1975)","Scale: CHOP780101<br>Name: β-turn<br>Description: Normalized frequency of beta-turn (Chou-Fasman, 1978a)","Scale: CHOP780203<br>Name: β-turn<br>Description: Normalized frequency of beta-turn (Chou-Fasman, 1978b)","Scale: CRAJ730103<br>Name: β-turn<br>Description: Normalized frequency of turn (Crawford et al., 1973)","Scale: ISOY800103<br>Name: β-turn<br>Description: Normalized relative frequency of bend (Isogai et al., 1980)","Scale: LEVM780103<br>Name: β-turn<br>Description: Normalized frequency of reverse turn, with weights (Levitt, 1978)","Scale: LEVM780106<br>Name: β-turn<br>Description: Normalized frequency of reverse turn, unweighted (Levitt, 1978)","Scale: LEWP710101<br>Name: β-turn<br>Description: Frequency of occurrence in beta-bends (Lewis et al., 1971)","Scale: PALJ810105<br>Name: β-turn<br>Description: Normalized frequency of turn from LG (Palau et al., 1981)","Scale: PALJ810106<br>Name: β-turn<br>Description: Normalized frequency of turn from CF (Palau et al., 1981)","Scale: PALJ810114<br>Name: β-turn<br>Description: Normalized frequency of turn in all-beta class (Palau et al., 1981)","Scale: PALJ810115<br>Name: β-turn<br>Description: Normalized frequency of turn in alpha+beta class (Palau et al., 1981)","Scale: PRAM900104<br>Name: β-turn<br>Description: Relative frequency in reverse-turn (Prabhakaran, 1990)","Scale: ROBB760113<br>Name: β-turn<br>Description: Information measure for loop (Robson-Suzuki, 1976)","Scale: TANS770110<br>Name: β-turn<br>Description: Normalized frequency of chain reversal (Tanaka-Scheraga, 1977)","Scale: CHOP780216<br>Name: β-turn (middle)<br>Description: Normalized frequency of the 2nd and 3rd residues in turn (Chou-Fasman, 1978b)","Scale: ROBB760110<br>Name: β-turn (middle)<br>Description: Information measure for middle turn (Robson-Suzuki, 1976)","Scale: PALJ810116<br>Name: β-turn (α/β class)<br>Description: Normalized frequency of turn in alpha/beta class (Palau et al., 1981)","Scale: RACS820107<br>Name: α-helix with flat 4-Cα structure (i+1)<br>Description: Average relative fractional occurrence in A0(i-1) (Rackovsky-Scheraga, 1982)","Scale: ROBB760108<br>Name: β-turn<br>Description: Information measure for turn (Robson-Suzuki, 1976)","Scale: CHOP780214<br>Name: β-turn (3rd residue)<br>Description: Frequency of the 3rd residue in turn (Chou-Fasman, 1978b)","Scale: TANS770105<br>Name: β-turn (3rd residue)<br>Description: Normalized frequency of chain reversal S (Tanaka-Scheraga, 1977)","Scale: CHOP780215<br>Name: β-turn (4th residue)<br>Description: Frequency of the 4th residue in turn (Chou-Fasman, 1978b)","Scale: ROBB760111<br>Name: β-turn (C-terminal)<br>Description: Information measure for C-terminal turn (Robson-Suzuki, 1976)","Scale: GEIM800109<br>Name: Aperiodic indices<br>Description: Aperiodic indices for alpha-proteins (Geisow-Roberts, 1980)","Scale: CHOP780210<br>Name: Non β region (N-terminal)<br>Description: Normalized frequency of N-terminal non beta region (Chou-Fasman, 1978b)","Scale: ROBB760109<br>Name: β-turn (1st residue)<br>Description: Information measure for N-terminal turn (Robson-Suzuki, 1976)","Scale: CHOP780213<br>Name: β-turn (2nd residue)<br>Description: Frequency of the 2nd residue in turn (Chou-Fasman, 1978b)","Scale: ISOY800104<br>Name: β-turn (2nd residue)<br>Description: Normalized relative frequency of bend R (Isogai et al., 1980)","Scale: TANS770104<br>Name: β-turn (2nd residue)<br>Description: Normalized frequency of chain reversal R (Tanaka-Scheraga, 1977)","Scale: MONM990101<br>Name: β-turn (TM helix)<br>Description: Turn propensity scale for transmembrane helices (Monne et al., 1999)","Scale: MONM990201<br>Name: β-turn (TM helix)<br>Description: Averaged turn propensities in a transmembrane helix (Monne et al., 1999)","Scale: TANS770106<br>Name: β-turn in double bend<br>Description: Normalized frequency of chain reversal D (Tanaka-Scheraga, 1977)","Scale: SUYM030101<br>Name: Linker index (negative linker propensity, 10-100 AA long region)<br>Description: Linker propensity index (Suyama-Ohara, 2003)","Scale: RICJ880107<br>Name: α-helix<br>Description: Relative preference value at N4 (Richardson-Richardson, 1988)","Scale: RICJ880111<br>Name: α-helix<br>Description: Relative preference value at C4 (Richardson-Richardson, 1988)","Scale: QIAN880113<br>Name: α-helix (C-terminal)<br>Description: Weights for alpha-helix at the window position of 6 (Qian-Sejnowski, 1988)","Scale: FODM020101<br>Name: π-helix<br>Description: Propensity of amino acids within pi-helices (Fodje-Al-Karadaghi, 2002)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#62C923","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(148,103,189,1)"}},"showlegend":true,"error_y":{"color":"rgba(148,103,189,1)"},"error_x":{"color":"rgba(148,103,189,1)"},"line":{"color":"rgba(148,103,189,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[-15.973319976023602,-17.127773307006663,-15.859763852200453,-16.240782822928093,-16.7489338551764,-16.718880777143806,-17.3830465510739,-16.858795847202295,-16.355051730919865,-17.961233682186883,-17.059743527659347,-18.441850819585486,-18.684485821607122,-16.137780723759796,-15.53681932567188,-16.836536395058459,-17.354945474040996,-15.831374799200855,-15.907104685944034,-15.671565760984382,-15.302046747795533,-14.287109333924921,-16.231647553945773,-15.51202198396674,-4.7236389603306437,-8.7862500432378816,-4.0426563393032957,-3.6574182931148909,-16.794864855750692,-16.18752346676105,-15.647882937427083,-16.411869817504321,-15.39207302549636,-15.453937823564113,-17.072807131992505,-17.435873448909014,-16.111704530382802,-14.358082615079427,-15.82406080473813,-16.007663134921774,-15.530124514894343,-2.6428230307058609,-16.963800957921674,-2.8562029205790971,-1.8039804813603231,-2.7973807159589481,-2.6003985981363655,-5.4026978652203566,-5.6545160320222259,-7.5941913208006397,-6.3125190865067635,-6.3119627219058012,-6.3220097052708342,-6.4099870531387264,-5.5141632391120963,-6.1884891847141938,-14.2731311827499,-4.9135929158560039],"y":[6.9636860631294919,6.6363830158188097,6.2130359387751355,6.1919773420737654,7.2140563059799838,7.8872522278096469,7.0971854422614129,6.00729872436251,7.6915069994827858,5.8098819544111659,8.86177164898365,6.6799054220954845,7.0656957768556321,8.7069096333715041,9.0574614363557089,8.8945105716531341,8.1119250509733192,5.7224255331786873,5.0326172615005555,7.1465173275527523,6.1636199619850149,8.0961363465469205,7.8100670081016359,9.9864317315462827,17.36829726021168,17.169996149980459,17.253066361718606,17.12900989811024,4.6151638234526065,4.2484423478106637,2.2510733766472191,1.6913233617687133,1.9832511397047323,2.7387175162285833,1.9602805589077406,2.484506840027163,2.8117301680378897,7.9931598462297453,3.1200903479097453,3.4024991565308778,3.6906536774701846,-7.9199700631362813,1.6871917448223821,-7.8469913996482825,-8.2943945115868214,18.959537268550434,17.215520456159801,16.995062636184386,20.010307732127426,18.413771060859236,17.23976219039735,17.431487639082786,18.888498974161067,18.604426813125766,18.430051881824962,14.94556769293564,-5.3210600263311845,-9.1341065188736401],"z":[-4.1994624488948586,-5.7011346494767503,-4.8203566793358501,-3.9422709165576024,-5.6237913836979621,-3.1480062408168235,-4.85325340979188,-4.7670767646078476,-5.1717618797678577,-5.3809766935782699,-4.5826160581736852,-4.3319416444771361,-5.3434489871792081,-4.0999079240093801,-4.0248808968083365,-3.5423574376674294,-5.8330788216416734,-5.777120321650191,-4.3980582117521294,-6.2667866676046513,-6.3501495364307567,-7.6200216833622036,-6.8963971237831396,-2.5756926943734779,7.3046159664539596,9.3684393619269759,7.3221684630802519,7.0299597025445522,-7.7902980294385049,-7.6671311619580509,-9.1652881021941113,-10.936912939593395,-9.7252913792071141,-9.7616040408568612,-10.816689473398771,-10.21365641961742,-9.5479585129349189,-7.582011111893646,-7.7540698130241799,-8.664312828324805,-8.943425324419497,-19.845636511901539,-11.123679728677741,-19.822155816146655,-19.982056916053324,14.69957268367196,13.919931876078559,10.824875812363144,13.901316509193483,10.157297261334493,13.106394807475274,13.086289937590792,11.728178392426846,11.221018601631862,11.540809493217957,13.725698854078228,-12.138479069639816,11.574175237239418],"name":"Composition","text":["Scale: DAYM780101<br>Name: AA composition<br>Description: Amino acid composition (Dayhoff et al., 1978a)","Scale: JOND920101<br>Name: AA composition<br>Description: Relative frequency of occurrence (Jones et al., 1992)","Scale: JUKT750101<br>Name: AA composition<br>Description: Amino acid distribution (Jukes et al., 1975)","Scale: JUNJ780101<br>Name: AA composition<br>Description: Sequence frequency (Jungck, 1978)","Scale: NAKH900101<br>Name: AA composition<br>Description: AA composition of total proteins (Nakashima et al., 1990)","Scale: NAKH900102<br>Name: AA composition<br>Description: SD of AA composition of total proteins (Nakashima et al., 1990)","Scale: CEDJ970102<br>Name: AA composition (anchored)<br>Description: Composition of amino acids in anchored proteins [%] (Cedano et al., 1997)","Scale: CEDJ970101<br>Name: AA composition (EXT)<br>Description: Composition of amino acids in extracellular proteins [%] (Cedano et al., 1997)","Scale: CEDJ970104<br>Name: AA composition (INT)<br>Description: Composition of amino acids in intracellular proteins [%] (Cedano et al., 1997)","Scale: NAKH920107<br>Name: Membrane proteins (multi-spanning, EXT)<br>Description: AA composition of EXT of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920106<br>Name: Membrane proteins (multi-spanning, INT)<br>Description: AA composition of CYT of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920103<br>Name: Membrane proteins (single-spanning, EXT)<br>Description: AA composition of EXT of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920104<br>Name: Membrane proteins (single-spanning, EXT)<br>Description: AA composition of EXT2 of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920101<br>Name: Membrane proteins (single-spanning, INT)<br>Description: AA composition of CYT of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920102<br>Name: Membrane proteins (single-spanning, INT)<br>Description: AA composition of CYT2 of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: CEDJ970105<br>Name: Nuclear proteins<br>Description: Composition of amino acids in nuclear proteins [%] (Cedano et al., 1997)","Scale: FUKS010112<br>Name: Nuclear proteins<br>Description: Entire chain composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000102<br>Name: Proteins of mesophiles<br>Description: Distribution of amino acid residues in the 18 non-redundant families of mesophilic proteins (Kumar et al., 2000)","Scale: FUKS010111<br>Name: Proteins of mesophiles (EXT)<br>Description: Entire chain composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010110<br>Name: Proteins of mesophiles (INT)<br>Description: Entire chain composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000101<br>Name: Proteins of thermophiles<br>Description: Distribution of amino acid residues in the 18 non-redundant families of thermophilic proteins (Kumar et al., 2000)","Scale: KUMS000103<br>Name: Proteins of thermophiles<br>Description: Distribution of amino acid residues in the alpha-helices in thermophilic proteins (Kumar et al., 2000)","Scale: FUKS010109<br>Name: Proteins of thermophiles (INT)<br>Description: Entire chain composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: DIGM050101<br>Name: Hydrostatic pressure<br>Description: Hydrostatic Pressure asymmetry index, PAI (Di Giulio, 2005)","Scale: FUKS010104<br>Name: Nuclear proteins<br>Description: Surface composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010103<br>Name: Proteins of mesophiles (EXT)<br>Description: Surface composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010102<br>Name: Proteins of mesophiles (INT)<br>Description: Surface composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010101<br>Name: Proteins of thermophiles (INT)<br>Description: Surface composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: CEDJ970103<br>Name: Membrane proteins<br>Description: Composition of amino acids in membrane proteins [%] (Cedano et al., 1997)","Scale: NAKH900109<br>Name: Membrane proteins<br>Description: AA composition of membrane proteins (Nakashima et al., 1990)","Scale: NAKH900111<br>Name: Membrane proteins<br>Description: Transmembrane regions of non-mt-proteins (Nakashima et al., 1990)","Scale: NAKH900112<br>Name: Membrane proteins<br>Description: Transmembrane regions of mt-proteins (Nakashima et al., 1990)","Scale: NAKH920108<br>Name: Membrane proteins (multi-spanning)<br>Description: AA composition of MEM of multi-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH920105<br>Name: Membrane proteins (single-spanning)<br>Description: AA composition of MEM of single-spanning proteins (Nakashima-Nishikawa, 1992)","Scale: NAKH900103<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins (Nakashima et al., 1990)","Scale: NAKH900107<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins from fungi and plant (Nakashima et al., 1990)","Scale: FUKS010108<br>Name: Nuclear proteins (interior)<br>Description: Interior composition of amino acids in nuclear proteins [%] (Fukuchi-Nishikawa, 2001)","Scale: KUMS000104<br>Name: Proteins of mesophiles<br>Description: Distribution of amino acid residues in the alpha-helices in mesophilic proteins (Kumar et al., 2000)","Scale: FUKS010107<br>Name: Proteins of mesophiles (EXT)<br>Description: Interior composition of amino acids in extracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010106<br>Name: Proteins of mesophiles (INT)<br>Description: Interior composition of amino acids in intracellular proteins of mesophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: FUKS010105<br>Name: Proteins of thermophiles (INT)<br>Description: Interior composition of amino acids in intracellular proteins of thermophiles [%] (Fukuchi-Nishikawa, 2001)","Scale: NAKH900104<br>Name: Mitochondrial proteins<br>Description: Normalized composition of mt-proteins (Nakashima et al., 1990)","Scale: NAKH900105<br>Name: Mitochondrial proteins<br>Description: AA composition of mt-proteins from animal (Nakashima et al., 1990)","Scale: NAKH900106<br>Name: Mitochondrial proteins<br>Description: Normalized composition from animal (Nakashima et al., 1990)","Scale: NAKH900108<br>Name: Mitochondrial proteins<br>Description: Normalized composition from fungi and plant (Nakashima et al., 1990)","Scale: PUNT030101<br>Name: Membrane proteins (anchor)<br>Description: Knowledge-based membrane-propensity scale from 1D_Helix in MPtopo databases (Punta-Maritan, 2003)","Scale: PUNT030102<br>Name: Membrane proteins (anchor)<br>Description: Knowledge-based membrane-propensity scale from 3D_Helix in MPtopo databases (Punta-Maritan, 2003)","Scale: OOBM770103<br>Name: Non-bonded Energy per Atom<br>Description: Long range non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: GUYH850101<br>Name: Partition energy<br>Description: Partition energy (Guy, 1985)","Scale: GUYH850102<br>Name: Partition energy<br>Description: Apparent partition energies calculated from Wertz-Scheraga index (Guy, 1985)","Scale: MIYS990101<br>Name: Partition energy<br>Description: Relative partition energies derived by the Bethe approximation (Miyazawa-Jernigan, 1999)","Scale: MIYS990102<br>Name: Partition energy<br>Description: Optimized relative partition energies - method A (Miyazawa-Jernigan, 1999)","Scale: MIYS990103<br>Name: Partition energy<br>Description: Optimized relative partition energies - method B (Miyazawa-Jernigan, 1999)","Scale: MIYS990104<br>Name: Partition energy<br>Description: Optimized relative partition energies - method C (Miyazawa-Jernigan, 1999)","Scale: MIYS990105<br>Name: Partition energy<br>Description: Optimized relative partition energies - method D (Miyazawa-Jernigan, 1999)","Scale: BULH740101<br>Name: TFE to surface<br>Description: Transfer free energy to surface (Bull-Breese, 1974)","Scale: NAKH900113<br>Name: Average Composition<br>Description: Ratio of average and computed composition (Nakashima et al., 1990)","Scale: GRAR740101<br>Name: Substitution Frequency<br>Description: Composition (Grantham, 1974)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#0A9A9B","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(140,86,75,1)"}},"showlegend":true,"error_y":{"color":"rgba(140,86,75,1)"},"error_x":{"color":"rgba(140,86,75,1)"},"line":{"color":"rgba(140,86,75,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[10.667579396028895,7.3639068647293655,2.23967809825526,11.087389019782304,10.07584860084035,8.7379296360853225,-6.1485869973710257,-0.035309899062564647,-4.3550687128093646,-4.2101267861864864,-2.3634363830114045,-2.446780781911654,-3.4020434494912797,-6.0625009758975681,-8.8842631491387589,-8.9409877807518754,-2.5722841223531629,-7.0047146652419654,-7.154178024283631,-0.12908835162980861,-0.58741104330486271,-11.800874852120179,3.1321258925532418,0.59243733445464763,-5.9180837601883436,-2.6568129042296618,-1.3547751436888249,0.57049426368353906,-1.8121261171587966,-1.6196102270825063,1.7622816300709538,-0.92176990494137789,0.20357667029036977,-1.483930771409786,-0.59216292878353394,-2.0935570177470377,2.9638299683892515,-1.1622552255125964,0.59666124927736031,-1.715795418661251,0.60608339195705085,2.2406907443691049,1.644099191780517,-7.6358832281985798,-3.6848493795170634,-1.1820996976747253,4.3992449648569645,0.79172807901810893,0.58126944643496181,-0.87629318834367953,-1.9594981450601092,-0.50152806047582754,-11.219156561070617,-8.8317819753442652,-3.0397752802630795,4.8916642125181129,4.4113372844292842,-10.192813036511279,-10.286425076627893,-8.368704751995109,-7.7157124570696363,-7.5134044946170686,-4.3030111788828442,-5.1153184276233636,-0.044562956083341827,-1.7095922150288076,-0.92757590758327857,4.8437710082501759,1.3424831740137815,-1.5788315227010674,2.3727772696522402,1.5385073006128478,-0.52648150882325218,-0.66873966454552247,1.9528691234316051,-1.0177776720163494,1.1820897661603325,4.5985325600737168,-11.423009015427262,0.34381314855535799,4.3780179056682256,1.9887421111903947,-6.1071150560612866,-3.8253538623605507,-6.0796441960671475,-13.059732465690777,-12.734298679596744,-5.4631406882827518,-3.3038858984514334,-9.4928162645857377,-7.818696177643246,-7.8656137482814543,-11.98493821770505,-2.6064094134939908,-4.471980739473346,-5.7365323902929273,-6.3856918502644637,-6.9302821610978702,-13.256113782607436,-8.1921094720925112,-6.3304652177404135,-6.0129430358917757,-12.671027889195946,-7.9715157937622587,-13.669927318783813,15.768830374484169,-13.510954613471498,1.376730081148978],"y":[-6.6059182947894381,-9.1352621964204808,-10.925438324199018,-9.894273788864874,-11.56767042717472,-9.3747874231654507,-15.29613946440403,-12.928992227633403,-14.933162905338163,-14.755348415888468,-15.423877069930583,-15.346327011034568,-15.061257827734352,-13.481916390044189,-13.964723041534512,-13.36334381558561,-6.554156232495064,-12.552267561092751,-9.4766055354394467,14.749949678027646,21.440883646017259,10.088338302355977,17.618570138759171,18.617505664008334,19.528458539082582,16.363245731302353,16.559014386592789,18.966445045068699,15.532604842672159,18.106096133637937,20.593596290557297,18.989374759779814,22.369906717912514,18.956069350479556,18.801657357647418,15.215624785154413,18.812826888933259,16.706242397126328,18.621999589310963,15.815420946936623,16.076137099533472,19.099598467372328,19.483831610600166,17.179676731804776,-12.622998642339764,-13.446521043000519,-10.500497319358361,-13.471578584531937,-12.328538708315822,-13.138497017871297,-13.227562685013298,-12.822269729785127,-8.6966298881352682,-10.695284878271039,-10.62673431465686,-9.7880608132165623,-10.465071800528344,-11.044296820385391,-11.051821409373604,-13.033378603502554,-13.445227972414497,-13.700763361645018,-15.654298802145167,-12.944846426043206,-12.931081390311302,-7.9805564146554504,-7.642533587701549,-10.846274434193077,-7.0383426307193444,-9.3592783534256228,-8.2461156187225768,-8.6745938552009925,-10.202223094255848,-9.7054605598190822,-7.0958799910054138,-9.0619731791342151,-9.6082688922850128,-9.2596080398225009,-8.7447352567186503,-9.287296658345749,-10.355618835361067,-10.857516151086436,-14.994568547442562,-9.97430074640382,-7.9784373972275935,-9.2632740025192213,-9.4205290874830734,-14.424505156019556,-8.5459274626250341,-11.553275174003474,-14.134688124182382,-14.406156735845441,-8.051231616210929,-9.3153037693702547,-15.747202395736993,-15.409367785184342,-15.206976553471735,-14.848600556808417,-12.200360052310693,-15.900698546025307,-14.825647347886287,-16.157745059785114,-14.939364698800215,-14.637368224887316,-5.1691247091606156,12.200506059259764,-5.1718491276737826,1.493767631583895],"z":[-3.1195297898969825,-7.9461542509161616,-9.8401186826010179,-4.6462111209826382,-4.9710863015406419,-7.0367131089740536,-15.144359890130803,-17.159087148257669,-18.383558310837344,-18.621849705484511,-19.294322614474073,-19.241112066570626,-18.723937523348546,-15.376419529516694,-17.14032874848964,-16.620578602555259,-15.949134235686866,-16.549642465496596,-19.631228633630425,12.943980625174824,12.766051794532363,5.4992656267321909,12.896584187038549,13.764298612733974,12.484092156711466,12.661242821655506,10.046069590536312,10.952693059201994,11.208072325068233,14.97006746601466,12.416340932100198,16.132743585411564,12.657772347527734,14.750673004880232,14.507886486500373,10.441256974645517,13.092695708840312,9.9624084930916865,13.758293303648131,12.438804088496189,13.019985739680889,12.34658433343265,11.610919455049608,12.824033844920733,-15.481177281074508,-13.202632070897213,-14.422361541253984,-13.534582088133924,-13.566121445113628,-14.872523824622252,-14.824339555962618,-14.279473386313727,-15.626322872399758,-15.823956677286173,-15.211313123521212,-15.041258576049598,-14.40227430987218,-15.237638718409883,-15.28163291830511,-15.129786783615852,-14.471692364072167,-13.924901154022372,-15.792241359849294,-16.022641079008878,-17.162429171489645,-13.483566433920901,-12.917274263346069,-16.495659333077107,-15.569302942056519,-13.91523577729328,-16.061612491312907,-15.541590846146782,-13.279960912601148,-13.321365526944357,-15.689882842805385,-16.146288944362144,-13.679629040262268,-16.45272282042934,-15.086409842778979,-15.481198912274793,-15.196198396044041,-10.177476475336737,-9.6323823035307949,11.416246671428901,8.751645875579305,-15.363712912094742,-15.195972750928298,-13.77311014923664,-14.066400566280887,-14.136960560546759,-12.504663841151014,-11.741283578319267,-15.349708996217011,-15.957656736789701,-16.490433727372441,-16.976265970016669,-16.768102419565349,-16.621896819569375,-16.064211287518237,-17.829803611968305,-18.719285565981615,-15.498905729360828,-12.831532224140851,-10.65362049488226,-13.402770311056051,-3.7327430509861177,-13.80824469569176,17.99909361419726],"name":"Polarity","text":["Scale: MITS020101<br>Name: Amphiphilicity<br>Description: Amphiphilicity index (Mitaku et al., 2002)","Scale: GARJ730101<br>Name: Partition coefficient<br>Description: Partition Coefficient (Garel et al., 1973)","Scale: ZASB820101<br>Name: Partition coefficient<br>Description: Dependence of Partition Coefficient on ionic strength (Zaslavsky et al., 1982)","Scale: CHAM820101<br>Name: Polarizability<br>Description: Polarizability parameter (Charton-Charton, 1982)","Scale: MCMT640101<br>Name: Refractivity<br>Description: Refractivity (McMeekin et al., 1964), Cited by Jones (1975)","Scale: LEVM760107<br>Name: van der Waals parameter epsilon<br>Description: van der Waals parameter epsilon (Levitt, 1976)","Scale: CORJ870101<br>Name: Amphiphilicity (α-helix)<br>Description: NNEIG index (Cornette et al., 1987)","Scale: CORJ870102<br>Name: Amphiphilicity (α-helix)<br>Description: SWEIG index (Cornette et al., 1987)","Scale: CORJ870103<br>Name: Amphiphilicity (α-helix)<br>Description: PRIFT index (Cornette et al., 1987)","Scale: CORJ870104<br>Name: Amphiphilicity (α-helix)<br>Description: PRILS index (Cornette et al., 1987)","Scale: CORJ870105<br>Name: Amphiphilicity (α-helix)<br>Description: ALTFT index (Cornette et al., 1987)","Scale: CORJ870106<br>Name: Amphiphilicity (α-helix)<br>Description: ALTLS index (Cornette et al., 1987)","Scale: CORJ870107<br>Name: Amphiphilicity (α-helix)<br>Description: TOTFT index (Cornette et al., 1987)","Scale: ROSG850102<br>Name: Mean fractional area loss<br>Description: Mean fractional area loss (Rose et al., 1985)","Scale: DESM900101<br>Name: Membrane preference<br>Description: Membrane preference for cytochrome b: MPH89 (Degli Esposti et al., 1990)","Scale: DESM900102<br>Name: Membrane preference<br>Description: Average membrane preference: AMP07 (Degli Esposti et al., 1990)","Scale: NAKH900110<br>Name: Membrane preference<br>Description: Normalized composition of Membrane proteins (Nakashima et al., 1990)","Scale: MEIH800103<br>Name: Side chain angle (phi)<br>Description: Average side chain orientation angle (Meirovitch et al., 1980)","Scale: ARGP820102<br>Name: Signal sequence helical potential<br>Description: Signal sequence helical potential (Argos et al., 1982)","Scale: HOPA770101<br>Name: Hydration Number<br>Description: Hydration Number (Hopfinger, 1971), Cited by Charton-Charton (1982)","Scale: OOBM770101<br>Name: Non-bonded energy per atom<br>Description: Average non-bonded energy per atom (Oobatake-Ooi, 1977)","Scale: OOBM850104<br>Name: Non-bonded energy per atom<br>Description: Optimized average non-bonded energy per atom (Oobatake et al., 1985)","Scale: FAUJ880110<br>Name: Number full non-bonding orbitals<br>Description: Number of full nonbonding orbitals (Fauchere et al., 1988)","Scale: ENGD860101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity index (Engelman et al., 1986)","Scale: FASG890101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity index (Fasman, 1989)","Scale: GRAR740102<br>Name: Polarity (hydrophilicity)<br>Description: Polarity (Grantham, 1974)","Scale: HOPT810101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophilicity value (Hopp-Woods, 1981)","Scale: KIDA850101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity-related index (Kidera et al., 1985)","Scale: KOEH090101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Wimley-White, 1996), Cited by Koehler et al. (2009)","Scale: KOEH090102<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Hessa et al., 2007), Cited by Koehler et al. (2009)","Scale: KOEH090103<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Eisenberg-Weiss, 1984), Inverted to match the direction of transfer from water to bilayer by Koehler et al. (2009)","Scale: KOEH090104<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Kyte-Doolittle, 1982), Inverted to match the direction of transfer from water to bilayer by Koehler et al. (2009)","Scale: KOEH090106<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale (Janin, 1979), Cited by Koehler et al. (2009)","Scale: KOEH090109<br>Name: Polarity (hydrophilicity)<br>Description: Unified Hydrophobicity Scale, UHS  (Koehler et al., 2009)","Scale: KOEH090110<br>Name: Polarity (hydrophilicity)<br>Description: Mammalian Hydrophobicity Scale, MHS (Koehler et al., 2009)","Scale: KOEH090111<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity scale transition between water-interface from (Wimley-White, 1998), Cited by Koehler et al. (2009)","Scale: KUHL950101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophilicity scale (Kuhn et al., 1995)","Scale: LEVM760101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobic parameter (Levitt, 1976)","Scale: PRAM900101<br>Name: Polarity (hydrophilicity)<br>Description: Hydrophobicity (Prabhakaran, 1990)","Scale: WOEC730101<br>Name: Polarity (hydrophilicity)<br>Description: Polar requirement (Woese, 1973)","Scale: ZIMJ680103<br>Name: Polarity (hydrophilicity)<br>Description: Polarity (Zimmerman et al., 1968)","Scale: ROSM880101<br>Name: Side chain hydropathy<br>Description: Side chain hydropathy, uncorrected for solvation (Roseman, 1988)","Scale: ROSM880102<br>Name: Side chain hydropathy<br>Description: Side chain hydropathy, corrected for solvation (Roseman, 1988)","Scale: CORJ870108<br>Name: TOTLS index<br>Description: TOTLS index (Cornette et al., 1987)","Scale: MIYS850101<br>Name: Effective partition energy<br>Description: Effective partition energy (Miyazawa-Jernigan, 1985)","Scale: ROBB790101<br>Name: Hydration free energy<br>Description: Hydration free energy (Robson-Osguthorpe, 1979)","Scale: ARGP820101<br>Name: Hydrophobicity<br>Description: Hydrophobicity index (Argos et al., 1982)","Scale: CIDH920101<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha-proteins (Cid et al., 1992)","Scale: CIDH920102<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for beta-proteins (Cid et al., 1992)","Scale: CIDH920103<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha+beta-proteins (Cid et al., 1992)","Scale: CIDH920104<br>Name: Hydrophobicity<br>Description: Normalized hydrophobicity scales for alpha/beta-proteins (Cid et al., 1992)","Scale: CIDH920105<br>Name: Hydrophobicity<br>Description: Normalized average hydrophobicity scales (Cid et al., 1992)","Scale: EISD840101<br>Name: Hydrophobicity<br>Description: Consensus normalized hydrophobicity scale (Eisenberg, 1984)","Scale: EISD860103<br>Name: Hydrophobicity<br>Description: Direction of hydrophobic moment (Eisenberg-McLachlan, 1986)","Scale: FAUJ830101<br>Name: Hydrophobicity<br>Description: Hydrophobic parameter π (Fauchere-Pliska, 1983)","Scale: GOLD730101<br>Name: Hydrophobicity<br>Description: Hydrophobicity factor (Goldsack-Chalifoux, 1973)","Scale: JOND750101<br>Name: Hydrophobicity<br>Description: Hydrophobicity (Jones, 1975)","Scale: JURD980101<br>Name: Hydrophobicity<br>Description: Modified Kyte-Doolittle hydrophobicity scale (Juretic et al., 1998)","Scale: KYTJ820101<br>Name: Hydrophobicity<br>Description: Hydropathy index (Kyte-Doolittle, 1982)","Scale: NADH010102<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (9% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010103<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (16% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010104<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (20% accessibility) (Naderi-Manesh et al., 2001)","Scale: PONP930101<br>Name: Hydrophobicity<br>Description: Hydrophobicity scales (Ponnuswamy, 1993)","Scale: RADA880108<br>Name: Hydrophobicity<br>Description: Mean polarity (Radzicka-Wolfenden, 1988)","Scale: SWER830101<br>Name: Hydrophobicity<br>Description: Optimal matching hydrophobicity (Sweet-Eisenberg, 1983)","Scale: WILM950101<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C18 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: WILM950102<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C8 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: ZIMJ680101<br>Name: Hydrophobicity<br>Description: Hydrophobicity (Zimmerman et al., 1968)","Scale: BROC820101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in TFA (Browne et al., 1982)","Scale: GUOD860101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient at pH 2 (Guo et al., 1986)","Scale: MEEJ800101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in HPLC, pH7.4 (Meek, 1980)","Scale: MEEJ800102<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in HPLC, pH2.1 (Meek, 1980)","Scale: MEEJ810101<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in NaClO4 (Meek-Rossetti, 1981)","Scale: MEEJ810102<br>Name: Hydrophobicity (retention coefficient)<br>Description: Retention Coefficient in NaH2PO4 (Meek-Rossetti, 1981)","Scale: BROC820102<br>Name: Retention coefficient<br>Description: Retention Coefficient in HFBA (Browne et al., 1982)","Scale: EISD860101<br>Name: Solvation free energy<br>Description: Solvation free energy (Eisenberg-McLachlan, 1986)","Scale: NOZY710101<br>Name: Transfer free energy (TFE)<br>Description: Transfer energy, organic solvent/water (Nozaki-Tanford, 1971)","Scale: LAWE840101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy, CHP/water (Lawson et al., 1984)","Scale: RADA880101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy from chx to wat (Radzicka-Wolfenden, 1988)","Scale: RADA880102<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy from oct to wat (Radzicka-Wolfenden, 1988)","Scale: SIMZ760101<br>Name: Transfer free energy (TFE) to outside<br>Description: Transfer free energy (Simon, 1976), Cited by Charton-Charton (1982)","Scale: WIMW960101<br>Name: Transfer free energy (TFE) to outside<br>Description: Free energies of transfer of AcWl-X-LL peptides from bilayer interface to water (Wimley-White, 1996)","Scale: CHAM820102<br>Name: Free energy (interface)<br>Description: Free energy of solution in water, kcal/mole (Charton-Charton, 1982)","Scale: KOEH090112<br>Name: Hydrophobicity (interface)<br>Description: Hydrophobicity scale transition between water-interface, Unified Hydrophobicity Scale interface, UHSint  (Koehler et al., 2009)","Scale: WILM950103<br>Name: Hydrophobicity (interface)<br>Description: Hydrophobicity Coefficient in RP-HPLC, C4 with 0.1%TFA/MeCN/H2O (Wilce et al., 1995)","Scale: WOLR790101<br>Name: Hydration potential<br>Description: Hydrophobicity index (Wolfenden et al., 1979)","Scale: WOLR810101<br>Name: Hydration potential<br>Description: Hydration potential (Wolfenden et al., 1981)","Scale: CASG920101<br>Name: Hydrophobicity<br>Description: Hydrophobicity scale from native protein structures (Casari-Sippl, 1992)","Scale: COWR900101<br>Name: Hydrophobicity<br>Description: Hydrophobicity index, 3.0 pH (Cowan-Whittaker, 1990)","Scale: NADH010101<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (5% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010105<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (25% accessibility) (Naderi-Manesh et al., 2001)","Scale: NADH010106<br>Name: Hydrophobicity<br>Description: Hydropathy scale based on self-information values in the two-state model (36% accessibility) (Naderi-Manesh et al., 2001)","Scale: JACR890101<br>Name: Hydrophobicity (surrounding)<br>Description: Weights from the IFH scale (Jacobs-White, 1989)","Scale: BLAS910101<br>Name: Side chain hydrophobicity<br>Description: Scaled side chain hydrophobicity values (Black-Mould, 1991)","Scale: MANP780101<br>Name: Surrounding hydrophobicity<br>Description: Average surrounding hydrophobicity (Manavalan-Ponnuswamy, 1978)","Scale: PONP800101<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in folded form (Ponnuswamy et al., 1980)","Scale: PONP800102<br>Name: Surrounding hydrophobicity<br>Description: Average gain in surrounding hydrophobicity (Ponnuswamy et al., 1980)","Scale: PONP800103<br>Name: Surrounding hydrophobicity<br>Description: Average gain ratio in surrounding hydrophobicity (Ponnuswamy et al., 1980)","Scale: PONP800104<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in alpha-helix (Ponnuswamy et al., 1980)","Scale: PONP800105<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in beta-sheet (Ponnuswamy et al., 1980)","Scale: PONP800106<br>Name: Surrounding hydrophobicity<br>Description: Surrounding hydrophobicity in turn (Ponnuswamy et al., 1980)","Scale: PONP800108<br>Name: Surrounding hydrophobicity<br>Description: Average number of surrounding residues (Ponnuswamy et al., 1980)","Scale: WILM950104<br>Name: Hydrophobicity<br>Description: Hydrophobicity Coefficient in RP-HPLC, C18 with 0.1%TFA/2-PrOH/MeCN/H2O (Wilce et al., 1995)","Scale: NADH010107<br>Name: Hydrophobicity (based on self-information values)<br>Description: Hydropathy scale based on self-information values in the two-state model (50% accessibility) (Naderi-Manesh et al., 2001)","Scale: FASG760105<br>Name: pK-C<br>Description: pK-C (Fasman, 1976)","Scale: FAUJ880113<br>Name: pK-C<br>Description: pK-a(RCOOH) (Fauchere et al., 1988)","Scale: JOND750102<br>Name: pK-C<br>Description: pK (-COOH) (Jones, 1975)","Scale: FASG760104<br>Name: pK-N<br>Description: pK-N (Fasman, 1976)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#1958C3","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(227,119,194,1)"}},"showlegend":true,"error_y":{"color":"rgba(227,119,194,1)"},"error_x":{"color":"rgba(227,119,194,1)"},"line":{"color":"rgba(227,119,194,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[11.860395566992791,7.759352326939319,7.8021543790600223,13.624357804838741,14.951301966814789,-9.8686317336438236,-6.8270422152026473,-4.9241094776331353,-6.2738139297293589,-8.9855066885007933,-7.4821575139102041,-7.6657846314481279,-8.9845856915994311,-7.4784023628624858,-6.4463972945571504,-8.0386519402449608,-6.0003351445754207,-5.8771786342223251,-2.9303736605299866,-2.3723167474776194,3.3629301507797047,0.77399946665783226,6.3167260331114718,2.3288953779339292,-0.38176390204479088,6.9249647204173925,22.923547280962751,19.893809388761429,17.564991129757807,19.308816956470384,1.2251133528024756],"y":[-9.6880350448847548,-9.7442417479244963,-9.7665693515536258,13.291786193029477,12.061950966281829,18.920180710977856,16.103890546700416,18.297993113601617,17.715388540527908,16.599263113228602,16.859842701458597,19.841028802561475,18.806290838380924,21.10364019110791,20.45128279514514,20.806786788854719,17.342000847240879,21.619806459826336,22.443847828880102,-12.871636919853218,-9.8302193570185512,-11.046080143623737,-12.138373351056794,-18.184445624587141,-19.212980292276292,-16.948686764106487,7.8633156584288946,7.963449541576316,9.8008724886762515,8.3997966847214798,-6.6729706888639946],"z":[2.2804992330617786,3.8111481226043606,3.7664131564800334,-2.2400855942619851,-2.5714495855287076,11.158482437948109,9.9794583524735696,8.3058587156653587,9.8824046108431407,10.605988803594858,9.9675853994482679,10.691545908776481,10.955050317385657,10.089615145145467,10.674438064824116,9.0953370002677563,6.2922623109148468,11.566543488341548,13.678612155959172,-13.62206024828769,-14.73572447640146,-12.887157390437533,-10.774594299666445,-19.965290390823363,-17.145861263026593,-15.109397775526272,-1.5627632305318744,-2.5084433425442083,-2.491269274441223,-2.8968439838897857,-10.144176534187025],"name":"Structure-Activity","text":["Scale: ROSM880103<br>Name: Loss of hydropathy by α-helix formation<br>Description: Loss of Side chain hydropathy by helix formation (Roseman, 1988)","Scale: ANDN920101<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: alpha-CH chemical shifts (Andersen et al., 1992)","Scale: BUNA790102<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: alpha-CH chemical shifts (Bundi-Wuthrich, 1979)","Scale: BUNA790103<br>Name: Spin-spin coupling<br>Description: Spin-spin coupling constants 3JHalpha-NH (Bundi-Wuthrich, 1979)","Scale: BUNA790101<br>Name: α-NH chemical shifts (backbone-dynamics)<br>Description: alpha-NH chemical shifts (Bundi-Wuthrich, 1979)","Scale: BHAR880101<br>Name: Flexibility<br>Description: Average flexibility indices (Bhaskaran-Ponnuswamy, 1988)","Scale: PARS000101<br>Name: Flexibility<br>Description: p-Values of mesophilic proteins based on the distributions of B values (Parthasarathy-Murthy, 2000)","Scale: PARS000102<br>Name: Flexibility<br>Description: p-Values of thermophilic proteins based on the distributions of B values (Parthasarathy-Murthy, 2000)","Scale: VINM940101<br>Name: Flexibility<br>Description: Normalized flexibility parameters (B-values), average (Vihinen et al., 1994)","Scale: KARP850101<br>Name: Flexibility (0 rigid neighbors)<br>Description: Flexibility parameter for no rigid neighbors (Karplus-Schulz, 1985)","Scale: VINM940102<br>Name: Flexibility (0 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by none rigid neighbours (Vihinen et al., 1994)","Scale: KARP850102<br>Name: Flexibility (1 rigid neighbors)<br>Description: Flexibility parameter for one rigid neighbor (Karplus-Schulz, 1985)","Scale: VINM940103<br>Name: Flexibility (1 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by one rigid neighbours (Vihinen et al., 1994)","Scale: KRIW710101<br>Name: Side chain interaction<br>Description: Side chain interaction parameter (Krigbaum-Rubin, 1971)","Scale: KRIW790101<br>Name: Side chain interaction<br>Description: Side chain interaction parameter (Krigbaum-Komoriya, 1979)","Scale: OOBM850105<br>Name: Side chain interaction<br>Description: Optimized side chain interaction parameter (Oobatake et al., 1985)","Scale: KARP850103<br>Name: Flexibility (2 rigid neighbors)<br>Description: Flexibility parameter for two rigid neighbors (Karplus-Schulz, 1985)","Scale: KRIW790102<br>Name: Flexibility (2 rigid neighbors)<br>Description: Fraction of site occupied by water (Krigbaum-Komoriya, 1979)","Scale: VINM940104<br>Name: Flexibility (2 rigid neighbors)<br>Description: Normalized flexibility parameters (B-values) for each residue surrounded by two rigid neighbours (Vihinen et al., 1994)","Scale: BASU050102<br>Name: Interactivity<br>Description: Interactivity scale obtained by maximizing the mean of correlation Coefficient over single-domain globular proteins (Bastolla et al., 2005)","Scale: TAKK010101<br>Name: Stability<br>Description: Side-chain contribution to protein stability (kJ/mol) (Takano-Yutani, 2001)","Scale: ZHOH040101<br>Name: Stability<br>Description: The stability scale from the knowledge-based atom-atom potential (Zhou-Zhou, 2004)","Scale: ZHOH040102<br>Name: Stability<br>Description: The relative stability scale extracted from mutation experiments (Zhou-Zhou, 2004)","Scale: OOBM850101<br>Name: Stability (extended-coil)<br>Description: Optimized beta-structure-coil equilibrium constant (Oobatake et al., 1985)","Scale: PTIO830102<br>Name: Stability (extended-coil)<br>Description: Beta-coil equilibrium constant (Ptitsyn-Finkelstein, 1983)","Scale: FAUJ880107<br>Name: α-CH chemical shifts (backbone-dynamics)<br>Description: N.m.r. chemical shift of alpha-carbon (Fauchere et al., 1988)","Scale: GEOR030108<br>Name: Linker (helical)<br>Description: Linker propensity from helical (annotated by DSSP) dataset (George-Heringa, 2003)","Scale: FINA770101<br>Name: Stability (helix-coil)<br>Description: Helix-coil equilibrium constant (Finkelstein-Ptitsyn, 1977)","Scale: PTIO830101<br>Name: Stability (helix-coil)<br>Description: Helix-coil equilibrium constant (Ptitsyn-Finkelstein, 1983)","Scale: SUEM840101<br>Name: Stability (helix-coil)<br>Description: Zimm-Bragg parameter s at 20 C (Sueki et al., 1984)","Scale: SUEM840102<br>Name: Stability (extended-coil)<br>Description: Zimm-Bragg parameter sigma x 1.0E4 (Sueki et al., 1984)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#8000D3","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(127,127,127,1)"}},"showlegend":true,"error_y":{"color":"rgba(127,127,127,1)"},"error_x":{"color":"rgba(127,127,127,1)"},"line":{"color":"rgba(127,127,127,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[-10.854216021237971,-10.891434007654006,-13.497972081250296,16.831298658653168,1.0966927454935129,-14.328562893278539,13.201738393336424,8.7802634608591479,-6.0643442702343346,-8.9094121166936571,-6.0350332288594259,-5.4887391359184186,1.1494957056235684,8.4839519522050537,0.81682601930706067,15.060281091795344,16.022377601865948],"y":[14.366338622340853,14.333775541690784,9.5564899738609341,6.092995007957275,-8.1124210801257419,-9.8809162412135496,-9.2194763508386863,-8.7748658265978516,-7.5006443081937588,-6.439708492881361,15.703522794781156,15.32783656392207,-9.1124406053928038,12.64912783249566,-5.0534316111940862,12.602055016031732,11.564946476092651],"z":[6.5699597824712157,6.5267751833532746,4.0358070771851704,-4.3678366748741304,-16.151719752717639,-14.623239099735706,-0.80364256430688907,-5.2578466437387927,9.3621079226324824,10.414116146355942,13.141763936480375,13.435874768998039,-13.012804003931601,3.9402015539333126,-14.581231136880838,-3.780474289472894,-2.6029999332407781],"name":"Others","text":["Scale: DAYM780201<br>Name: Mutability<br>Description: Relative mutability (Dayhoff et al., 1978b)","Scale: JOND920102<br>Name: Mutability<br>Description: Relative mutability (Jones et al., 1992)","Scale: WEBA780101<br>Name: RF value<br>Description: RF value in high salt chromatography (Weber-Lacey, 1978)","Scale: SNEP660101<br>Name: Principal Component 1 (Sneath)<br>Description: Principal component I (Sneath, 1966)","Scale: ZIMJ680105<br>Name: Principal Component 1 (Zimmerman)<br>Description: RF rank (Zimmerman et al., 1968)","Scale: SNEP660102<br>Name: Principal Component 2 (Sneath)<br>Description: Principal component II (Sneath, 1966)","Scale: WOLS870102<br>Name: Principal Component 2 (Wold)<br>Description: Principal property value z2 (Wold et al., 1987)","Scale: SNEP660103<br>Name: Principal Component 3 (Sneath)<br>Description: Principal component III (Sneath, 1966)","Scale: WOLS870103<br>Name: Principal Component 3 (Wold)<br>Description: Principal property value z3 (Wold et al., 1987)","Scale: SNEP660104<br>Name: Principal Component 4 (Sneath)<br>Description: Principal component IV (Sneath, 1966)","Scale: PARJ860101<br>Name: HPLC parameter<br>Description: HPLC parameter (Parker et al., 1986)","Scale: WOLS870101<br>Name: Principal Component 1 (Wold)<br>Description: Principal property value z1 (Wold et al., 1987)","Scale: VENT840101<br>Name: Bitterness<br>Description: Bitterness (Venanzi, 1984)","Scale: KHAG800101<br>Name: Kerr-constant<br>Description: The Kerr-constant increments (Khanarian-Moore, 1980)","Scale: FASG760102<br>Name: Melting point<br>Description: Melting Point (Fasman, 1976)","Scale: FASG760103<br>Name: Optical rotation<br>Description: Optical rotation (Fasman, 1976)","Scale: ONEK900101<br>Name: ΔG values in peptides<br>Description: Delta G values for the peptides extrapolated to 0 M urea (O'Neil-DeGrado, 1990)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#999999","size":4,"opacity":0.69999999999999996,"symbol":"circle","line":{"color":"rgba(188,189,34,1)"}},"showlegend":true,"error_y":{"color":"rgba(188,189,34,1)"},"error_x":{"color":"rgba(188,189,34,1)"},"line":{"color":"rgba(188,189,34,1)"},"frame":null},{"mode":"markers","type":"scatter3d","x":[-6.3863391037695489,-14.77481958677191,13.449055111856786,-5.7676751372295616,5.3808817953649184,-16.519397623670603,18.462945942463403,2.8166701516351687,4.4365733607322735,5.8341875427404046,4.2797242715650876,-5.8693743286430529],"y":[18.220941143134386,7.0003932064910721,11.298525853024117,-6.2458174113031513,10.505872697178841,-1.6350199906414125,15.539146117662201,-23.689163192294821,12.287290195383211,-6.9568358242513089,3.144273397234572,-14.015901969600678],"z":[15.531883111461823,2.5937725633475912,-2.2940381550516444,9.4355107571757237,12.464930621238519,10.834609581021821,2.1342753505092573,-13.601014044663255,15.549388977765394,6.697225423355591,8.3672408440712562,-3.9524142102396271],"text":["Feature-space axis PC1 (my.PC1)","Feature-space axis PC2 (my.PC2)","Feature-space axis PC3 (my.PC3)","Feature-space axis PC4 (my.PC4)","Feature-space axis PC5 (my.PC5)","Feature-space axis PC6 (my.PC6)","Feature-space axis PC7 (my.PC7)","Feature-space axis PC8 (my.PC8)","Feature-space axis PC9 (my.PC9)","Feature-space axis PC10 (my.PC10)","Feature-space axis PC11 (my.PC11)","Feature-space axis PC12 (my.PC12)"],"hoverinfo":["text","text","text","text","text","text","text","text","text","text","text","text"],"marker":{"color":"#ff295f","size":6,"symbol":"square","line":{"color":"rgba(23,190,207,1)"}},"name":"Feature-space PCs","showlegend":true,"error_y":{"color":"rgba(23,190,207,1)"},"error_x":{"color":"rgba(23,190,207,1)"},"line":{"color":"rgba(23,190,207,1)"},"frame":null},{"mode":"text","type":"scatter3d","x":[-6.3863391037695489,-14.77481958677191,13.449055111856786,-5.7676751372295616,5.3808817953649184,-16.519397623670603,18.462945942463403,2.8166701516351687,4.4365733607322735,5.8341875427404046,4.2797242715650876,-5.8693743286430529],"y":[18.220941143134386,7.0003932064910721,11.298525853024117,-6.2458174113031513,10.505872697178841,-1.6350199906414125,15.539146117662201,-23.689163192294821,12.287290195383211,-6.9568358242513089,3.144273397234572,-14.015901969600678],"z":[15.531883111461823,2.5937725633475912,-2.2940381550516444,9.4355107571757237,12.464930621238519,10.834609581021821,2.1342753505092573,-13.601014044663255,15.549388977765394,6.697225423355591,8.3672408440712562,-3.9524142102396271],"text":[1,2,3,4,5,6,7,8,9,10,11,12],"textposition":["top center","top center","top center","top center","top center","top center","top center","top center","top center","top center","top center","top center"],"textfont":{"color":"#ff295f","size":12},"showlegend":false,"marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"error_y":{"color":"rgba(31,119,180,1)"},"error_x":{"color":"rgba(31,119,180,1)"},"line":{"color":"rgba(31,119,180,1)"},"frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p><strong>Figure:</strong> 3D t-SNE of amino acid scales and Feature Space PCs.</p>
</div>
</div>
<p>You can rotate the plot to see the 3d relationships, as well as zoom in and out with “pinching” gestures. In some aspects the plot is similar to the 2d plot - after all, it displays the same neighbour-relationships. But much overlap is resolved, and it is easier to see which of the source scales best represent the PCs.</p>
<p>A note on coding these plots: this entire visualization was developed with ChatGPT-5. A quick visualization was easy to obtain, but getting it exactly right took quite a bit of back-and-forth. We are making extensive use of the <code>plotly::</code> package, which I haven’t used very much at all, so my knowledge of the R language did not help that much. It was very much a process of explaining what I want, discussing strategies and alternatives to get it, patiently revising, refining, copying and pasting error messages, copying and pasting lots of code to ensure we are talking about the same thing etc. Throughout, I had the feeling to be pair-programming with a competent colleague. This is absolutely something that you should be able to do even as a programming novice, and that you need to practice. You need language skills, and structured thinking, not coding skills.</p>
<p>The result however is quite convincing, it presents exactly the right kind of “best practicy” that make data analysis meaningful, to unravel relationships between facts and interpret them.</p>
<p>We are ready to move on: use our feature space to define a function with which we can compute “similarity” between amino acid pairs.</p>
</section>
</section>
</section>
<section id="amino-acid-similarity---the-function-aasim" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Amino Acid Similarity - the Function <code>aaSim()</code></h1>
<p>We are now ready to define:</p>
<blockquote class="blockquote">
<p><strong>The similarity between two amino acids is the Euclidian distance between their positions in an amino acid feature space.</strong></p>
</blockquote>
<p>This maps the diverse properties of a pair of amino acids to a single number.</p>
<p>I construct this function as a “closure” - an R specialty which allows us to package code and data into a convenient unit. You can skip this section if you are only interested in the result.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>  aaSimConstructor <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Value: a function that computes pairwise amino acid distances.</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note:</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This function returns another function as its output. The returned</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># function is a so-called "closure" - a combination of data in its</span></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># environment, and code instructions. By doing this, the auxiliary data for</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># our function is created only once, when the function is defined, not every</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># time it is called. Yet, since the objects are defined locally, we do not</span></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># risk overwriting objects in our workspace when we define the function.</span></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This constructor function does not use external parameters that are </span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># "passed in" in the function call, but has the information</span></span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># it needs written into the function body.</span></span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># === Parameters ==============================</span></span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Feature Space object:</span></span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a>    SPACEFILE <span class="ot">&lt;-</span> <span class="st">"dat/aaFeatureSpace.4.1.Rds"</span></span>
<span id="cb68-21"><a href="#cb68-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span> <span class="fu">file.exists</span>(SPACEFILE)) {</span>
<span id="cb68-22"><a href="#cb68-22" aria-hidden="true" tabindex="-1"></a>      GHurl  <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"https://raw.githubusercontent.com/hyginn/CSB195/main/"</span>,</span>
<span id="cb68-23"><a href="#cb68-23" aria-hidden="true" tabindex="-1"></a>                         SPACEFILE)</span>
<span id="cb68-24"><a href="#cb68-24" aria-hidden="true" tabindex="-1"></a>        <span class="fu">download.file</span>(<span class="at">url =</span> GHurl, <span class="at">destfile =</span> SPACEFILE, <span class="at">mode =</span> <span class="st">"wb"</span>)</span>
<span id="cb68-25"><a href="#cb68-25" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb68-26"><a href="#cb68-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-27"><a href="#cb68-27" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Recreate the Feature Space</span></span>
<span id="cb68-28"><a href="#cb68-28" aria-hidden="true" tabindex="-1"></a>    AASPACE <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(SPACEFILE)</span>
<span id="cb68-29"><a href="#cb68-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-30"><a href="#cb68-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Stop codon distance:</span></span>
<span id="cb68-31"><a href="#cb68-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The distance of an amino acid to a stop codon is STOPDIST times the</span></span>
<span id="cb68-32"><a href="#cb68-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># maximum distance in the distance matrix.</span></span>
<span id="cb68-33"><a href="#cb68-33" aria-hidden="true" tabindex="-1"></a>    STOPDIST <span class="ot">&lt;-</span> <span class="fl">1.5</span></span>
<span id="cb68-34"><a href="#cb68-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-35"><a href="#cb68-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Compute a 21 x 21 matrix of Euclidian distances between any pair of vectors</span></span>
<span id="cb68-36"><a href="#cb68-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># in AASPACE.</span></span>
<span id="cb68-37"><a href="#cb68-37" aria-hidden="true" tabindex="-1"></a>    AADMAT <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">numeric</span>(<span class="dv">21</span> <span class="sc">*</span> <span class="dv">21</span>), <span class="at">nrow =</span> <span class="dv">21</span>)</span>
<span id="cb68-38"><a href="#cb68-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Fill the first 20 x 20 values with amino acid pair distances</span></span>
<span id="cb68-39"><a href="#cb68-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>) {</span>
<span id="cb68-40"><a href="#cb68-40" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>) {</span>
<span id="cb68-41"><a href="#cb68-41" aria-hidden="true" tabindex="-1"></a>        AADMAT[i, j] <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">sum</span>((AASPACE[i, ] <span class="sc">-</span> AASPACE[j, ])<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb68-42"><a href="#cb68-42" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb68-43"><a href="#cb68-43" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb68-44"><a href="#cb68-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-45"><a href="#cb68-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define distance to stop codons: distance of stop codon "*" to</span></span>
<span id="cb68-46"><a href="#cb68-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># any other codon as STOPDIST times the maximum distance in the</span></span>
<span id="cb68-47"><a href="#cb68-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># distance matrix.</span></span>
<span id="cb68-48"><a href="#cb68-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-49"><a href="#cb68-49" aria-hidden="true" tabindex="-1"></a>    stopDist <span class="ot">&lt;-</span> STOPDIST <span class="sc">*</span> <span class="fu">max</span>(AADMAT)</span>
<span id="cb68-50"><a href="#cb68-50" aria-hidden="true" tabindex="-1"></a>    AADMAT[ <span class="dv">21</span>, ] <span class="ot">&lt;-</span> stopDist</span>
<span id="cb68-51"><a href="#cb68-51" aria-hidden="true" tabindex="-1"></a>    AADMAT[ , <span class="dv">21</span>] <span class="ot">&lt;-</span> stopDist</span>
<span id="cb68-52"><a href="#cb68-52" aria-hidden="true" tabindex="-1"></a>    AADMAT[<span class="dv">21</span>,<span class="dv">21</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb68-53"><a href="#cb68-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-54"><a href="#cb68-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(AADMAT) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rownames</span>(AASPACE), <span class="st">"*"</span>)</span>
<span id="cb68-55"><a href="#cb68-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(AADMAT) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rownames</span>(AASPACE), <span class="st">"*"</span>)</span>
<span id="cb68-56"><a href="#cb68-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-57"><a href="#cb68-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define a function to return the pairwise distance between two points</span></span>
<span id="cb68-58"><a href="#cb68-58" aria-hidden="true" tabindex="-1"></a>    <span class="co"># in the distance space</span></span>
<span id="cb68-59"><a href="#cb68-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-60"><a href="#cb68-60" aria-hidden="true" tabindex="-1"></a>    <span class="co">#  The  function takes as its input two amino acid</span></span>
<span id="cb68-61"><a href="#cb68-61" aria-hidden="true" tabindex="-1"></a>    <span class="co">#  one-letter symbols and returns the Euclidian distance between</span></span>
<span id="cb68-62"><a href="#cb68-62" aria-hidden="true" tabindex="-1"></a>    <span class="co">#  the two vectors in the feature space defined in SPACEFILE.</span></span>
<span id="cb68-63"><a href="#cb68-63" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb68-64"><a href="#cb68-64" aria-hidden="true" tabindex="-1"></a>    <span class="co">#  Parameters:</span></span>
<span id="cb68-65"><a href="#cb68-65" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    a1, a1: two letters from "ACDEFGHIKLMNPQRSTVWY*"</span></span>
<span id="cb68-66"><a href="#cb68-66" aria-hidden="true" tabindex="-1"></a>    <span class="co">#  Value:  distance between a1, a2</span></span>
<span id="cb68-67"><a href="#cb68-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-68"><a href="#cb68-68" aria-hidden="true" tabindex="-1"></a>    myFun <span class="ot">&lt;-</span> <span class="cf">function</span>(a1, a2) {</span>
<span id="cb68-69"><a href="#cb68-69" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(AADMAT[a1, a2])</span>
<span id="cb68-70"><a href="#cb68-70" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb68-71"><a href="#cb68-71" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb68-72"><a href="#cb68-72" aria-hidden="true" tabindex="-1"></a>    <span class="co"># attach the matrix and alphabet as parameters so we can inspect them</span></span>
<span id="cb68-73"><a href="#cb68-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">attr</span>(myFun, <span class="st">"AADMAT"</span>) <span class="ot">&lt;-</span> AADMAT</span>
<span id="cb68-74"><a href="#cb68-74" aria-hidden="true" tabindex="-1"></a>    <span class="fu">attr</span>(myFun, <span class="st">"alphabet"</span>) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(AADMAT)</span>
<span id="cb68-75"><a href="#cb68-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-76"><a href="#cb68-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(myFun)  <span class="co"># Return the function. This is the key move. This wraps up</span></span>
<span id="cb68-77"><a href="#cb68-77" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># the function's environment (i.e. the work space we just</span></span>
<span id="cb68-78"><a href="#cb68-78" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># created for it with the SPACEFILE) into the constructor</span></span>
<span id="cb68-79"><a href="#cb68-79" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># function's output.</span></span>
<span id="cb68-80"><a href="#cb68-80" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb68-81"><a href="#cb68-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-82"><a href="#cb68-82" aria-hidden="true" tabindex="-1"></a>  aaSim <span class="ot">&lt;-</span> <span class="fu">aaSimConstructor</span>()   <span class="co"># Define the function aaSim() so we can use it.</span></span>
<span id="cb68-83"><a href="#cb68-83" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(aaSimConstructor)          <span class="co"># The constructor function is no longer needed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="test-aasim-and-validate" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="test-aasim-and-validate"><span class="header-section-number">8.1</span> Test <code>aaSim()</code> and validate</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Our similarities induce a METRIC space</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"Q"</span>, <span class="st">"Q"</span>) <span class="co">#  IDENTITY: d(x,y) == 0: x == y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">any</span>(<span class="fu">attr</span>(aaSim, <span class="st">"AADMAT"</span>) <span class="sc">&lt;</span> <span class="dv">0</span>)  <span class="co"># FALSE; POSITIVITY: no d(x,y) &lt; 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"F"</span>, <span class="st">"I"</span>) <span class="co">#  Similar amino acids: distance is small.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 11.32606</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"Q"</span>, <span class="st">"F"</span>) <span class="co">#  Dissimilar amino acids: distance is large.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 27.83712</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"F"</span>, <span class="st">"Q"</span>) <span class="co">#  SYMMETRY: d(x, y) == d(y, x)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 27.83712</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"Q"</span>, <span class="st">"*"</span>) <span class="co">#  Distance between any amino acid and a stop codon is large.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 53.93576</code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"*"</span>, <span class="st">"*"</span>) <span class="co">#  Two stop codons: distance is zero.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"G"</span>, <span class="st">"C"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25.05452</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"C"</span>, <span class="st">"W"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25.90865</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"G"</span>, <span class="st">"W"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 35.95717</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aaSim</span>(<span class="st">"G"</span>, <span class="st">"C"</span>) <span class="sc">+</span> <span class="fu">aaSim</span>(<span class="st">"C"</span>, <span class="st">"W"</span>) <span class="sc">&lt;=</span> <span class="fu">aaSim</span>(<span class="st">"G"</span>, <span class="st">"W"</span>) <span class="co"># TRIANGLE INEQUALITY</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<section id="distinct-amino-acids" class="level3" data-number="8.1.1">
<h3 data-number="8.1.1" class="anchored" data-anchor-id="distinct-amino-acids"><span class="header-section-number">8.1.1</span> Distinct Amino Acids</h3>
<p>Let’s do a quick check of which amino acid is the most distinct, and which one is the most “plain vanilla” among the twenty. We develop this from pseudocode:</p>
<pre><code>For each amino acid
  For each of the 19 other amino acids
    Accumulate the distance of the amino acid pair
  Compute the mean
Sort the results and show them</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>  mySims <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="dv">20</span>)  <span class="co"># create an empty 20-element vector of numbers</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(mySims) <span class="ot">&lt;-</span> aaLabels</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#&gt; For each amino acid</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (thisA <span class="cf">in</span> aaLabels) {</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&gt;   For each of the 19 other amino acids</span></span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (otherA <span class="cf">in</span> aaLabels) {</span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>      <span class="co">#&gt;     Add the distance between the pair to the other distances</span></span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a>      mySims[thisA] <span class="ot">&lt;-</span> mySims[thisA] <span class="sc">+</span> <span class="fu">aaSim</span>(thisA, otherA)</span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb92-11"><a href="#cb92-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#&gt;   Compute the mean</span></span>
<span id="cb92-12"><a href="#cb92-12" aria-hidden="true" tabindex="-1"></a>    mySims[thisA] <span class="ot">&lt;-</span> mySims[thisA] <span class="sc">/</span> <span class="dv">20</span></span>
<span id="cb92-13"><a href="#cb92-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb92-14"><a href="#cb92-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-15"><a href="#cb92-15" aria-hidden="true" tabindex="-1"></a>  mySims <span class="ot">&lt;-</span> <span class="fu">sort</span>(mySims)                             <span class="co"># sort the vector </span></span>
<span id="cb92-16"><a href="#cb92-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">"</span><span class="sc">\t</span><span class="st">%s: %5.2f</span><span class="sc">\n</span><span class="st">"</span>, <span class="fu">names</span>(mySims), mySims)) <span class="co"># show the results</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    T: 17.07
    S: 18.02
    H: 18.26
    Q: 18.35
    A: 18.49
    N: 18.98
    M: 19.22
    V: 19.61
    L: 20.05
    Y: 20.10
    E: 20.75
    D: 21.22
    I: 21.68
    F: 21.68
    P: 22.09
    C: 22.27
    G: 23.71
    K: 23.75
    W: 24.61
    R: 25.20</code></pre>
</div>
</div>
<p>The most “distinct” amino acid is “R” (arginine), the most average amino acid is “T” (threonine).</p>
<p>Now plot this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">barplot</span>(mySims,</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">main =</span> <span class="st">"Amino Acid Similarity"</span>,</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylab =</span> <span class="st">"Mean pairwise distances in feature space (AU)"</span>,</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.0</span>, <span class="dv">30</span>),</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">col =</span> AACOLS[<span class="fu">names</span>(mySims)],</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">cex.names =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="computingAminoAcidSimilarity_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>… Done.</strong></p>
<!-- [END] -->
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>