<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Arjun Mehandiratta">
<meta name="dcterms.date" content="2025-10-10">

<title>Report 1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="Arjun_Mehandiratta_Report_1_files/libs/clipboard/clipboard.min.js"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/popper.min.js"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/anchor.min.js"></script>
<link href="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Arjun_Mehandiratta_Report_1_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Arjun_Mehandiratta_Report_1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Arjun_Mehandiratta_Report_1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Arjun_Mehandiratta_Report_1_files/libs/bootstrap/bootstrap-01837272e09a7ffd4015b00fc1d96172.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<style>
  .version-info {
    font-size: 70%;
    color: #BBBBBB;
    margin-top: -10px;
  }
  .quarto-code-block {
    background: #222;
    color: #f8f8f2;
    border-radius: 6px;
    padding: 1em;
    border: 1px solid #444;
    font-size: 1em;
  }
  .quarto-code-block code {
    background: none;
    color: inherit;
  }
</style>


</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview"><span class="header-section-number">1</span> Overview</a></li>
  <li><a href="#specification" id="toc-specification" class="nav-link" data-scroll-target="#specification"><span class="header-section-number">2</span> Specification</a></li>
  <li><a href="#setting-up-the-environment" id="toc-setting-up-the-environment" class="nav-link" data-scroll-target="#setting-up-the-environment"><span class="header-section-number">3</span> Setting Up the Environment</a>
  <ul class="collapse">
  <li><a href="#assumptions" id="toc-assumptions" class="nav-link" data-scroll-target="#assumptions"><span class="header-section-number">3.1</span> Assumptions</a></li>
  <li><a href="#some-constants" id="toc-some-constants" class="nav-link" data-scroll-target="#some-constants"><span class="header-section-number">3.2</span> Some constants</a></li>
  <li><a href="#importing-the-data" id="toc-importing-the-data" class="nav-link" data-scroll-target="#importing-the-data"><span class="header-section-number">3.3</span> Importing the data</a></li>
  <li><a href="#generator-and-scorer" id="toc-generator-and-scorer" class="nav-link" data-scroll-target="#generator-and-scorer"><span class="header-section-number">3.4</span> Generator and Scorer</a></li>
  <li><a href="#ensuring-baseline" id="toc-ensuring-baseline" class="nav-link" data-scroll-target="#ensuring-baseline"><span class="header-section-number">3.5</span> Ensuring Baseline</a></li>
  </ul></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="header-section-number">4</span> Data</a>
  <ul class="collapse">
  <li><a href="#comparing-random-codes-to-the-sgc" id="toc-comparing-random-codes-to-the-sgc" class="nav-link" data-scroll-target="#comparing-random-codes-to-the-sgc"><span class="header-section-number">4.1</span> Comparing Random Codes to the SGC</a></li>
  </ul></li>
  <li><a href="#optimization-extras" id="toc-optimization-extras" class="nav-link" data-scroll-target="#optimization-extras"><span class="header-section-number">5</span> Optimization <em>(extras)</em></a>
  <ul class="collapse">
  <li><a href="#random-optimization" id="toc-random-optimization" class="nav-link" data-scroll-target="#random-optimization"><span class="header-section-number">5.1</span> Random “Optimization”</a></li>
  <li><a href="#simulated-annealing" id="toc-simulated-annealing" class="nav-link" data-scroll-target="#simulated-annealing"><span class="header-section-number">5.2</span> Simulated Annealing</a></li>
  </ul></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions"><span class="header-section-number">6</span> Conclusions</a>
  <ul class="collapse">
  <li><a href="#limitations-of-the-experiment" id="toc-limitations-of-the-experiment" class="nav-link" data-scroll-target="#limitations-of-the-experiment"><span class="header-section-number">6.1</span> Limitations of the experiment</a></li>
  </ul></li>
  <li><a href="#appendix" id="toc-appendix" class="nav-link" data-scroll-target="#appendix"><span class="header-section-number">7</span> Appendix</a>
  <ul class="collapse">
  <li><a href="#ai-chat-links" id="toc-ai-chat-links" class="nav-link" data-scroll-target="#ai-chat-links"><span class="header-section-number">7.1</span> AI Chat Links</a></li>
  <li><a href="#preparing-the-report" id="toc-preparing-the-report" class="nav-link" data-scroll-target="#preparing-the-report"><span class="header-section-number">7.2</span> Preparing the report</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">7.3</span> References</a></li>
  <li><a href="#self-assessment" id="toc-self-assessment" class="nav-link" data-scroll-target="#self-assessment"><span class="header-section-number">7.4</span> Self-Assessment</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content page-columns page-full column-body" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Report 1</h1>
<p class="subtitle lead">Determining the efficacy of the Standard Genetic Code against Point-Form Mutations</p>
</div>



<div class="quarto-title-meta column-body">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Arjun Mehandiratta </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 10, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="version-info">
<p>Last updated October 10, 2025.<br> <a href="https://github.com/SmatMan/CSB195/blob/main/docs/Arjun_Mehandiratta_Report_1.qmd">Source</a></p>
</div>
<section id="overview" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Overview</h1>
<ul>
<li>This project’s goal is to evaluate how optimal the Standard Genetic Code <strong>(SGC)</strong> is when compared to other random genetic codes.</li>
<li>The analysis displays that the SGC is much more well optimized against point mutations than random alternatives are.</li>
<li>Therefore, the SGC must have evolved under selection to prevent against such mutations.</li>
</ul>
</section>
<section id="specification" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Specification</h1>
<ul>
<li><p>My specification for the primary experiment is below, I will summarize it briefly underneath.</p>
<pre><code>The goal of this project is to compare the Standard Genetic Code (SGC) to other, 
randomly generated codes that are biologically plausible. 
We will do so by the following general specification.


Variables:
  - BASE_DIR = base directory of project
  - aaSimilarity = PCAs of amino acid similarities
  - dtable = lookup table of single letter amino acids showing distance from each other
  - SGC = standard genetic code of form: {'CODON': "AMINO ACID"} 
          where amino acids are in single letter form 
Functions:
  - ALREADY BUILT! construct_aa_table(dat: pd.DataFrame) -&gt; pd.DataFrame
    - Constructs a pandas dataframe from imported csv stored in aaSimilarity, 
      containing a 21x21 matrix of Euclidean distances between amino acids
      The function calculates the distance between 20 standard amino acids based on 
      12 principal components and adds a 'STOP' codon.
  - ALREADY BUILT! aaSim(aa1: str, aa2: str) -&gt; np.float64
    - Looks up the values in the dataframe constructed earlier in `dtable`
      and returns the applicable stored distance between aa1, aa2; amino acids
  - create_genetic_code(sc_limit: int) -&gt; dict
    - Creates a random genetic code config that maps all 64 codons to all 
      20 amino acids (in single letter form) + at least one stop codon 
      (allow 1-`sc_limit` stop codons).
      Return as a python dict e.g. {'ATA': 'A', 'ATT': 'W'}
  - score_gc(gc: dict) -&gt; np.float64
    - Provides a 'score' of the provided genetic code in `gc`. 
    It must calculate the cost of every possible codon mutation (3 codon mutations 
    per 3 codons), then use aaSim() between the gc-mapped amino acid of the original codon 
    and the gc-mapped amino acid of the mutated codon 
    to determine what the distance between each mutation is. 
    Sum up the distances of every possible mutation and return as np.float64 </code></pre></li>
<li><p>construct_aa_table() is a function that creates a <code>pandas</code> dataframe using the data from class. This used to be written in R but I ported it to python with the help of my AI. It creates a table showing similarity between amino acids.</p></li>
<li><p>aaSim() uses the dataframe from above as a lookup table to quickly get the similarity value between two amino acids passed into func</p></li>
<li><p>create_genetic_code() generates a random valid genetic code. Valid in this case means all 20 amino acids are mapped to at least one codon, and one stop codon is mapped.</p></li>
<li><p>score_gc() uses aaSim() to measure the point mutations of <em>every</em> codon to see how similar/different the post-mutation amino acid is to the pre-mutated amino acid</p>
<ul>
<li>The benchmark set in class is <code>9856.116</code> which is the score of the SGC.</li>
</ul></li>
</ul>
</section>
<section id="setting-up-the-environment" class="level1 page-columns page-full" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Setting Up the Environment</h1>
<section id="assumptions" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="assumptions"><span class="header-section-number">3.1</span> Assumptions</h2>
<ul>
<li>I define my computable measure of tolerance of point mutations as a score that, for any genetic code, determines all possible point shift mutations for each codon, calculates the distances between the original and mutated amino acid (using lookup table) and sums the distances to return the score.</li>
<li>I explain below how the function create_genetic_code() creates a valid genetic code. To summarize: 20 amino acids are assigned to 20 codons, a random amount of stop codons (as specified) are added, the rest are filled with random amino acids. This ensures the code is biologically useful as it maps all amino acids.
<ul>
<li>There are other constraints that exist in biology, like tRNA availability and codon usage bias in organisms, however we ignore these in this experiment for simplicity.</li>
</ul></li>
<li>This experiment will be conducted by building the functions and logic, using</li>
</ul>
</section>
<section id="some-constants" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="some-constants"><span class="header-section-number">3.2</span> Some constants</h2>
<ul>
<li><p>I import some packages and add some constants to begin. STOP_CODON_LIMIT determines the max of how many stop codons the generator will try to assign.</p>
<div id="89e6aaeb" class="cell" data-execution_count="2">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statistics <span class="im">as</span> stats</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> itertools <span class="im">import</span> product</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> Counter</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>BASE_DIR <span class="op">=</span> <span class="st">"~/Documents/School/CSB195"</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>STOP_CODON_LIMIT <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>SGC <span class="op">=</span> {</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">'ATA'</span>: <span class="st">'I'</span>, <span class="st">'ATC'</span>: <span class="st">'I'</span>, <span class="st">'ATT'</span>: <span class="st">'I'</span>, <span class="st">'ATG'</span>: <span class="st">'M'</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">'ACA'</span>: <span class="st">'T'</span>, <span class="st">'ACC'</span>: <span class="st">'T'</span>, <span class="st">'ACG'</span>: <span class="st">'T'</span>, <span class="st">'ACT'</span>: <span class="st">'T'</span>,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">'AAC'</span>: <span class="st">'N'</span>, <span class="st">'AAT'</span>: <span class="st">'N'</span>, <span class="st">'AAA'</span>: <span class="st">'K'</span>, <span class="st">'AAG'</span>: <span class="st">'K'</span>,</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">'AGC'</span>: <span class="st">'S'</span>, <span class="st">'AGT'</span>: <span class="st">'S'</span>, <span class="st">'AGA'</span>: <span class="st">'R'</span>, <span class="st">'AGG'</span>: <span class="st">'R'</span>,</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CTA'</span>: <span class="st">'L'</span>, <span class="st">'CTC'</span>: <span class="st">'L'</span>, <span class="st">'CTG'</span>: <span class="st">'L'</span>, <span class="st">'CTT'</span>: <span class="st">'L'</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CCA'</span>: <span class="st">'P'</span>, <span class="st">'CCC'</span>: <span class="st">'P'</span>, <span class="st">'CCG'</span>: <span class="st">'P'</span>, <span class="st">'CCT'</span>: <span class="st">'P'</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CAC'</span>: <span class="st">'H'</span>, <span class="st">'CAT'</span>: <span class="st">'H'</span>, <span class="st">'CAA'</span>: <span class="st">'Q'</span>, <span class="st">'CAG'</span>: <span class="st">'Q'</span>,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">'CGA'</span>: <span class="st">'R'</span>, <span class="st">'CGC'</span>: <span class="st">'R'</span>, <span class="st">'CGG'</span>: <span class="st">'R'</span>, <span class="st">'CGT'</span>: <span class="st">'R'</span>,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">'GTA'</span>: <span class="st">'V'</span>, <span class="st">'GTC'</span>: <span class="st">'V'</span>, <span class="st">'GTG'</span>: <span class="st">'V'</span>, <span class="st">'GTT'</span>: <span class="st">'V'</span>,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">'GCA'</span>: <span class="st">'A'</span>, <span class="st">'GCC'</span>: <span class="st">'A'</span>, <span class="st">'GCG'</span>: <span class="st">'A'</span>, <span class="st">'GCT'</span>: <span class="st">'A'</span>,</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">'GAC'</span>: <span class="st">'D'</span>, <span class="st">'GAT'</span>: <span class="st">'D'</span>, <span class="st">'GAA'</span>: <span class="st">'E'</span>, <span class="st">'GAG'</span>: <span class="st">'E'</span>,</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>    <span class="st">'GGA'</span>: <span class="st">'G'</span>, <span class="st">'GGC'</span>: <span class="st">'G'</span>, <span class="st">'GGG'</span>: <span class="st">'G'</span>, <span class="st">'GGT'</span>: <span class="st">'G'</span>,</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    <span class="st">'TCA'</span>: <span class="st">'S'</span>, <span class="st">'TCC'</span>: <span class="st">'S'</span>, <span class="st">'TCG'</span>: <span class="st">'S'</span>, <span class="st">'TCT'</span>: <span class="st">'S'</span>,</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">'TTC'</span>: <span class="st">'F'</span>, <span class="st">'TTT'</span>: <span class="st">'F'</span>, <span class="st">'TTA'</span>: <span class="st">'L'</span>, <span class="st">'TTG'</span>: <span class="st">'L'</span>,</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    <span class="st">'TAC'</span>: <span class="st">'Y'</span>, <span class="st">'TAT'</span>: <span class="st">'Y'</span>, <span class="st">'TAA'</span>: <span class="st">'*'</span>, <span class="st">'TAG'</span>: <span class="st">'*'</span>,</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="st">'TGC'</span>: <span class="st">'C'</span>, <span class="st">'TGT'</span>: <span class="st">'C'</span>, <span class="st">'TGA'</span>: <span class="st">'*'</span>, <span class="st">'TGG'</span>: <span class="st">'W'</span>,</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></li>
<li><p>Let’s also create a function (with the help of my AI) to nicely display any genetic codes we create</p>
<div id="ad9780f5" class="cell" data-execution_count="3">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> format_genetic_code_table(gc: <span class="bu">dict</span>) <span class="op">-&gt;</span> pd.DataFrame:</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">"""Formats a genetic code dictionary into a standard 4x4 table."""</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  bases <span class="op">=</span> [<span class="st">'T'</span>, <span class="st">'C'</span>, <span class="st">'A'</span>, <span class="st">'G'</span>]</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  aa_map <span class="op">=</span> {</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>      <span class="st">'A'</span>: <span class="st">'Ala'</span>, <span class="st">'R'</span>: <span class="st">'Arg'</span>, <span class="st">'N'</span>: <span class="st">'Asn'</span>, <span class="st">'D'</span>: <span class="st">'Asp'</span>, <span class="st">'C'</span>: <span class="st">'Cys'</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>      <span class="st">'E'</span>: <span class="st">'Glu'</span>, <span class="st">'Q'</span>: <span class="st">'Gln'</span>, <span class="st">'G'</span>: <span class="st">'Gly'</span>, <span class="st">'H'</span>: <span class="st">'His'</span>, <span class="st">'I'</span>: <span class="st">'Ile'</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>      <span class="st">'L'</span>: <span class="st">'Leu'</span>, <span class="st">'K'</span>: <span class="st">'Lys'</span>, <span class="st">'M'</span>: <span class="st">'Met'</span>, <span class="st">'F'</span>: <span class="st">'Phe'</span>, <span class="st">'P'</span>: <span class="st">'Pro'</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>      <span class="st">'S'</span>: <span class="st">'Ser'</span>, <span class="st">'T'</span>: <span class="st">'Thr'</span>, <span class="st">'W'</span>: <span class="st">'Trp'</span>, <span class="st">'Y'</span>: <span class="st">'Tyr'</span>, <span class="st">'V'</span>: <span class="st">'Val'</span>,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>      <span class="st">'*'</span>: <span class="st">'*'</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  table_data <span class="op">=</span> []</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> first_base <span class="kw">in</span> bases:</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>      row <span class="op">=</span> []</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> second_base <span class="kw">in</span> bases:</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>          cell_text <span class="op">=</span> []</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> third_base <span class="kw">in</span> bases:</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>              codon <span class="op">=</span> first_base <span class="op">+</span> second_base <span class="op">+</span> third_base</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>              single_letter_aa <span class="op">=</span> gc.get(codon, <span class="st">'?'</span>)</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>              three_letter_aa <span class="op">=</span> aa_map.get(single_letter_aa, <span class="st">'???'</span>)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>              cell_text.append(<span class="ss">f"</span><span class="sc">{</span>three_letter_aa<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span>third_base<span class="sc">}</span><span class="ss">)"</span>)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>          row.append(<span class="st">"&lt;br&gt;"</span>.join(cell_text))</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>      table_data.append(row)</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  df <span class="op">=</span> pd.DataFrame(table_data, </span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>                    index<span class="op">=</span>pd.Index(bases, name<span class="op">=</span><span class="st">"First Base"</span>), </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>                    columns<span class="op">=</span>pd.Index(bases, name<span class="op">=</span><span class="st">"Second Base"</span>))</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>  <span class="im">from</span> IPython.display <span class="im">import</span> HTML</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> HTML(df.to_html(escape<span class="op">=</span><span class="va">False</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></li>
<li><p>And a helper function to show the amino acid distribution for any inputted code.</p>
<div id="e546f01f" class="cell" data-execution_count="4">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_aa_distribution(gc: <span class="bu">dict</span>, title: <span class="bu">str</span> <span class="op">=</span> <span class="st">'Amino Acid Distribution in Genetic Code'</span>):</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">"""</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">  Renders a bar chart showing the frequency of each amino acid in a given genetic code.</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">  The chart layout is kept consistent for easy comparison between different codes.</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">  Args:</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">      gc: A dictionary mapping 64 codons to single-letter amino acid codes (e.g., 'A', 'C', '*').</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co">      title: The title for the chart.</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co">  """</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define a fixed order for amino acids to ensure consistent chart layouts</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  aa_order <span class="op">=</span> <span class="bu">list</span>(<span class="st">'ACDEFGHIKLMNPQRSTVWY'</span>) <span class="op">+</span> [<span class="st">'*'</span>]</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Count the occurrences of each amino acid in the genetic code</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  counts <span class="op">=</span> Counter(gc.values())</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create a pandas Series from the counts, reindexing to the fixed order</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This ensures that all amino acids are present in the data, with a count of 0 if absent.</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>  dist_series <span class="op">=</span> pd.Series(counts).reindex(aa_order, fill_value<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plotting</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>  fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">14</span>, <span class="dv">7</span>))</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  dist_series.plot(kind<span class="op">=</span><span class="st">'bar'</span>, ax<span class="op">=</span>ax, color<span class="op">=</span><span class="st">'skyblue'</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>, width<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Chart Formatting for Clarity and Consistency ---</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  ax.set_title(title, fontsize<span class="op">=</span><span class="dv">16</span>, weight<span class="op">=</span><span class="st">'bold'</span>)</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  ax.set_xlabel(<span class="st">'Amino Acid'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>  ax.set_ylabel(<span class="st">'Number of Codons (Frequency)'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set a fixed y-axis limit for direct visual comparison of codon counts.</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># A safe upper limit is 8, as it's rare for one amino acid to have more codons.</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>  ax.set_ylim(<span class="dv">0</span>, <span class="dv">8</span>)</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>  ax.set_yticks(<span class="bu">range</span>(<span class="dv">9</span>))</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Rotate x-axis labels for better readability</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>  plt.xticks(rotation<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add a grid for easier reading of values</span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>  ax.grid(axis<span class="op">=</span><span class="st">'y'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add count labels on top of each bar</span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> i, count <span class="kw">in</span> <span class="bu">enumerate</span>(dist_series):</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> count <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>          ax.text(i, count <span class="op">+</span> <span class="fl">0.1</span>, <span class="bu">str</span>(count), ha<span class="op">=</span><span class="st">'center'</span>, va<span class="op">=</span><span class="st">'bottom'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure a clean layout</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>  plt.tight_layout()</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Display the plot</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>  plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div></li>
</ul>
</section>
<section id="importing-the-data" class="level2 page-columns page-full" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="importing-the-data"><span class="header-section-number">3.3</span> Importing the data</h2>
<ul class="page-columns page-full">
<li><p>Since the original data was in .Rds format, I used R’s write.csv() command to export <code>dat/aaFeatureSpace.2025.Rds</code> to <code>dat/aaFeatureSpace.csv</code></p></li>
<li><p>Python accesses this data and constructs a similarity lookup table using the following code: <sup><a href="https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#0">AI Source</a></sup></p>
<div id="47ae6004" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Show the code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> construct_aa_table(dat: pd.DataFrame) <span class="op">-&gt;</span> pd.DataFrame:</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">"""</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">  Constructs a 21x21 matrix of Euclidean distances between amino acids.</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">  The function calculates the distance between 20 standard amino acids based on </span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">  12 principal components and adds a pseudo 'STOP' codon.</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">  Args:</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co">      dat: A pandas DataFrame with amino acid single-letter codes as the index </span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co">           and 12 principal components (PC1-PC12) as columns.</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns:</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co">      A 21x21 pandas DataFrame containing the Euclidean distances between all </span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">      20 amino acids and a 'STOP' codon.</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co">  """</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- FIX ---</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure that the dataframe values are numeric before calculations</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This prevents the TypeError during vector subtraction.</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>  dat_numeric <span class="op">=</span> dat.<span class="bu">apply</span>(pd.to_numeric, errors<span class="op">=</span><span class="st">'coerce'</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># It's good practice to check if coercion created any NaN values</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> dat_numeric.isnull().values.<span class="bu">any</span>():</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>      <span class="bu">print</span>(<span class="st">"Warning: Non-numeric values were found and converted to NaN."</span>)</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define the constant for calculating the distance to the stop codon</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  STOP_DIST <span class="op">=</span> <span class="fl">1.5</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract the list of amino acids from the input DataFrame's index</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>  amino_acids <span class="op">=</span> dat_numeric.index.tolist()</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize an empty square DataFrame to hold the distances</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  dist_matrix <span class="op">=</span> pd.DataFrame(index<span class="op">=</span>amino_acids, columns<span class="op">=</span>amino_acids, dtype<span class="op">=</span><span class="bu">float</span>)</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate the Euclidean distance for each pair of amino acids</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> aa1 <span class="kw">in</span> amino_acids:</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> aa2 <span class="kw">in</span> amino_acids:</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Retrieve the 12-dimensional vectors for the two amino acids</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>          vec1 <span class="op">=</span> dat_numeric.loc[aa1].values</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>          vec2 <span class="op">=</span> dat_numeric.loc[aa2].values</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Compute the Euclidean distance and store it in the matrix</span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>          distance <span class="op">=</span> np.linalg.norm(vec1 <span class="op">-</span> vec2)</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>          dist_matrix.loc[aa1, aa2] <span class="op">=</span> distance</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find the maximum distance value among the amino acids</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>  max_dist <span class="op">=</span> dist_matrix.values.<span class="bu">max</span>()</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate the distance for the stop codon</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>  stop_codon_dist <span class="op">=</span> STOP_DIST <span class="op">*</span> max_dist</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add a 'STOP' column and row to the matrix</span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>  stop_label <span class="op">=</span> <span class="st">'*'</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>  dist_matrix[stop_label] <span class="op">=</span> stop_codon_dist</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>  dist_matrix.loc[stop_label] <span class="op">=</span> stop_codon_dist</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The distance from 'STOP' to itself is 0</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>  dist_matrix.loc[stop_label, stop_label] <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> dist_matrix</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>aaSimilarity <span class="op">=</span> pd.read_csv(<span class="ss">f"</span><span class="sc">{</span>BASE_DIR<span class="sc">}</span><span class="ss">/dat/aaFeatureSpace.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>dtable <span class="op">=</span> construct_aa_table(aaSimilarity)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div></li>
<li class="page-columns page-full"><p>Now the lookup table is constructed nicely and stored in <code>dtable</code>:</p>
<div id="4e348181" class="cell page-columns page-full" data-execution_count="6">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>dtable</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display column-page" data-execution_count="29">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">D</th>
<th data-quarto-table-cell-role="th">E</th>
<th data-quarto-table-cell-role="th">F</th>
<th data-quarto-table-cell-role="th">G</th>
<th data-quarto-table-cell-role="th">H</th>
<th data-quarto-table-cell-role="th">I</th>
<th data-quarto-table-cell-role="th">K</th>
<th data-quarto-table-cell-role="th">L</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">N</th>
<th data-quarto-table-cell-role="th">P</th>
<th data-quarto-table-cell-role="th">Q</th>
<th data-quarto-table-cell-role="th">R</th>
<th data-quarto-table-cell-role="th">S</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">V</th>
<th data-quarto-table-cell-role="th">W</th>
<th data-quarto-table-cell-role="th">Y</th>
<th data-quarto-table-cell-role="th">*</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>0.000000</td>
<td>18.176334</td>
<td>21.211890</td>
<td>21.923157</td>
<td>22.371972</td>
<td>11.482274</td>
<td>18.900629</td>
<td>19.645559</td>
<td>27.036963</td>
<td>17.392675</td>
<td>...</td>
<td>18.462433</td>
<td>19.083326</td>
<td>19.461327</td>
<td>29.894757</td>
<td>10.715891</td>
<td>10.600212</td>
<td>14.435550</td>
<td>28.323295</td>
<td>23.341264</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>18.176334</td>
<td>0.000000</td>
<td>28.319709</td>
<td>29.728459</td>
<td>20.041857</td>
<td>25.054519</td>
<td>21.739702</td>
<td>17.581695</td>
<td>33.867608</td>
<td>17.658081</td>
<td>...</td>
<td>25.926758</td>
<td>27.312201</td>
<td>27.016538</td>
<td>33.663847</td>
<td>22.000385</td>
<td>20.114841</td>
<td>13.842399</td>
<td>25.908648</td>
<td>22.289136</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">D</th>
<td>21.211890</td>
<td>28.319709</td>
<td>0.000000</td>
<td>7.859468</td>
<td>32.456299</td>
<td>23.219087</td>
<td>18.435790</td>
<td>32.599804</td>
<td>18.891600</td>
<td>30.707916</td>
<td>...</td>
<td>11.938848</td>
<td>20.851585</td>
<td>11.273871</td>
<td>21.535519</td>
<td>13.592779</td>
<td>15.809162</td>
<td>28.758401</td>
<td>33.169889</td>
<td>25.982454</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">E</th>
<td>21.923157</td>
<td>29.728459</td>
<td>7.859468</td>
<td>0.000000</td>
<td>31.412447</td>
<td>25.289262</td>
<td>17.162447</td>
<td>31.094875</td>
<td>15.017994</td>
<td>28.952665</td>
<td>...</td>
<td>12.332626</td>
<td>22.363162</td>
<td>8.935904</td>
<td>18.980827</td>
<td>16.066922</td>
<td>17.324399</td>
<td>27.976748</td>
<td>31.759465</td>
<td>24.054114</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">F</th>
<td>22.371972</td>
<td>20.041857</td>
<td>32.456299</td>
<td>31.412447</td>
<td>0.000000</td>
<td>31.582276</td>
<td>21.746057</td>
<td>11.326064</td>
<td>34.060936</td>
<td>9.293158</td>
<td>...</td>
<td>28.611496</td>
<td>28.291862</td>
<td>27.837117</td>
<td>32.524010</td>
<td>27.225385</td>
<td>23.223101</td>
<td>14.060650</td>
<td>12.321010</td>
<td>14.091799</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>11.482274</td>
<td>25.054519</td>
<td>23.219087</td>
<td>25.289262</td>
<td>31.582276</td>
<td>0.000000</td>
<td>25.097377</td>
<td>28.775849</td>
<td>30.197159</td>
<td>27.214751</td>
<td>...</td>
<td>20.767591</td>
<td>20.145725</td>
<td>23.307118</td>
<td>33.426858</td>
<td>13.828360</td>
<td>17.115369</td>
<td>23.954105</td>
<td>35.957174</td>
<td>30.544664</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">H</th>
<td>18.900629</td>
<td>21.739702</td>
<td>18.435790</td>
<td>17.162447</td>
<td>21.746057</td>
<td>25.097377</td>
<td>0.000000</td>
<td>23.465681</td>
<td>20.076748</td>
<td>21.236117</td>
<td>...</td>
<td>14.913534</td>
<td>20.765900</td>
<td>14.389646</td>
<td>19.461579</td>
<td>16.819221</td>
<td>15.078192</td>
<td>20.613181</td>
<td>22.622245</td>
<td>15.307727</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">I</th>
<td>19.645559</td>
<td>17.581695</td>
<td>32.599804</td>
<td>31.094875</td>
<td>11.326064</td>
<td>28.775849</td>
<td>23.465681</td>
<td>0.000000</td>
<td>34.227760</td>
<td>6.733640</td>
<td>...</td>
<td>28.616005</td>
<td>27.744302</td>
<td>28.779372</td>
<td>34.776820</td>
<td>26.230282</td>
<td>22.495364</td>
<td>7.080977</td>
<td>19.449383</td>
<td>19.319285</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">K</th>
<td>27.036963</td>
<td>33.867608</td>
<td>18.891600</td>
<td>15.017994</td>
<td>34.060936</td>
<td>30.197159</td>
<td>20.076748</td>
<td>34.227760</td>
<td>0.000000</td>
<td>32.194150</td>
<td>...</td>
<td>17.420960</td>
<td>23.411609</td>
<td>12.381269</td>
<td>16.471912</td>
<td>20.806245</td>
<td>20.609675</td>
<td>31.602886</td>
<td>32.079634</td>
<td>25.243565</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">L</th>
<td>17.392675</td>
<td>17.658081</td>
<td>30.707916</td>
<td>28.952665</td>
<td>9.293158</td>
<td>27.214751</td>
<td>21.236117</td>
<td>6.733640</td>
<td>32.194150</td>
<td>0.000000</td>
<td>...</td>
<td>27.084261</td>
<td>25.801083</td>
<td>26.292854</td>
<td>32.900377</td>
<td>24.287408</td>
<td>20.874809</td>
<td>8.152492</td>
<td>18.577124</td>
<td>16.909402</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">M</th>
<td>17.359260</td>
<td>15.066169</td>
<td>27.880345</td>
<td>26.681749</td>
<td>11.131475</td>
<td>27.144879</td>
<td>17.351416</td>
<td>13.573533</td>
<td>29.369891</td>
<td>8.799197</td>
<td>...</td>
<td>24.597132</td>
<td>24.960747</td>
<td>23.340748</td>
<td>29.676611</td>
<td>22.260555</td>
<td>19.448368</td>
<td>12.687781</td>
<td>17.599251</td>
<td>15.416300</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">N</th>
<td>18.462433</td>
<td>25.926758</td>
<td>11.938848</td>
<td>12.332626</td>
<td>28.611496</td>
<td>20.767591</td>
<td>14.913534</td>
<td>28.616005</td>
<td>17.420960</td>
<td>27.084261</td>
<td>...</td>
<td>0.000000</td>
<td>19.324180</td>
<td>7.580542</td>
<td>19.788073</td>
<td>10.776410</td>
<td>13.261159</td>
<td>25.619874</td>
<td>29.981411</td>
<td>22.550224</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">P</th>
<td>19.083326</td>
<td>27.312201</td>
<td>20.851585</td>
<td>22.363162</td>
<td>28.291862</td>
<td>20.145725</td>
<td>20.765900</td>
<td>27.744302</td>
<td>23.411609</td>
<td>25.801083</td>
<td>...</td>
<td>19.324180</td>
<td>0.000000</td>
<td>20.001033</td>
<td>28.278856</td>
<td>16.696112</td>
<td>17.105724</td>
<td>24.659890</td>
<td>30.363342</td>
<td>24.694739</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Q</th>
<td>19.461327</td>
<td>27.016538</td>
<td>11.273871</td>
<td>8.935904</td>
<td>27.837117</td>
<td>23.307118</td>
<td>14.389646</td>
<td>28.779372</td>
<td>12.381269</td>
<td>26.292854</td>
<td>...</td>
<td>7.580542</td>
<td>20.001033</td>
<td>0.000000</td>
<td>16.300134</td>
<td>12.449408</td>
<td>13.382975</td>
<td>25.742964</td>
<td>28.484868</td>
<td>20.051376</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">R</th>
<td>29.894757</td>
<td>33.663847</td>
<td>21.535519</td>
<td>18.980827</td>
<td>32.524010</td>
<td>33.426858</td>
<td>19.461579</td>
<td>34.776820</td>
<td>16.471912</td>
<td>32.900377</td>
<td>...</td>
<td>19.788073</td>
<td>28.278856</td>
<td>16.300134</td>
<td>0.000000</td>
<td>24.334498</td>
<td>23.602768</td>
<td>32.849640</td>
<td>31.215589</td>
<td>24.343017</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">S</th>
<td>10.715891</td>
<td>22.000385</td>
<td>13.592779</td>
<td>16.066922</td>
<td>27.225385</td>
<td>13.828360</td>
<td>16.819221</td>
<td>26.230282</td>
<td>20.806245</td>
<td>24.287408</td>
<td>...</td>
<td>10.776410</td>
<td>16.696112</td>
<td>12.449408</td>
<td>24.334498</td>
<td>0.000000</td>
<td>6.167702</td>
<td>21.321638</td>
<td>30.880924</td>
<td>23.941856</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">T</th>
<td>10.600212</td>
<td>20.114841</td>
<td>15.809162</td>
<td>17.324399</td>
<td>23.223101</td>
<td>17.115369</td>
<td>15.078192</td>
<td>22.495364</td>
<td>20.609675</td>
<td>20.874809</td>
<td>...</td>
<td>13.261159</td>
<td>17.105724</td>
<td>13.382975</td>
<td>23.602768</td>
<td>6.167702</td>
<td>0.000000</td>
<td>17.554950</td>
<td>27.092283</td>
<td>20.496910</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">V</th>
<td>14.435550</td>
<td>13.842399</td>
<td>28.758401</td>
<td>27.976748</td>
<td>14.060650</td>
<td>23.954105</td>
<td>20.613181</td>
<td>7.080977</td>
<td>31.602886</td>
<td>8.152492</td>
<td>...</td>
<td>25.619874</td>
<td>24.659890</td>
<td>25.742964</td>
<td>32.849640</td>
<td>21.321638</td>
<td>17.554950</td>
<td>0.000000</td>
<td>22.190822</td>
<td>19.065321</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">W</th>
<td>28.323295</td>
<td>25.908648</td>
<td>33.169889</td>
<td>31.759465</td>
<td>12.321010</td>
<td>35.957174</td>
<td>22.622245</td>
<td>19.449383</td>
<td>32.079634</td>
<td>18.577124</td>
<td>...</td>
<td>29.981411</td>
<td>30.363342</td>
<td>28.484868</td>
<td>31.215589</td>
<td>30.880924</td>
<td>27.092283</td>
<td>22.190822</td>
<td>0.000000</td>
<td>14.294620</td>
<td>53.935761</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Y</th>
<td>23.341264</td>
<td>22.289136</td>
<td>25.982454</td>
<td>24.054114</td>
<td>14.091799</td>
<td>30.544664</td>
<td>15.307727</td>
<td>19.319285</td>
<td>25.243565</td>
<td>16.909402</td>
<td>...</td>
<td>22.550224</td>
<td>24.694739</td>
<td>20.051376</td>
<td>24.343017</td>
<td>23.941856</td>
<td>20.496910</td>
<td>19.065321</td>
<td>14.294620</td>
<td>0.000000</td>
<td>53.935761</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">*</th>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>...</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>53.935761</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

<p>21 rows × 21 columns</p>
</div>
</div>
</div></li>
</ul>
</section>
<section id="generator-and-scorer" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="generator-and-scorer"><span class="header-section-number">3.4</span> Generator and Scorer</h2>
<ul>
<li><p>Now we can define the generator function with the following: <sup><a href="https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#4">AI source</a></sup></p>
<div id="18dfde58" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Show the code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_genetic_code(sc_limit: <span class="bu">int</span>) <span class="op">-&gt;</span> <span class="bu">dict</span>:</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">"""</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">  Creates a random genetic code config that maps all 64 codons to all </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">  20 amino acids and at least one stop codon.</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">  Args:</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">      sc_limit: The maximum number of stop codons allowed (the actual</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">                number will be chosen randomly between 1 and sc_limit).</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns:</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co">      A dictionary mapping each of the 64 codons to a single-letter </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co">      amino acid code or a stop codon ('*').</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="co">  Raises:</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="co">      ValueError: If sc_limit is not within a reasonable range (1 to 10).</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co">  """</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Validate the stop codon limit to a sensible range</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="kw">not</span> <span class="dv">1</span> <span class="op">&lt;=</span> sc_limit <span class="op">&lt;=</span> <span class="dv">10</span>:</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>      <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="st">"sc_limit must be between 1 and 10 for practical genetic codes."</span>)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Define the building blocks for the genetic code</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  bases <span class="op">=</span> [<span class="st">'A'</span>, <span class="st">'T'</span>, <span class="st">'G'</span>, <span class="st">'C'</span>]</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  amino_acids <span class="op">=</span> <span class="bu">list</span>(<span class="st">"ACDEFGHIKLMNPQRSTVWY"</span>)  <span class="co"># 20 standard amino acids</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>  stop_symbol <span class="op">=</span> <span class="st">'*'</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 1. Generate all 64 possible codons</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>  all_codons <span class="op">=</span> [<span class="st">''</span>.join(p) <span class="cf">for</span> p <span class="kw">in</span> product(bases, repeat<span class="op">=</span><span class="dv">3</span>)]</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>  random.shuffle(all_codons)</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 2. Determine the number of stop codons</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>  num_stop_codons <span class="op">=</span> random.randint(<span class="dv">1</span>, sc_limit)</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 3. Create the pool of assignments (amino acids + stops)</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure all 20 amino acids are present at least once</span></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>  assignments <span class="op">=</span> amino_acids.copy()</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add the stop codons</span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>  assignments.extend([stop_symbol] <span class="op">*</span> num_stop_codons)</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Fill the remaining slots with random amino acids to reach 64 total assignments</span></span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>  remaining_slots <span class="op">=</span> <span class="dv">64</span> <span class="op">-</span> <span class="bu">len</span>(assignments)</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>  assignments.extend(random.choices(amino_acids, k<span class="op">=</span>remaining_slots))</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Shuffle the assignments to ensure randomness</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>  random.shuffle(assignments)</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 4. Create the dictionary by mapping codons to assignments</span></span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>  genetic_code <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(all_codons, assignments))</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> genetic_code</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div></li>
<li><p>And we can see that running this function generates us a nice randomly assigned genetic code:</p>
<div id="dcd38e91" class="cell" data-execution_count="8">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>test_code <span class="op">=</span> create_genetic_code(<span class="dv">2</span>) <span class="co"># with max 2 stop codons</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>format_genetic_code_table(test_code) <span class="co"># display nicely</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Second Base</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">G</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">First Base</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">T</th>
<td>Pro (T)<br>
Met (C)<br>
Asn (A)<br>
Tyr (G)</td>
<td>Gly (T)<br>
Thr (C)<br>
Ile (A)<br>
Trp (G)</td>
<td>Val (T)<br>
Trp (C)<br>
Ala (A)<br>
Gln (G)</td>
<td>Glu (T)<br>
Arg (C)<br>
Trp (A)<br>
Gly (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>* (T)<br>
Asn (C)<br>
Gly (A)<br>
Tyr (G)</td>
<td>His (T)<br>
Ile (C)<br>
Pro (A)<br>
Gly (G)</td>
<td>Glu (T)<br>
Ile (C)<br>
Tyr (A)<br>
His (G)</td>
<td>Ile (T)<br>
Leu (C)<br>
Pro (A)<br>
Arg (G)</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>Phe (T)<br>
Val (C)<br>
Thr (A)<br>
Ser (G)</td>
<td>Ile (T)<br>
Cys (C)<br>
Ser (A)<br>
Trp (G)</td>
<td>Ala (T)<br>
Leu (C)<br>
Lys (A)<br>
Phe (G)</td>
<td>Gly (T)<br>
Cys (C)<br>
Tyr (A)<br>
Arg (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>Trp (T)<br>
Phe (C)<br>
Gln (A)<br>
Ala (G)</td>
<td>Met (T)<br>
Val (C)<br>
Gln (A)<br>
Asp (G)</td>
<td>Glu (T)<br>
Gln (C)<br>
Ala (A)<br>
Glu (G)</td>
<td>Lys (T)<br>
Asn (C)<br>
Ser (A)<br>
Gln (G)</td>
</tr>
</tbody>
</table>
</div>
</div></li>
<li><p>Next is the scorer function, defined as follows: <sup><a href="https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#6">AI source</a></sup></p>
<div id="8a5bf9aa" class="cell" data-execution_count="9">
<details class="code-fold">
<summary>Show the code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> score_gc(gc: <span class="bu">dict</span>) <span class="op">-&gt;</span> np.float64:</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">"""</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">  Calculates a 'robustness' score for a genetic code.</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">  The score is the sum of physicochemical distances between amino acids </span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">  resulting from all possible single-nucleotide point mutations across all </span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">  64 codons. A lower score indicates a more robust code where mutations </span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">  tend to result in similar amino acids.</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">  Args:</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">      gc: A dictionary mapping 64 codons to single-letter amino acid codes.</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">      dtable: A DataFrame containing the pairwise distances between amino acids,</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">              as created by the `construct_aa_table` function.</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns:</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="co">      A np.float64 value representing the total cost of all possible mutations.</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="co">  Raises:</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="co">      ValueError: If a codon in the genetic code is not found in the distance table.</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a><span class="co">  """</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  bases <span class="op">=</span> [<span class="st">'A'</span>, <span class="st">'T'</span>, <span class="st">'G'</span>, <span class="st">'C'</span>]</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  total_distance_cost: <span class="bu">float</span> <span class="op">=</span> <span class="fl">0.0</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Generate all 64 codons</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>  all_codons <span class="op">=</span> [<span class="st">''</span>.join(p) <span class="cf">for</span> p <span class="kw">in</span> product(bases, repeat<span class="op">=</span><span class="dv">3</span>)]</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> original_codon <span class="kw">in</span> all_codons:</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Get the amino acid for the original, unmutated codon</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>      original_aa <span class="op">=</span> gc.get(original_codon)</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> original_aa <span class="kw">is</span> <span class="va">None</span>:</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>          <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="ss">f"Codon '</span><span class="sc">{</span>original_codon<span class="sc">}</span><span class="ss">' not found in the provided genetic code."</span>)</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Iterate through each of the 3 positions in the codon</span></span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">3</span>):</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>          original_base <span class="op">=</span> original_codon[i]</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Iterate through each possible mutation at that position</span></span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> mutated_base <span class="kw">in</span> bases:</span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>              <span class="co"># We only care about actual mutations, not silent changes</span></span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>              <span class="cf">if</span> original_base <span class="op">!=</span> mutated_base:</span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>                  <span class="co"># Construct the mutated codon</span></span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>                  mutated_codon_list <span class="op">=</span> <span class="bu">list</span>(original_codon)</span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>                  mutated_codon_list[i] <span class="op">=</span> mutated_base</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>                  mutated_codon <span class="op">=</span> <span class="st">""</span>.join(mutated_codon_list)</span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>                  <span class="co"># Get the amino acid for the new, mutated codon</span></span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>                  mutated_aa <span class="op">=</span> gc.get(mutated_codon)</span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>                  <span class="cf">if</span> mutated_aa <span class="kw">is</span> <span class="va">None</span>:</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a>                      <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="ss">f"Mutated codon '</span><span class="sc">{</span>mutated_codon<span class="sc">}</span><span class="ss">' not found in the provided genetic code."</span>)</span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>                  <span class="co"># Use aaSim to get the distance and add it to the total cost</span></span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a>                  <span class="co"># Note: We are directly using the dtable here as aaSim is just a lookup</span></span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a>                  distance: np.float64 <span class="op">=</span> dtable.at[original_aa, mutated_aa]</span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>                  total_distance_cost <span class="op">+=</span> distance</span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> np.float64(total_distance_cost)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div></li>
<li><p>To ensure this function works, we will test the Standard Genetic Code against the baseline in the next section</p></li>
</ul>
</section>
<section id="ensuring-baseline" class="level2 page-columns page-full" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="ensuring-baseline"><span class="header-section-number">3.5</span> Ensuring Baseline</h2>
<ul class="page-columns page-full">
<li><p>The baseline score for the SGC was defined in class as <code>9856.116</code>. In order to ensure nothing has gone wrong, we will test our functions to see if it returns the same score.</p></li>
<li class="page-columns page-full"><p>Since we already have the SGC defined in <code>SGC</code> from the constants earlier, we will use that here.</p>
<div id="c41d9a9d" class="cell page-columns page-full" data-execution_count="10">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>format_genetic_code_table(SGC)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display column-page" data-execution_count="33">
<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Second Base</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">G</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">First Base</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">T</th>
<td>Phe (T)<br>
Phe (C)<br>
Leu (A)<br>
Leu (G)</td>
<td>Ser (T)<br>
Ser (C)<br>
Ser (A)<br>
Ser (G)</td>
<td>Tyr (T)<br>
Tyr (C)<br>
* (A)<br>
* (G)</td>
<td>Cys (T)<br>
Cys (C)<br>
* (A)<br>
Trp (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>Leu (T)<br>
Leu (C)<br>
Leu (A)<br>
Leu (G)</td>
<td>Pro (T)<br>
Pro (C)<br>
Pro (A)<br>
Pro (G)</td>
<td>His (T)<br>
His (C)<br>
Gln (A)<br>
Gln (G)</td>
<td>Arg (T)<br>
Arg (C)<br>
Arg (A)<br>
Arg (G)</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>Ile (T)<br>
Ile (C)<br>
Ile (A)<br>
Met (G)</td>
<td>Thr (T)<br>
Thr (C)<br>
Thr (A)<br>
Thr (G)</td>
<td>Asn (T)<br>
Asn (C)<br>
Lys (A)<br>
Lys (G)</td>
<td>Ser (T)<br>
Ser (C)<br>
Arg (A)<br>
Arg (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>Val (T)<br>
Val (C)<br>
Val (A)<br>
Val (G)</td>
<td>Ala (T)<br>
Ala (C)<br>
Ala (A)<br>
Ala (G)</td>
<td>Asp (T)<br>
Asp (C)<br>
Glu (A)<br>
Glu (G)</td>
<td>Gly (T)<br>
Gly (C)<br>
Gly (A)<br>
Gly (G)</td>
</tr>
</tbody>
</table>
</div>
</div></li>
<li><p>Scoring it provides us with the following:</p>
<div id="705486e5" class="cell" data-execution_count="11">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>score_gc(SGC)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display" data-execution_count="34">
<pre><code>np.float64(9856.116291937447)</code></pre>
</div>
</div></li>
<li><p>Hooray! We can see the code has scored the Standard Genetic Code equal to the baseline!! (to an even more precise figure!).</p></li>
</ul>
</section>
</section>
<section id="data" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Data</h1>
<section id="comparing-random-codes-to-the-sgc" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="comparing-random-codes-to-the-sgc"><span class="header-section-number">4.1</span> Comparing Random Codes to the SGC</h2>
<ul>
<li><p>Let’s create a helper function to generate and test random GCs called <code>gen_and_score(times)</code>; where <code>times</code> is the amount of trials to run.</p>
<div id="2d7ea837" class="cell" data-execution_count="12">
<details class="code-fold">
<summary>Show the code</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> gen_and_score(times: <span class="bu">int</span>) <span class="op">-&gt;</span> <span class="bu">list</span>:</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    scores <span class="op">=</span> []</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> <span class="bu">len</span>(scores) <span class="op">&lt;</span> times:</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>        gc <span class="op">=</span> create_genetic_code(STOP_CODON_LIMIT)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>        scores.append(<span class="bu">float</span>(score_gc(gc)))</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> scores</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</div></li>
<li><p>Now, let’s generate 1000 random genetic codes, score them, and plot them on a graph.</p>
<div id="932a671d" class="cell" data-execution_count="13">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>attempts <span class="op">=</span> <span class="dv">1000</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>scores <span class="op">=</span> gen_and_score(attempts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="f4286d40" class="cell" data-execution_count="14">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>plt.hist(scores, bins<span class="op">=</span><span class="dv">40</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f'Distribution of Genetic Code Scores (n=</span><span class="sc">{</span>attempts<span class="sc">}</span><span class="ss">)'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Score'</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-14-output-1.png" width="585" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div></li>
<li><p>This is nice! We get a nice bell curve, and we can see the scores average between 12,000 and 15,000. These scores are MUCH higher than the SGC’s score, however. Let’s plot the SCG score as well to visualize this.</p>
<div id="e0755b1d" class="cell" data-execution_count="15">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>scores.append(score_gc(SGC))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>plt.hist(scores, bins<span class="op">=</span><span class="dv">50</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f'Distribution of Genetic Code Scores w/ SGC (n=</span><span class="sc">{</span>attempts<span class="sc">}</span><span class="ss">)'</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Score'</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-15-output-1.png" width="585" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div></li>
<li><p>Our random codes score <em>significantly</em> worse than the standard genetic code. Let’s try generating more codes (5000) to try to brute-force optimize…</p>
<div id="7a84e536" class="cell" data-execution_count="16">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>attempts <span class="op">=</span> <span class="dv">5000</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>scores <span class="op">=</span> gen_and_score(attempts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="e3c368bc" class="cell" data-execution_count="17">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>plt.hist(scores, bins<span class="op">=</span><span class="dv">40</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f'Distribution of Genetic Code Scores (n=</span><span class="sc">{</span>attempts<span class="sc">}</span><span class="ss">)'</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Score'</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-17-output-1.png" width="593" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div></li>
<li><p>We get a wider distribution with more attempts, however when compared to the SGC:</p>
<div id="1d7d7c43" class="cell" data-execution_count="18">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>scores.append(score_gc(SGC))</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>plt.hist(scores, bins<span class="op">=</span><span class="dv">60</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f'Distribution of Genetic Code Scores w/ SGC (n=</span><span class="sc">{</span>attempts<span class="sc">}</span><span class="ss">)'</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Score'</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-18-output-1.png" width="614" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div></li>
<li><p>We can clearly see that the Standard Genetic Code is extremely well optimized when compared to random genetic codes.</p></li>
</ul>
</section>
</section>
<section id="optimization-extras" class="level1 page-columns page-full" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Optimization <em>(extras)</em></h1>
<section id="random-optimization" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="random-optimization"><span class="header-section-number">5.1</span> Random “Optimization”</h2>
<ul>
<li><p>I tried to make an “optimization” function. Truly all it is is a while True loop that randomly generates codes, keeps score of the best one its found, and keeps going until the user keyboard interrupts. So it’s not really optimization, rather brute force.</p></li>
<li><p>Regardless, let’s see how low of a score it can find. Since this quarto document doesn’t have a way to signal interrupt, I will limit the function to 10000 runs.</p>
<div id="542ca25b" class="cell" data-execution_count="19">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> random_optimization(attempts) <span class="op">-&gt;</span> np.float64:</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  best: <span class="bu">dict</span> <span class="op">=</span> create_genetic_code(STOP_CODON_LIMIT)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  best_score <span class="op">=</span> score_gc(best)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  tries <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">while</span> tries <span class="op">&lt;=</span> attempts:</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>        tries <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>        test <span class="op">=</span> create_genetic_code(STOP_CODON_LIMIT)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>        test_score <span class="op">=</span> score_gc(test)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> test_score <span class="op">&lt;</span> best_score:</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>            best <span class="op">=</span> test.copy()</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>            best_score <span class="op">=</span> test_score</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> <span class="pp">KeyboardInterrupt</span>:</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>tries<span class="sc">}</span><span class="ss"> total tries"</span>)</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>        sys.exit()</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"Best code: </span><span class="sc">{</span>best<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"Best score: </span><span class="sc">{</span>best_score<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>random_optimization(<span class="dv">10000</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Best code: {'TTT': 'T', 'GGT': 'V', 'AAT': 'I', 'GCT': 'H', 'AAA': 'I', 'ACT': 'P', 'TGC': 'I', 'AGC': 'E', 'GTG': 'Q', 'TAG': 'G', 'TGG': 'K', 'GGA': 'F', 'ACA': 'T', 'TTA': 'K', 'CGA': 'R', 'ATC': 'M', 'CCC': 'P', 'CCA': 'I', 'AAG': 'L', 'TGA': 'P', 'CCT': 'E', 'ACG': 'P', 'TCG': 'N', 'GCC': 'W', 'ATA': 'A', 'CAA': 'F', 'CGT': 'K', 'CGC': 'E', 'GTA': 'Q', 'TGT': 'E', 'CAC': 'H', 'AGA': 'K', 'ATT': 'N', 'CCG': 'G', 'CGG': '*', 'GAG': 'N', 'GCA': 'T', 'AGG': 'C', 'GAA': 'A', 'TAC': 'M', 'AAC': 'V', 'GAC': 'Y', 'GGC': 'M', 'GTT': 'K', 'TTG': 'N', 'TCC': 'S', 'CTG': 'T', 'CAG': 'S', 'GGG': 'H', 'CTT': 'Q', 'GTC': 'M', 'TAT': 'D', 'TCA': 'S', 'TTC': 'T', 'TAA': 'C', 'GAT': 'D', 'ATG': 'D', 'CAT': 'E', 'CTC': 'W', 'ACC': 'S', 'GCG': 'D', 'CTA': 'A', 'TCT': 'S', 'AGT': 'N'}
Best score: 11420.356037821826</code></pre>
</div>
</div></li>
<li><p>I have no way of knowing the exact amount when the quarto doc renders, but in my testing I rarely got better than 11000. Let’s try 100,000 attempts…</p>
<div id="a901a049" class="cell" data-execution_count="20">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>random_optimization(<span class="dv">100000</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Best code: {'CGG': 'N', 'AAG': 'I', 'AAC': 'G', 'TAT': 'E', 'GTG': 'L', 'GGC': 'H', 'CTT': '*', 'TTC': 'F', 'TGC': 'F', 'TGA': 'L', 'GTT': 'Q', 'CCT': 'Q', 'GGG': 'N', 'ATT': 'A', 'AGT': 'I', 'ACC': 'N', 'GAA': 'M', 'CGT': 'I', 'CAA': 'K', 'AAT': 'W', 'CGC': 'H', 'TCG': 'C', 'TTT': 'E', 'TGG': 'Y', 'CCC': 'S', 'TCC': 'K', 'GGA': 'P', 'TAC': 'T', 'TAG': 'M', 'ATA': 'D', 'CTC': 'L', 'ACA': 'T', 'ACT': 'Q', 'CCG': 'S', 'GAG': 'C', 'TAA': 'M', 'GTC': 'H', 'GAT': 'M', 'GAC': 'R', 'ACG': 'A', 'GCT': 'T', 'GTA': 'Y', 'AGA': 'R', 'ATG': 'Y', 'AGG': 'N', 'CAT': 'D', 'CCA': 'L', 'GGT': 'V', 'CGA': 'L', 'TTG': 'M', 'GCA': 'S', 'TCA': 'A', 'CAG': 'M', 'AGC': 'Q', 'AAA': 'T', 'TTA': 'A', 'GCG': 'A', 'ATC': 'S', 'CTG': 'M', 'TGT': 'L', 'TCT': 'Q', 'CTA': 'F', 'CAC': 'S', 'GCC': 'E'}
Best score: 10901.939353916418</code></pre>
</div>
</div></li>
<li><p>This has even more chance of randomly stumbling upon a valid genetic code, however in my testing it rarely gets any better than it did in 10,000 attempts. This <em>significantly</em> displays how unlikely the SGC’s highly optimized score is.</p></li>
<li><p>So let’s try something better.</p></li>
</ul>
</section>
<section id="simulated-annealing" class="level2 page-columns page-full" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="simulated-annealing"><span class="header-section-number">5.2</span> Simulated Annealing</h2>
<ul class="page-columns page-full">
<li class="page-columns page-full"><p>We discussed simulated annealing in class. I could NOT write this function myself, so I got Gemini 2.5 Pro to write it for me. My prompt was:</p>
<pre class="{markdown}"><code># PROMPT
I created an example "optimization" function that attempts to find the lowest score:

def random_optimization() -&gt; np.float64:
&nbsp; &nbsp; best: dict = create_genetic_code(STOP_CODON_LIMIT)
&nbsp; &nbsp; best_score = score_gc(best)
&nbsp; &nbsp; tries = 0
&nbsp; &nbsp; while True:
&nbsp; &nbsp; &nbsp; &nbsp; try:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; tries += 1
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; test = create_genetic_code(STOP_CODON_LIMIT)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; test_score = score_gc(test)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; if test_score &lt; best_score:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; best = test.copy()
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; best_score = test_score
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; print(f"New best! {best_score}")
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; print(best)
&nbsp; &nbsp; &nbsp; &nbsp; except KeyboardInterrupt:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; print(f"{tries} total tries")
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; sys.exit()

I would like a function similar to this that, rather than creating random genetic codes, generates one first and then performs "simulated annealing" to optimize the code intelligently.</code></pre>
<ul>
<li>And running it results in: <sup><a href="https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#7">AI Source</a></sup></li>
</ul>
<div id="0a953dfd" class="cell page-columns page-full" data-execution_count="21">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> simulated_annealing_optimization(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    stop_codon_limit: <span class="bu">int</span>, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    initial_temp: <span class="bu">float</span> <span class="op">=</span> <span class="fl">10000.0</span>, </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    cooling_rate: <span class="bu">float</span> <span class="op">=</span> <span class="fl">0.999</span>, </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    min_temp: <span class="bu">float</span> <span class="op">=</span> <span class="fl">1e-4</span>, </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    max_iterations: <span class="bu">int</span> <span class="op">=</span> <span class="dv">50000</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>) <span class="op">-&gt;</span> <span class="bu">tuple</span>[<span class="bu">list</span>, <span class="bu">list</span>, <span class="bu">list</span>]:</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co">    Optimizes a genetic code using a simulated annealing algorithm to minimize mutation cost.</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co">        stop_codon_limit: The maximum number of stop codons for creating new codes.</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="co">        initial_temp: The starting temperature for the annealing process.</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co">        cooling_rate: The rate at which the temperature decreases.</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="co">        min_temp: The temperature at which to stop the process.</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="co">        max_iterations: The maximum number of iterations to run.</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="co">        A tuple containing the final best score and a list of scores at each iteration.</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 1. Initialization</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>    current_code <span class="op">=</span> create_genetic_code(stop_codon_limit)</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>    current_score <span class="op">=</span> score_gc(current_code)</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    best_code <span class="op">=</span> current_code.copy()</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>    best_score <span class="op">=</span> current_score</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># List to track the score at every iteration</span></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>    score_history <span class="op">=</span> [current_score]</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Initial score: </span><span class="sc">{</span>best_score<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>    T <span class="op">=</span> initial_temp</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a>    iteration <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>        <span class="co"># 2. Main Loop</span></span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a>        <span class="cf">while</span> T <span class="op">&gt;</span> min_temp <span class="kw">and</span> iteration <span class="op">&lt;</span> max_iterations:</span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>            iteration <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>            <span class="co"># 3. Generate a Neighbor State</span></span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a>            new_code <span class="op">=</span> current_code.copy()</span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>            codons <span class="op">=</span> <span class="bu">list</span>(new_code.keys())</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a>            codon1, codon2 <span class="op">=</span> random.sample(codons, <span class="dv">2</span>)</span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a>            new_code[codon1], new_code[codon2] <span class="op">=</span> new_code[codon2], new_code[codon1]</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a>            <span class="co"># 4. Score the New State</span></span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a>            new_score <span class="op">=</span> score_gc(new_code)</span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a>            <span class="co"># 5. Decide Whether to Accept the New State</span></span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a>            score_delta <span class="op">=</span> new_score <span class="op">-</span> current_score</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> score_delta <span class="op">&lt;</span> <span class="dv">0</span> <span class="kw">or</span> random.random() <span class="op">&lt;</span> np.exp(<span class="op">-</span>score_delta <span class="op">/</span> T):</span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a>                current_code <span class="op">=</span> new_code</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a>                current_score <span class="op">=</span> new_score</span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Update the best-ever score if the current one is better</span></span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> current_score <span class="op">&lt;</span> best_score:</span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a>                best_score <span class="op">=</span> current_score</span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a>                best_code <span class="op">=</span> current_code.copy()</span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"Iter </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss"> | Temp </span><span class="sc">{</span>T<span class="sc">:.4f}</span><span class="ss"> | New best score: </span><span class="sc">{</span>best_score<span class="sc">:.4f}</span><span class="ss">"</span>)</span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Track the score of the accepted state for this iteration</span></span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a>            score_history.append(<span class="bu">float</span>(current_score))</span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a>            <span class="co"># 6. Cool Down</span></span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a>            T <span class="op">*=</span> cooling_rate</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> <span class="pp">KeyboardInterrupt</span>:</span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Optimization stopped by user."</span>)</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a>    <span class="cf">finally</span>:</span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\n</span><span class="ss">Finished after </span><span class="sc">{</span>iteration<span class="sc">}</span><span class="ss"> iterations."</span>)</span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Final best score: </span><span class="sc">{</span>best_score<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> (best_score, score_history, best_code)</span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a>best_found_score, score_runs, best_found_code <span class="op">=</span> simulated_annealing_optimization(<span class="dv">1</span>)</span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_optimization_progress(runs: <span class="bu">list</span>, sgc_score: <span class="bu">float</span>):</span>
<span id="cb25-80"><a href="#cb25-80" aria-hidden="true" tabindex="-1"></a>  plt.figure(figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">7</span>))</span>
<span id="cb25-81"><a href="#cb25-81" aria-hidden="true" tabindex="-1"></a>  plt.plot(runs, linestyle<span class="op">=</span><span class="st">'-'</span>, label<span class="op">=</span><span class="st">'Score per Iteration'</span>)</span>
<span id="cb25-82"><a href="#cb25-82" aria-hidden="true" tabindex="-1"></a>  plt.axhline(y<span class="op">=</span>sgc_score, color<span class="op">=</span><span class="st">'r'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, label<span class="op">=</span><span class="ss">f'SGC Score (</span><span class="sc">{</span>sgc_score<span class="sc">:.2f}</span><span class="ss">)'</span>)</span>
<span id="cb25-83"><a href="#cb25-83" aria-hidden="true" tabindex="-1"></a>  plt.title(<span class="st">'Simulated Annealing Optimization Progress'</span>)</span>
<span id="cb25-84"><a href="#cb25-84" aria-hidden="true" tabindex="-1"></a>  plt.xlabel(<span class="st">'Iteration'</span>)</span>
<span id="cb25-85"><a href="#cb25-85" aria-hidden="true" tabindex="-1"></a>  plt.ylabel(<span class="st">'Genetic Code Score'</span>)</span>
<span id="cb25-86"><a href="#cb25-86" aria-hidden="true" tabindex="-1"></a>  plt.legend()</span>
<span id="cb25-87"><a href="#cb25-87" aria-hidden="true" tabindex="-1"></a>  plt.grid(<span class="va">True</span>, which<span class="op">=</span><span class="st">'both'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb25-88"><a href="#cb25-88" aria-hidden="true" tabindex="-1"></a>  plt.show()</span>
<span id="cb25-89"><a href="#cb25-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-90"><a href="#cb25-90" aria-hidden="true" tabindex="-1"></a>sgc_baseline_score <span class="op">=</span> score_gc(SGC)</span>
<span id="cb25-91"><a href="#cb25-91" aria-hidden="true" tabindex="-1"></a>plot_optimization_progress(score_runs, sgc_baseline_score)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Initial score: 12372.079401860823
Iter 15 | Temp 9860.9064 | New best score: 12371.8847
Iter 212 | Temp 8096.8858 | New best score: 12351.4544
Iter 256 | Temp 7748.1764 | New best score: 12311.5163
Iter 257 | Temp 7740.4282 | New best score: 12179.7363
Iter 258 | Temp 7732.6878 | New best score: 12145.6108
Iter 698 | Temp 4979.0363 | New best score: 12117.9525
Iter 883 | Temp 4137.7154 | New best score: 12047.0104
Iter 899 | Temp 4072.0062 | New best score: 12037.7607
Iter 1590 | Temp 2039.6741 | New best score: 11920.5708
Iter 2064 | Temp 1269.4109 | New best score: 11761.9230
Iter 4112 | Temp 163.5769 | New best score: 11747.2051
Iter 4116 | Temp 162.9236 | New best score: 11679.7306
Iter 4120 | Temp 162.2729 | New best score: 11665.2736
Iter 4126 | Temp 161.3017 | New best score: 11631.1224
Iter 4132 | Temp 160.3363 | New best score: 11539.4542
Iter 4597 | Temp 100.6894 | New best score: 11506.9874
Iter 4816 | Temp 80.8771 | New best score: 11426.0930
Iter 4817 | Temp 80.7963 | New best score: 11371.8591
Iter 4820 | Temp 80.5541 | New best score: 11357.0222
Iter 4823 | Temp 80.3127 | New best score: 11344.2659
Iter 4826 | Temp 80.0720 | New best score: 11295.0432
Iter 4829 | Temp 79.8320 | New best score: 11255.5136
Iter 5199 | Temp 55.1325 | New best score: 11250.5399
Iter 5209 | Temp 54.5837 | New best score: 11248.2436
Iter 5244 | Temp 52.7054 | New best score: 11189.6858
Iter 5245 | Temp 52.6527 | New best score: 11112.4823
Iter 5250 | Temp 52.3899 | New best score: 11086.7683
Iter 5255 | Temp 52.1285 | New best score: 11070.5808
Iter 5258 | Temp 51.9723 | New best score: 10967.0100
Iter 5381 | Temp 45.9544 | New best score: 10928.6566
Iter 5446 | Temp 43.0609 | New best score: 10912.1827
Iter 5449 | Temp 42.9319 | New best score: 10837.8828
Iter 5460 | Temp 42.4620 | New best score: 10816.4406
Iter 5488 | Temp 41.2890 | New best score: 10808.8015
Iter 5557 | Temp 38.5348 | New best score: 10752.8332
Iter 5574 | Temp 37.8849 | New best score: 10743.7261
Iter 5590 | Temp 37.2832 | New best score: 10714.5227
Iter 5868 | Temp 28.2306 | New best score: 10713.8604
Iter 5872 | Temp 28.1178 | New best score: 10690.8884
Iter 5917 | Temp 26.8799 | New best score: 10598.3325
Iter 5941 | Temp 26.2422 | New best score: 10560.3261
Iter 5947 | Temp 26.0851 | New best score: 10483.9403
Iter 5955 | Temp 25.8772 | New best score: 10386.7225
Iter 6061 | Temp 23.2733 | New best score: 10380.2706
Iter 6063 | Temp 23.2268 | New best score: 10370.4082
Iter 6067 | Temp 23.1340 | New best score: 10345.4300
Iter 6079 | Temp 22.8580 | New best score: 10343.1673
Iter 6113 | Temp 22.0935 | New best score: 10289.9151
Iter 6158 | Temp 21.1208 | New best score: 10252.8197
Iter 6162 | Temp 21.0365 | New best score: 10232.9317
Iter 6180 | Temp 20.6610 | New best score: 10204.1849
Iter 6187 | Temp 20.5168 | New best score: 10162.7808
Iter 6194 | Temp 20.3736 | New best score: 10133.7249
Iter 6216 | Temp 19.9301 | New best score: 10114.5148
Iter 6265 | Temp 18.9766 | New best score: 10106.0416
Iter 6386 | Temp 16.8129 | New best score: 10062.5382
Iter 6389 | Temp 16.7625 | New best score: 10014.0053
Iter 6435 | Temp 16.0085 | New best score: 10001.2224
Iter 6469 | Temp 15.4731 | New best score: 9970.7323
Iter 6523 | Temp 14.6593 | New best score: 9927.6565
Iter 6577 | Temp 13.8883 | New best score: 9913.5893
Iter 6578 | Temp 13.8745 | New best score: 9898.7170
Iter 6604 | Temp 13.5182 | New best score: 9889.1792
Iter 6819 | Temp 10.9018 | New best score: 9889.0088
Iter 6842 | Temp 10.6538 | New best score: 9886.0255
Iter 6854 | Temp 10.5267 | New best score: 9875.2984
Iter 6860 | Temp 10.4637 | New best score: 9849.5208
Iter 6913 | Temp 9.9233 | New best score: 9842.7831
Iter 6961 | Temp 9.4580 | New best score: 9832.7832
Iter 6972 | Temp 9.3545 | New best score: 9828.7113
Iter 7002 | Temp 9.0779 | New best score: 9803.4960
Iter 7008 | Temp 9.0235 | New best score: 9798.6055
Iter 7049 | Temp 8.6609 | New best score: 9792.3815
Iter 7087 | Temp 8.3378 | New best score: 9761.2572
Iter 7132 | Temp 7.9707 | New best score: 9760.4469
Iter 7186 | Temp 7.5515 | New best score: 9659.8182
Iter 7192 | Temp 7.5063 | New best score: 9657.0629
Iter 7203 | Temp 7.4241 | New best score: 9652.1670
Iter 7292 | Temp 6.7916 | New best score: 9583.9723
Iter 7308 | Temp 6.6838 | New best score: 9568.2792
Iter 7328 | Temp 6.5514 | New best score: 9525.2606
Iter 7347 | Temp 6.4280 | New best score: 9519.1197
Iter 7428 | Temp 5.9276 | New best score: 9497.6751
Iter 7466 | Temp 5.7065 | New best score: 9490.8725
Iter 7534 | Temp 5.3312 | New best score: 9466.9073
Iter 7566 | Temp 5.1632 | New best score: 9464.9275
Iter 7576 | Temp 5.1118 | New best score: 9446.6144
Iter 7687 | Temp 4.5745 | New best score: 9438.7421
Iter 7926 | Temp 3.6016 | New best score: 9427.6479
Iter 7927 | Temp 3.5980 | New best score: 9398.6607
Iter 8189 | Temp 2.7683 | New best score: 9398.1500
Iter 8217 | Temp 2.6919 | New best score: 9384.2089
Iter 8229 | Temp 2.6597 | New best score: 9377.2121
Iter 8324 | Temp 2.4186 | New best score: 9358.1738
Iter 8442 | Temp 2.1492 | New best score: 9356.3802
Iter 8456 | Temp 2.1194 | New best score: 9331.5124
Iter 8479 | Temp 2.0711 | New best score: 9328.8700
Iter 8538 | Temp 1.9524 | New best score: 9320.0515
Iter 8585 | Temp 1.8627 | New best score: 9310.2985
Iter 8594 | Temp 1.8460 | New best score: 9307.7693
Iter 8646 | Temp 1.7525 | New best score: 9300.1163
Iter 8749 | Temp 1.5809 | New best score: 9289.9162
Iter 8750 | Temp 1.5793 | New best score: 9250.1911
Iter 8762 | Temp 1.5604 | New best score: 9245.6994
Iter 8796 | Temp 1.5082 | New best score: 9241.4399
Iter 9071 | Temp 1.1455 | New best score: 9196.8199
Iter 9258 | Temp 0.9500 | New best score: 9186.7269
Iter 9875 | Temp 0.5124 | New best score: 9174.7790
Iter 10175 | Temp 0.3796 | New best score: 9167.1638
Iter 10948 | Temp 0.1751 | New best score: 9129.4126
Iter 13020 | Temp 0.0220 | New best score: 9123.1791
Iter 13055 | Temp 0.0213 | New best score: 9095.3975

Finished after 18412 iterations.
Final best score: 9095.397502217338</code></pre>
</div>
<div class="cell-output cell-output-display page-columns page-full">
<div class="page-columns page-full">
<figure class="figure page-columns page-full">
<p class="page-columns page-full"><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-21-output-2.png" width="982" height="597" class="figure-img column-page"></p>
</figure>
</div>
</div>
</div></li>
<li><p>This is INCREDIBLY satisfying to watch happen, as once you hit a minimum point it <em>quickly</em> falls to a low value.</p></li>
<li><p>On the plot above, it’s clear that finding a code optimized better than the SGC isn’t that hard.</p></li>
<li><p>It’s important to note that this is only optimized by the standards of the scoring function. It’s possible that this code might suck by evolution’s true standards.</p></li>
<li class="page-columns page-full"><p>Let’s see what the codon map of this optimized code looks like:</p>
<div id="9e26d4f0" class="cell page-columns page-full" data-execution_count="22">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>format_genetic_code_table(best_found_code)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display column-page" data-execution_count="45">
<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Second Base</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">G</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">First Base</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">T</th>
<td>Glu (T)<br>
Gln (C)<br>
Asn (A)<br>
Asn (G)</td>
<td>Asp (T)<br>
Thr (C)<br>
Val (A)<br>
Thr (G)</td>
<td>Asp (T)<br>
Thr (C)<br>
Val (A)<br>
Cys (G)</td>
<td>Glu (T)<br>
Glu (C)<br>
Glu (A)<br>
Glu (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>Pro (T)<br>
Gly (C)<br>
Val (A)<br>
Gly (G)</td>
<td>Pro (T)<br>
Ala (C)<br>
Val (A)<br>
Thr (G)</td>
<td>Pro (T)<br>
Ala (C)<br>
Val (A)<br>
Cys (G)</td>
<td>Ser (T)<br>
Ser (C)<br>
Ser (A)<br>
Ser (G)</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>Trp (T)<br>
Met (C)<br>
Phe (A)<br>
Tyr (G)</td>
<td>Trp (T)<br>
Met (C)<br>
Ile (A)<br>
Tyr (G)</td>
<td>Trp (T)<br>
Met (C)<br>
Ile (A)<br>
Cys (G)</td>
<td>Trp (T)<br>
Met (C)<br>
Phe (A)<br>
Tyr (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>Lys (T)<br>
Gln (C)<br>
His (A)<br>
His (G)</td>
<td>Tyr (T)<br>
Tyr (C)<br>
Ile (A)<br>
His (G)</td>
<td>* (T)<br>
Leu (C)<br>
Ile (A)<br>
Cys (G)</td>
<td>Arg (T)<br>
Arg (C)<br>
Arg (A)<br>
Arg (G)</td>
</tr>
</tbody>
</table>
</div>
</div></li>
<li class="page-columns page-full"><p>And compared to the SCG:</p>
<div id="327bd29e" class="cell page-columns page-full" data-execution_count="23">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>format_genetic_code_table(SGC)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display column-page" data-execution_count="46">
<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Second Base</th>
<th data-quarto-table-cell-role="th">T</th>
<th data-quarto-table-cell-role="th">C</th>
<th data-quarto-table-cell-role="th">A</th>
<th data-quarto-table-cell-role="th">G</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">First Base</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">T</th>
<td>Phe (T)<br>
Phe (C)<br>
Leu (A)<br>
Leu (G)</td>
<td>Ser (T)<br>
Ser (C)<br>
Ser (A)<br>
Ser (G)</td>
<td>Tyr (T)<br>
Tyr (C)<br>
* (A)<br>
* (G)</td>
<td>Cys (T)<br>
Cys (C)<br>
* (A)<br>
Trp (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">C</th>
<td>Leu (T)<br>
Leu (C)<br>
Leu (A)<br>
Leu (G)</td>
<td>Pro (T)<br>
Pro (C)<br>
Pro (A)<br>
Pro (G)</td>
<td>His (T)<br>
His (C)<br>
Gln (A)<br>
Gln (G)</td>
<td>Arg (T)<br>
Arg (C)<br>
Arg (A)<br>
Arg (G)</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">A</th>
<td>Ile (T)<br>
Ile (C)<br>
Ile (A)<br>
Met (G)</td>
<td>Thr (T)<br>
Thr (C)<br>
Thr (A)<br>
Thr (G)</td>
<td>Asn (T)<br>
Asn (C)<br>
Lys (A)<br>
Lys (G)</td>
<td>Ser (T)<br>
Ser (C)<br>
Arg (A)<br>
Arg (G)</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">G</th>
<td>Val (T)<br>
Val (C)<br>
Val (A)<br>
Val (G)</td>
<td>Ala (T)<br>
Ala (C)<br>
Ala (A)<br>
Ala (G)</td>
<td>Asp (T)<br>
Asp (C)<br>
Glu (A)<br>
Glu (G)</td>
<td>Gly (T)<br>
Gly (C)<br>
Gly (A)<br>
Gly (G)</td>
</tr>
</tbody>
</table>
</div>
</div></li>
<li><p>Since the Quarto document generates the code output on each render, it is impossible to draw conclusions based on the specific code displayed in the current version. However, in my testing the optimized codes generally only posessed 1 stop codon, whereas the SGC has 3.</p>
<ul>
<li>This would likely due to pressures outside the scope of this assignment, perhaps for preventing against frameshift mutations by providing more options of stop codons.</li>
</ul></li>
<li><p>We can do some further analysis by comparing the amino acid frequency distribution between the SGC and the optimized code.</p></li>
<li class="page-columns page-full"><p>Let’s look at the SGC’s distribution:</p>
<div id="e8889232" class="cell page-columns page-full" data-execution_count="24">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>plot_aa_distribution(SGC)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display page-columns page-full">
<div class="page-columns page-full">
<figure class="figure page-columns page-full">
<p class="page-columns page-full"><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-24-output-1.png" width="1335" height="663" class="figure-img column-page"></p>
</figure>
</div>
</div>
</div></li>
<li class="page-columns page-full"><p>We can see Leucine, Argenine and Serine have the most codon mappings in the SGC. What about our optimized code?</p>
<div id="7e803dcf" class="cell page-columns page-full" data-execution_count="25">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>plot_aa_distribution(best_found_code)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display page-columns page-full">
<div class="page-columns page-full">
<figure class="figure page-columns page-full">
<p class="page-columns page-full"><img src="Arjun_Mehandiratta_Report_1_files/figure-html/cell-25-output-1.png" width="1335" height="663" class="figure-img column-page"></p>
</figure>
</div>
</div>
</div></li>
<li><p>Since Quarto rendering would find a new optimized code every time, it’s hard to predict what the distribution would exactly look like, but from my testing it’s generally a random mix, and doesn’t relate to the SGC’s distribution in any meaningful way</p>
<ul>
<li>This may be because there are other pressures involved in evolution but not looked at in this experiment which cause the SGC’s distribution to be optimal.</li>
</ul></li>
</ul>
</section>
</section>
<section id="conclusions" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Conclusions</h1>
<ul>
<li>The standard genetic code is clearly much better than a randomly generated genetic code.
<ul>
<li>This implies it was likely selected under evolutionary pressure for its ability to resist point-form mutations as tested in this code, among other criteria.</li>
</ul></li>
<li>However, by the metric we have chosen, it is definitely possible to find a more optimized code, with the earlier function able to find codes with scores lower than 8800.</li>
</ul>
<section id="limitations-of-the-experiment" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="limitations-of-the-experiment"><span class="header-section-number">6.1</span> Limitations of the experiment</h2>
<ul>
<li>It is definitely possible that these ‘optimized’ codes are not better than the SGC by other biological measures. A few of the limitations of this experiment (with some help identifying from the AI assistant) include:
<ul>
<li>The scoring function only gives low scores to codes that protect against <em>single point mutations</em>. It does not prevent against other mutations, like frameshifts or indels.</li>
<li>Amino acid similarity defined in <code>aaSim()</code> might not be the most optimal metric to measure.</li>
<li>Other biological pressures on the genetic code may exist, such as:
<ul>
<li>translational efficiency of the code</li>
<li>available resources (tRNA, proteins)</li>
</ul></li>
</ul></li>
</ul>
</section>
</section>
<section id="appendix" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Appendix</h1>
<section id="ai-chat-links" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="ai-chat-links"><span class="header-section-number">7.1</span> AI Chat Links</h2>
<ul>
<li><a href="https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#0">https://www.perplexity.ai/search/create-a-python-function-const-Y8zPZAC5ReKfUI1nNXVnYg#0</a></li>
<li><a href="https://www.perplexity.ai/search/i-will-provide-you-with-a-copy-8zXgHDAnQC62cn7lSM2tTw#0">https://www.perplexity.ai/search/i-will-provide-you-with-a-copy-8zXgHDAnQC62cn7lSM2tTw#0</a>
<ul>
<li>Model was always set to <code>Gemini 2.5 Pro</code></li>
</ul></li>
</ul>
</section>
<section id="preparing-the-report" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="preparing-the-report"><span class="header-section-number">7.2</span> Preparing the report</h2>
<ul>
<li>Preparing the report was relatively straightforward. I used my AI assistant to help me convert the dataset from R to Python. Then I used its help along with some of my own code to begin making aaSim(), and other functions.</li>
<li>Writing the code (90% of the report) was the easiest part since I had access to AI. Using extremely specific prompts helped a lot, and it rarely gave me a poor result; likely because of prompts
<ul>
<li>Working at a higher level is possible with AI but it definitely helps to still understand the lower levels of what’s going on so you can be specific in your prompts</li>
</ul></li>
<li>Most of my troubles were encountered in planning the assignment, and in understanding how Quarto works to translate what I wanted in my head into Quarto language.</li>
<li>I decided against using AI to write Quarto-specific syntax since I feel like I’ll be using this platform a lot in the future. I opted to read the docs instead! I’ve shockingly missed doing that…</li>
</ul>
</section>
<section id="references" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="references"><span class="header-section-number">7.3</span> References</h2>
<ul>
<li>
Kawashima, S. “AAindex: Amino Acid Index Database.” <i>Nucleic Acids Research</i> 28, no. 1 (January 1, 2000): 374. <a href="https://doi.org/10.1093/nar/28.1.374">https://doi.org/10.1093/nar/28.1.374</a>.
</li>
<li>
Kawashima, S., H. Ogata, and M. Kanehisa. “AAindex: Amino Acid Index Database.” <i>Nucleic Acids Research</i> 27, no. 1 (January 1, 1999): 368–69. <a href="https://doi.org/10.1093/nar/27.1.368">https://doi.org/10.1093/nar/27.1.368</a>.
</li>
<li>
Kawashima, S., P. Pokarowski, M. Pokarowska, A. Kolinski, T. Katayama, and M. Kanehisa. “AAindex: Amino Acid Index Database, Progress Report 2008.” <i>Nucleic Acids Research</i> 36, no. Database (November 13, 2007): D202–5. <a href="https://doi.org/10.1093/nar/gkm998">https://doi.org/10.1093/nar/gkm998</a>.
</li>
<li>
Nakai, Kenta, Akinori Kidera, and Minoru Kanehisa. “Cluster Analysis of Amino Acid Indices for Prediction of Protein Structure and Function.” <i>Protein Engineering Design and Selection</i> 2, no. 2 (January 1, 1988): 93–100. <a href="https://doi.org/10.1093/protein/2.2.93">https://doi.org/10.1093/protein/2.2.93</a>.
</li>
<li>
Tomii, Kentaro, and Minoru Kanehisa. “Analysis of Amino Acid Indices and Mutation Matrices for Sequence Comparison and Structure Prediction of Proteins.” <i>Protein Engineering Design and Selection</i> 9, no. 1 (January 1, 1996): 27–36. <a href="https://doi.org/10.1093/protein/9.1.27">https://doi.org/10.1093/protein/9.1.27</a>.
</li>
<li>
Boris Steipe. “GitHub - Hyginn/CSB195: Code and Data for CSB195 Computational Biology Foundations, University of Toronto.” GitHub, n.d. https://github.com/hyginn/CSB195. <a href="https://github.com/hyginn/CSB195">https://github.com/hyginn/CSB195</a>.
</li>
</ul>
</section>
<section id="self-assessment" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="self-assessment"><span class="header-section-number">7.4</span> Self-Assessment</h2>
<ul>
<li>90%</li>
</ul>
<table border="1">
<thead>
<tr>
<th>
Grade
</th>
<th>
Collaborative Improvement
</th>
<th>
Knowledge Accuracy &amp; Sources
</th>
<th>
Autonomous Voice
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Outstanding
</td>
<td>
Sophisticated iterative refinement transforms initial ideas and AI responses into original insights
</td>
<td>
All knowledge claims meticulously verified and sourced; novel contributions clearly identified and well-supported
</td>
<td style="background-color: #ffff99;">
Student’s intellectual agency drives the process; personal perspective unmistakably present; prompting shows strategic depth
</td>
</tr>
<tr>
<td>
Excellent
</td>
<td style="background-color: #ffff99;">
Clear evidence of improving ideas and AI outputs through thoughtful iteration and strategic redirection
</td>
<td style="background-color: #ffff99;">
Appropriate sourcing; novel insights generally well-identified and supported
</td>
<td>
Strong personal perspective maintained; clear student agency in guiding dialogue
</td>
</tr>
<tr>
<td>
Good
</td>
<td>
Some improvement of initial ideas and AI responses; basic understanding of refinement process
</td>
<td>
Most claims appropriately sourced; some novel contributions identified though presentation may lack clarity
</td>
<td>
Personal perspective present though inconsistent; some evidence of strategic engagement
</td>
</tr>
<tr>
<td>
Adequate
</td>
<td>
Minimal improvement of ideas or AI responses; limited understanding of iterative possibilities
</td>
<td>
Basic sourcing attempted with notable gaps; uncertain judgment about citation requirements
</td>
<td>
Weak personal perspective; limited agency in guiding collaboration
</td>
</tr>
<tr>
<td>
Marginal
</td>
<td>
No clear evidence of improving ideas or AI responses; accepts outputs without refinement
</td>
<td>
Inadequate sourcing; poor judgment about verification responsibilities
</td>
<td>
Student agency largely absent; minimal personal perspective
</td>
</tr>
<tr>
<td>
Inadequate
</td>
<td>
No improvement process evident; complete reliance on initial AI outputs
</td>
<td>
No source verification; factual claims accepted without responsibility
</td>
<td>
Complete absence of student agency; <em>work indistinguishable from pure AI output</em>
</td>
</tr>
</tbody>
</table>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>